{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import openai\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame, Series\n",
    "from sklearn.metrics.pairwise import  cosine_similarity\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import requests\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "sns.set_theme(style=\"whitegrid\")\n",
    "\n",
    "import os\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "\n",
    "openai.api_key = os.getenv(\"OPENAI_API_KEY\")\n",
    "HASURA_API_KEY = os.getenv(\"HASURA_API_KEY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.read_parquet(\"../data/scrna_seq_with_embs.pq\")\n",
    "df = df[~df.embedding.isna()].set_index('pubmed_id')\n",
    "df.index = df.index.map(lambda x: int(x.split(\"\\n\")[0]))\n",
    "embs = np.concatenate([np.atleast_2d(x) for x in df.embedding.values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 508 ms, sys: 103 ms, total: 611 ms\n",
      "Wall time: 11 s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "paps = requests.get(\n",
    "    \"https://sealver.in/hasura/api/rest/scrnaseq_papers/get\",\n",
    "    auth = requests.auth.HTTPBasicAuth('user', HASURA_API_KEY)\n",
    ")\n",
    "\n",
    "paps = DataFrame(paps.json()['ml_scrnaseq_papers']).set_index('pubmed_id')\n",
    "paps = paps[paps.is_computational][[\"title\", \"abstract\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.loc[paps.index]\n",
    "embs = np.concatenate([np.atleast_2d(x) for x in df.embedding.values])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "APIConnectionError",
     "evalue": "Error communicating with OpenAI",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mConnectionResetError\u001b[0m                      Traceback (most recent call last)",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/urllib3/connectionpool.py:703\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    702\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    704\u001b[0m     conn,\n\u001b[1;32m    705\u001b[0m     method,\n\u001b[1;32m    706\u001b[0m     url,\n\u001b[1;32m    707\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    708\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    709\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    710\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    711\u001b[0m )\n\u001b[1;32m    713\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    714\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    715\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    716\u001b[0m \u001b[39m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/urllib3/connectionpool.py:449\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    445\u001b[0m         \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    446\u001b[0m             \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    447\u001b[0m             \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    448\u001b[0m             \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n\u001b[0;32m--> 449\u001b[0m             six\u001b[39m.\u001b[39;49mraise_from(e, \u001b[39mNone\u001b[39;49;00m)\n\u001b[1;32m    450\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError, SocketError) \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[0;32m<string>:3\u001b[0m, in \u001b[0;36mraise_from\u001b[0;34m(value, from_value)\u001b[0m\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/urllib3/connectionpool.py:444\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    443\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 444\u001b[0m     httplib_response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49mgetresponse()\n\u001b[1;32m    445\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    446\u001b[0m     \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    447\u001b[0m     \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    448\u001b[0m     \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/http/client.py:1348\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1347\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1348\u001b[0m     response\u001b[39m.\u001b[39;49mbegin()\n\u001b[1;32m   1349\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/http/client.py:316\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    315\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m--> 316\u001b[0m     version, status, reason \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_read_status()\n\u001b[1;32m    317\u001b[0m     \u001b[39mif\u001b[39;00m status \u001b[39m!=\u001b[39m CONTINUE:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/http/client.py:277\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    276\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_read_status\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 277\u001b[0m     line \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfp\u001b[39m.\u001b[39;49mreadline(_MAXLINE \u001b[39m+\u001b[39;49m \u001b[39m1\u001b[39;49m), \u001b[39m\"\u001b[39m\u001b[39miso-8859-1\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    278\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(line) \u001b[39m>\u001b[39m _MAXLINE:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/socket.py:669\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    668\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 669\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sock\u001b[39m.\u001b[39;49mrecv_into(b)\n\u001b[1;32m    670\u001b[0m \u001b[39mexcept\u001b[39;00m timeout:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/ssl.py:1241\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1238\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m   1239\u001b[0m           \u001b[39m\"\u001b[39m\u001b[39mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m\n\u001b[1;32m   1240\u001b[0m           \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__class__\u001b[39m)\n\u001b[0;32m-> 1241\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mread(nbytes, buffer)\n\u001b[1;32m   1242\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/ssl.py:1099\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1098\u001b[0m \u001b[39mif\u001b[39;00m buffer \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m-> 1099\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sslobj\u001b[39m.\u001b[39;49mread(\u001b[39mlen\u001b[39;49m, buffer)\n\u001b[1;32m   1100\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "\u001b[0;31mConnectionResetError\u001b[0m: [Errno 104] Connection reset by peer",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mProtocolError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/requests/adapters.py:489\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    488\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 489\u001b[0m     resp \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49murlopen(\n\u001b[1;32m    490\u001b[0m         method\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mmethod,\n\u001b[1;32m    491\u001b[0m         url\u001b[39m=\u001b[39;49murl,\n\u001b[1;32m    492\u001b[0m         body\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mbody,\n\u001b[1;32m    493\u001b[0m         headers\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mheaders,\n\u001b[1;32m    494\u001b[0m         redirect\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    495\u001b[0m         assert_same_host\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    496\u001b[0m         preload_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    497\u001b[0m         decode_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    498\u001b[0m         retries\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmax_retries,\n\u001b[1;32m    499\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[1;32m    500\u001b[0m     )\n\u001b[1;32m    502\u001b[0m \u001b[39m# Send the request.\u001b[39;00m\n\u001b[1;32m    503\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/urllib3/connectionpool.py:787\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    785\u001b[0m     e \u001b[39m=\u001b[39m ProtocolError(\u001b[39m\"\u001b[39m\u001b[39mConnection aborted.\u001b[39m\u001b[39m\"\u001b[39m, e)\n\u001b[0;32m--> 787\u001b[0m retries \u001b[39m=\u001b[39m retries\u001b[39m.\u001b[39;49mincrement(\n\u001b[1;32m    788\u001b[0m     method, url, error\u001b[39m=\u001b[39;49me, _pool\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m, _stacktrace\u001b[39m=\u001b[39;49msys\u001b[39m.\u001b[39;49mexc_info()[\u001b[39m2\u001b[39;49m]\n\u001b[1;32m    789\u001b[0m )\n\u001b[1;32m    790\u001b[0m retries\u001b[39m.\u001b[39msleep()\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/urllib3/util/retry.py:550\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    549\u001b[0m \u001b[39mif\u001b[39;00m read \u001b[39mis\u001b[39;00m \u001b[39mFalse\u001b[39;00m \u001b[39mor\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_is_method_retryable(method):\n\u001b[0;32m--> 550\u001b[0m     \u001b[39mraise\u001b[39;00m six\u001b[39m.\u001b[39;49mreraise(\u001b[39mtype\u001b[39;49m(error), error, _stacktrace)\n\u001b[1;32m    551\u001b[0m \u001b[39melif\u001b[39;00m read \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/urllib3/packages/six.py:769\u001b[0m, in \u001b[0;36mreraise\u001b[0;34m(tp, value, tb)\u001b[0m\n\u001b[1;32m    768\u001b[0m \u001b[39mif\u001b[39;00m value\u001b[39m.\u001b[39m__traceback__ \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m tb:\n\u001b[0;32m--> 769\u001b[0m     \u001b[39mraise\u001b[39;00m value\u001b[39m.\u001b[39mwith_traceback(tb)\n\u001b[1;32m    770\u001b[0m \u001b[39mraise\u001b[39;00m value\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/urllib3/connectionpool.py:703\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    702\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    704\u001b[0m     conn,\n\u001b[1;32m    705\u001b[0m     method,\n\u001b[1;32m    706\u001b[0m     url,\n\u001b[1;32m    707\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    708\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    709\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    710\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    711\u001b[0m )\n\u001b[1;32m    713\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    714\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    715\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    716\u001b[0m \u001b[39m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/urllib3/connectionpool.py:449\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    445\u001b[0m         \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    446\u001b[0m             \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    447\u001b[0m             \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    448\u001b[0m             \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n\u001b[0;32m--> 449\u001b[0m             six\u001b[39m.\u001b[39;49mraise_from(e, \u001b[39mNone\u001b[39;49;00m)\n\u001b[1;32m    450\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError, SocketError) \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[0;32m<string>:3\u001b[0m, in \u001b[0;36mraise_from\u001b[0;34m(value, from_value)\u001b[0m\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/urllib3/connectionpool.py:444\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    443\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 444\u001b[0m     httplib_response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49mgetresponse()\n\u001b[1;32m    445\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    446\u001b[0m     \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    447\u001b[0m     \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    448\u001b[0m     \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/http/client.py:1348\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1347\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1348\u001b[0m     response\u001b[39m.\u001b[39;49mbegin()\n\u001b[1;32m   1349\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/http/client.py:316\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    315\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m--> 316\u001b[0m     version, status, reason \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_read_status()\n\u001b[1;32m    317\u001b[0m     \u001b[39mif\u001b[39;00m status \u001b[39m!=\u001b[39m CONTINUE:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/http/client.py:277\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    276\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_read_status\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 277\u001b[0m     line \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfp\u001b[39m.\u001b[39;49mreadline(_MAXLINE \u001b[39m+\u001b[39;49m \u001b[39m1\u001b[39;49m), \u001b[39m\"\u001b[39m\u001b[39miso-8859-1\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    278\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(line) \u001b[39m>\u001b[39m _MAXLINE:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/socket.py:669\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    668\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 669\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sock\u001b[39m.\u001b[39;49mrecv_into(b)\n\u001b[1;32m    670\u001b[0m \u001b[39mexcept\u001b[39;00m timeout:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/ssl.py:1241\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1238\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m   1239\u001b[0m           \u001b[39m\"\u001b[39m\u001b[39mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m\n\u001b[1;32m   1240\u001b[0m           \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__class__\u001b[39m)\n\u001b[0;32m-> 1241\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mread(nbytes, buffer)\n\u001b[1;32m   1242\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/ssl.py:1099\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1098\u001b[0m \u001b[39mif\u001b[39;00m buffer \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m-> 1099\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sslobj\u001b[39m.\u001b[39;49mread(\u001b[39mlen\u001b[39;49m, buffer)\n\u001b[1;32m   1100\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "\u001b[0;31mProtocolError\u001b[0m: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mConnectionError\u001b[0m                           Traceback (most recent call last)",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/openai/api_requestor.py:356\u001b[0m, in \u001b[0;36mAPIRequestor.request_raw\u001b[0;34m(self, method, url, params, supplied_headers, files, stream, request_id, request_timeout)\u001b[0m\n\u001b[1;32m    355\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 356\u001b[0m     result \u001b[39m=\u001b[39m _thread_context\u001b[39m.\u001b[39;49msession\u001b[39m.\u001b[39;49mrequest(\n\u001b[1;32m    357\u001b[0m         method,\n\u001b[1;32m    358\u001b[0m         abs_url,\n\u001b[1;32m    359\u001b[0m         headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    360\u001b[0m         data\u001b[39m=\u001b[39;49mdata,\n\u001b[1;32m    361\u001b[0m         files\u001b[39m=\u001b[39;49mfiles,\n\u001b[1;32m    362\u001b[0m         stream\u001b[39m=\u001b[39;49mstream,\n\u001b[1;32m    363\u001b[0m         timeout\u001b[39m=\u001b[39;49mrequest_timeout \u001b[39mif\u001b[39;49;00m request_timeout \u001b[39melse\u001b[39;49;00m TIMEOUT_SECS,\n\u001b[1;32m    364\u001b[0m     )\n\u001b[1;32m    365\u001b[0m \u001b[39mexcept\u001b[39;00m requests\u001b[39m.\u001b[39mexceptions\u001b[39m.\u001b[39mTimeout \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/requests/sessions.py:587\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    586\u001b[0m send_kwargs\u001b[39m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 587\u001b[0m resp \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(prep, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49msend_kwargs)\n\u001b[1;32m    589\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/requests/sessions.py:701\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m \u001b[39m# Send the request\u001b[39;00m\n\u001b[0;32m--> 701\u001b[0m r \u001b[39m=\u001b[39m adapter\u001b[39m.\u001b[39;49msend(request, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    703\u001b[0m \u001b[39m# Total elapsed time of the request (approximately)\u001b[39;00m\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/requests/adapters.py:547\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    546\u001b[0m \u001b[39mexcept\u001b[39;00m (ProtocolError, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m err:\n\u001b[0;32m--> 547\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m(err, request\u001b[39m=\u001b[39mrequest)\n\u001b[1;32m    549\u001b[0m \u001b[39mexcept\u001b[39;00m MaxRetryError \u001b[39mas\u001b[39;00m e:\n",
      "\u001b[0;31mConnectionError\u001b[0m: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mAPIConnectionError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[50], line 8\u001b[0m\n\u001b[1;32m      3\u001b[0m query \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mmeta-cells\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m      4\u001b[0m \u001b[39m# query = \"computational methods for analysis of case-control or cohort studies\"\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[39m# query = \"wavelet analysis\"\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[39m# query = \"epilepsy\"\u001b[39;00m\n\u001b[1;32m      7\u001b[0m \u001b[39m# query = \"Differential abundance testing on single-cell data using k-nearest neighbor graphs\"\u001b[39;00m\n\u001b[0;32m----> 8\u001b[0m response \u001b[39m=\u001b[39m openai\u001b[39m.\u001b[39;49mEmbedding\u001b[39m.\u001b[39;49mcreate(\u001b[39minput\u001b[39;49m\u001b[39m=\u001b[39;49mquery, model\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mtext-embedding-ada-002\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n\u001b[1;32m      9\u001b[0m query_emb \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39marray(response[\u001b[39m\"\u001b[39m\u001b[39mdata\u001b[39m\u001b[39m\"\u001b[39m][\u001b[39m0\u001b[39m][\u001b[39m\"\u001b[39m\u001b[39membedding\u001b[39m\u001b[39m\"\u001b[39m])\n\u001b[1;32m     10\u001b[0m query_sims \u001b[39m=\u001b[39m cosine_similarity(embs, np\u001b[39m.\u001b[39matleast_2d(query_emb))\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/openai/api_resources/embedding.py:34\u001b[0m, in \u001b[0;36mEmbedding.create\u001b[0;34m(cls, *args, **kwargs)\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m     33\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> 34\u001b[0m         response \u001b[39m=\u001b[39m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49mcreate(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m     36\u001b[0m         \u001b[39m# If a user specifies base64, we'll just return the encoded string.\u001b[39;00m\n\u001b[1;32m     37\u001b[0m         \u001b[39m# This is only for the default case.\u001b[39;00m\n\u001b[1;32m     38\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m user_provided_encoding_format:\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/openai/api_resources/abstract/engine_api_resource.py:115\u001b[0m, in \u001b[0;36mEngineAPIResource.create\u001b[0;34m(cls, api_key, api_base, api_type, request_id, api_version, organization, **params)\u001b[0m\n\u001b[1;32m    107\u001b[0m requestor \u001b[39m=\u001b[39m api_requestor\u001b[39m.\u001b[39mAPIRequestor(\n\u001b[1;32m    108\u001b[0m     api_key,\n\u001b[1;32m    109\u001b[0m     api_base\u001b[39m=\u001b[39mapi_base,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    112\u001b[0m     organization\u001b[39m=\u001b[39morganization,\n\u001b[1;32m    113\u001b[0m )\n\u001b[1;32m    114\u001b[0m url \u001b[39m=\u001b[39m \u001b[39mcls\u001b[39m\u001b[39m.\u001b[39mclass_url(engine, api_type, api_version)\n\u001b[0;32m--> 115\u001b[0m response, _, api_key \u001b[39m=\u001b[39m requestor\u001b[39m.\u001b[39;49mrequest(\n\u001b[1;32m    116\u001b[0m     \u001b[39m\"\u001b[39;49m\u001b[39mpost\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    117\u001b[0m     url,\n\u001b[1;32m    118\u001b[0m     params\u001b[39m=\u001b[39;49mparams,\n\u001b[1;32m    119\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    120\u001b[0m     stream\u001b[39m=\u001b[39;49mstream,\n\u001b[1;32m    121\u001b[0m     request_id\u001b[39m=\u001b[39;49mrequest_id,\n\u001b[1;32m    122\u001b[0m     request_timeout\u001b[39m=\u001b[39;49mrequest_timeout,\n\u001b[1;32m    123\u001b[0m )\n\u001b[1;32m    125\u001b[0m \u001b[39mif\u001b[39;00m stream:\n\u001b[1;32m    126\u001b[0m     \u001b[39m# must be an iterator\u001b[39;00m\n\u001b[1;32m    127\u001b[0m     \u001b[39massert\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(response, OpenAIResponse)\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/openai/api_requestor.py:171\u001b[0m, in \u001b[0;36mAPIRequestor.request\u001b[0;34m(self, method, url, params, headers, files, stream, request_id, request_timeout)\u001b[0m\n\u001b[1;32m    160\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mrequest\u001b[39m(\n\u001b[1;32m    161\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m    162\u001b[0m     method,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    169\u001b[0m     request_timeout: Optional[Union[\u001b[39mfloat\u001b[39m, Tuple[\u001b[39mfloat\u001b[39m, \u001b[39mfloat\u001b[39m]]] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    170\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Tuple[Union[OpenAIResponse, Iterator[OpenAIResponse]], \u001b[39mbool\u001b[39m, \u001b[39mstr\u001b[39m]:\n\u001b[0;32m--> 171\u001b[0m     result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mrequest_raw(\n\u001b[1;32m    172\u001b[0m         method\u001b[39m.\u001b[39;49mlower(),\n\u001b[1;32m    173\u001b[0m         url,\n\u001b[1;32m    174\u001b[0m         params\u001b[39m=\u001b[39;49mparams,\n\u001b[1;32m    175\u001b[0m         supplied_headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    176\u001b[0m         files\u001b[39m=\u001b[39;49mfiles,\n\u001b[1;32m    177\u001b[0m         stream\u001b[39m=\u001b[39;49mstream,\n\u001b[1;32m    178\u001b[0m         request_id\u001b[39m=\u001b[39;49mrequest_id,\n\u001b[1;32m    179\u001b[0m         request_timeout\u001b[39m=\u001b[39;49mrequest_timeout,\n\u001b[1;32m    180\u001b[0m     )\n\u001b[1;32m    181\u001b[0m     resp, got_stream \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_interpret_response(result, stream)\n\u001b[1;32m    182\u001b[0m     \u001b[39mreturn\u001b[39;00m resp, got_stream, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapi_key\n",
      "File \u001b[0;32m~/other/NLPProjects/GPTAreaOverview/env/lib/python3.8/site-packages/openai/api_requestor.py:368\u001b[0m, in \u001b[0;36mAPIRequestor.request_raw\u001b[0;34m(self, method, url, params, supplied_headers, files, stream, request_id, request_timeout)\u001b[0m\n\u001b[1;32m    366\u001b[0m     \u001b[39mraise\u001b[39;00m error\u001b[39m.\u001b[39mTimeout(\u001b[39m\"\u001b[39m\u001b[39mRequest timed out\u001b[39m\u001b[39m\"\u001b[39m) \u001b[39mfrom\u001b[39;00m \u001b[39me\u001b[39;00m\n\u001b[1;32m    367\u001b[0m \u001b[39mexcept\u001b[39;00m requests\u001b[39m.\u001b[39mexceptions\u001b[39m.\u001b[39mRequestException \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m--> 368\u001b[0m     \u001b[39mraise\u001b[39;00m error\u001b[39m.\u001b[39mAPIConnectionError(\u001b[39m\"\u001b[39m\u001b[39mError communicating with OpenAI\u001b[39m\u001b[39m\"\u001b[39m) \u001b[39mfrom\u001b[39;00m \u001b[39me\u001b[39;00m\n\u001b[1;32m    369\u001b[0m util\u001b[39m.\u001b[39mlog_info(\n\u001b[1;32m    370\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mOpenAI API response\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    371\u001b[0m     path\u001b[39m=\u001b[39mabs_url,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    374\u001b[0m     request_id\u001b[39m=\u001b[39mresult\u001b[39m.\u001b[39mheaders\u001b[39m.\u001b[39mget(\u001b[39m\"\u001b[39m\u001b[39mX-Request-Id\u001b[39m\u001b[39m\"\u001b[39m),\n\u001b[1;32m    375\u001b[0m )\n\u001b[1;32m    376\u001b[0m \u001b[39m# Don't read the whole stream for debug logging unless necessary.\u001b[39;00m\n",
      "\u001b[0;31mAPIConnectionError\u001b[0m: Error communicating with OpenAI"
     ]
    }
   ],
   "source": [
    "# query = \"non-negative matrix factorization (NMF)\"\n",
    "# query = \"comparative or case-control analysis\"\n",
    "query = \"meta-cells\"\n",
    "# query = \"computational methods for analysis of case-control or cohort studies\"\n",
    "# query = \"wavelet analysis\"\n",
    "# query = \"epilepsy\"\n",
    "# query = \"Differential abundance testing on single-cell data using k-nearest neighbor graphs\"\n",
    "response = openai.Embedding.create(input=query, model=\"text-embedding-ada-002\")\n",
    "query_emb = np.array(response[\"data\"][0][\"embedding\"])\n",
    "query_sims = cosine_similarity(embs, np.atleast_2d(query_emb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# plt.hist(query_sims, bins=50);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# query_sims = cosine_similarity(embs, np.atleast_2d(df.loc[34594043].embedding)) # Milo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.0\n",
      "# Differential abundance testing on single-cell data using k-nearest neighbor graphs.\n",
      "Current computational workflows for comparative analyses of single-cell datasets typically use discrete clusters as input when testing for differential abundance among experimental conditions. However, clusters do not always provide the appropriate resolution and cannot capture continuous trajectories. Here we present Milo, a scalable statistical framework that performs differential abundance testing by assigning cells to partially overlapping neighborhoods on a k-nearest neighbor graph. Using simulations and single-cell RNA sequencing (scRNA-seq) data, we show that Milo can identify perturbations that are obscured by discretizing cells into clusters, that it maintains false discovery rate control across batch effects and that it outperforms alternative differential abundance testing strategies. Milo identifies the decline of a fate-biased epithelial precursor in the aging mouse thymus and identifies perturbations to multiple lineages in human cirrhotic liver. As Milo is based on a cell-cell similarity structure, it might also be applicable to single-cell data other than scRNA-seq. Milo is provided as an open-source R software package at https://github.com/MarioniLab/miloR .\n",
      "\n",
      "0.874\n",
      "# MetaCell: analysis of single-cell RNA-seq data using K-nn graph partitions.\n",
      "scRNA-seq profiles each represent a highly partial sample of mRNA molecules from a unique cell that can never be resampled, and robust analysis must separate the sampling effect from biological variance. We describe a methodology for partitioning scRNA-seq datasets into metacells: disjoint and homogenous groups of profiles that could have been resampled from the same cell. Unlike clustering analysis, our algorithm specializes at obtaining granular as opposed to maximal groups. We show how to use metacells as building blocks for complex quantitative transcriptional maps while avoiding data smoothing. Our algorithms are implemented in the MetaCell R/C++ software package.\n",
      "\n",
      "0.868\n",
      "# scRNA-seq for Microcephaly Research [IV]: Dirichlet Regression for Single-Cell Population Differences.\n",
      "Analysis of single-cell RNA sequencing typically includes the clustering of cells and subsequent determination of the population size of each cluster, relative to the whole. In an experimental setting, two or more conditions are compared to assess changes in cellular composition of the sampled tissue. Cluster populations are frequently normalized to the total number of cells from each replicate in order to facilitate comparisons. After normalization, they become interdependent fractions and therefore cannot be compared using individual t-tests. Here we describe the use of Dirichlet regression to compare changes in cellular composition between two or more conditions when multiple biological replicates (three or more) are sampled under each condition. We provide an example of R code to conduct a similar analysis and interpret the results.\n",
      "\n",
      "0.867\n",
      "# Recommendations of scRNA-seq Differential Gene Expression Analysis Based on Comprehensive Benchmarking.\n",
      "To guide analysts to select the right tool and parameters in differential gene expression analyses of single-cell RNA sequencing (scRNA-seq) data, we developed a novel simulator that recapitulates the data characteristics of real scRNA-seq datasets while accounting for all the relevant sources of variation in a multi-subject, multi-condition scRNA-seq experiment: the cell-to-cell variation within a subject, the variation across subjects, the variability across cell types, the mean/variance relationship of gene expression across genes, library size effects, group effects, and covariate effects. By applying it to benchmark 12 differential gene expression analysis methods (including cell-level and pseudo-bulk methods) on simulated multi-condition, multi-subject data of the 10x Genomics platform, we demonstrated that methods originating from the negative binomial mixed model such as glmmTMB and NEBULA-HL outperformed other methods. Utilizing NEBULA-HL in a statistical analysis pipeline for single-cell analysis will enable scientists to better understand the cell-type-specific transcriptomic response to disease or treatment effects and to discover new drug targets. Further, application to two real datasets showed the outperformance of our differential expression (DE) pipeline, with unified findings of differentially expressed genes (DEG) and a pseudo-time trajectory transcriptomic result. In the end, we made recommendations for filtering strategies of cells and genes based on simulation results to achieve optimal experimental goals.\n",
      "\n",
      "0.865\n",
      "# Differential gene expression analysis for multi-subject single cell RNA sequencing studies with aggregateBioVar.\n",
      "Single cell RNA sequencing (scRNA-seq) studies provide more granular biological information than bulk RNA sequencing, but bulk RNA sequencing has remained popular due to relatively lower costs per sample, which has allowed investigators to process more biological replicates and design more powerful studies. As scRNA-seq costs have decreased, collecting data from more than one biological replicate has become more feasible, but careful modeling of different layers of biological variation remains challenging for many users. Here, we propose a statistical model for scRNA-seq gene counts, describe a simple method for estimating model parameters, and show that failing to account for additional biological variation in scRNA-seq studies can inflate false discovery rates of statistical tests.\n",
      "In a simulation study, we show that when the gene expression distribution of a population of cells varies between subjects, a naïve approach to differential expression analysis will inflate the false discovery rate. We also compare multiple differential expression testing methods on scRNA-seq data sets from human samples and from animal models. These analyses suggest that a naïve approach to differential expression testing could lead to many false discoveries; in contrast, an approach based on pseudobulk counts has better false discovery rate control.\n",
      "A software package, aggregateBioVar, is freely available on Bioconductor (https://www.bioconductor.org/packages/release/bioc/html/aggregateBioVar.html) to accommodate compatibility with upstream and downstream methods in scRNA-seq data analysis pipelines.\n",
      "Raw gene-by-cell count matrices for pig scRNA-seq data are available as GEO accession GSE150211. Supplementary data are available at Bioinformatics online.\n",
      "\n",
      "0.864\n",
      "# A statistical approach for identifying differential distributions in single-cell RNA-seq experiments.\n",
      "The ability to quantify cellular heterogeneity is a major advantage of single-cell technologies. However, statistical methods often treat cellular heterogeneity as a nuisance. We present a novel method to characterize differences in expression in the presence of distinct expression states within and among biological conditions. We demonstrate that this framework can detect differential expression patterns under a wide range of settings. Compared to existing approaches, this method has higher power to detect subtle differences in gene expression distributions that are more complex than a mean shift, and can characterize those differences. The freely available R package scDD implements the approach.\n",
      "\n",
      "0.864\n",
      "# A Hybrid Clustering Algorithm for Identifying Cell Types from Single-Cell RNA-Seq Data.\n",
      "Single-cell RNA sequencing (scRNA-seq) has recently brought new insight into cell differentiation processes and functional variation in cell subtypes from homogeneous cell populations. A lack of prior knowledge makes unsupervised machine learning methods, such as clustering, suitable for analyzing scRNA-seq . However, there are several limitations to overcome, including high dimensionality, clustering result instability, and parameter adjustment complexity. In this study, we propose a method by combining structure entropy and k nearest neighbor to identify cell subpopulations in scRNA-seq data. In contrast to existing clustering methods for identifying cell subtypes, minimized structure entropy results in natural communities without specifying the number of clusters. To investigate the performance of our model, we applied it to eight scRNA-seq datasets and compared our method with three existing methods (nonnegative matrix factorization, single-cell interpretation via multikernel learning, and structural entropy minimization principle). The experimental results showed that our approach achieves, on average, better performance in these datasets compared to the benchmark methods.\n",
      "\n",
      "0.862\n",
      "# Bioconductor workflow for single-cell RNA sequencing: Normalization, dimensionality reduction, clustering, and lineage inference.\n",
      "Novel single-cell transcriptome sequencing assays allow researchers to measure gene expression levels at the resolution of single cells and offer the unprecendented opportunity to investigate at the molecular level fundamental biological questions, such as stem cell differentiation or the discovery and characterization of rare cell types. However, such assays raise challenging statistical and computational questions and require the development of novel methodology and software. Using stem cell differentiation in the mouse olfactory epithelium as a case study, this integrated workflow provides a step-by-step tutorial to the methodology and associated software for the following four main tasks: (1) dimensionality reduction accounting for zero inflation and over dispersion and adjusting for gene and cell-level covariates; (2) cell clustering using resampling-based sequential ensemble clustering; (3) inference of cell lineages and pseudotimes; and (4) differential expression analysis along lineages.\n",
      "\n",
      "0.861\n",
      "# A discriminative learning approach to differential expression analysis for single-cell RNA-seq.\n",
      "Single-cell RNA-seq makes it possible to characterize the transcriptomes of cell types across different conditions and to identify their transcriptional signatures via differential analysis. Our method detects changes in transcript dynamics and in overall gene abundance in large numbers of cells to determine differential expression. When applied to transcript compatibility counts obtained via pseudoalignment, our approach provides a quantification-free analysis of 3' single-cell RNA-seq that can identify previously undetectable marker genes.\n",
      "\n",
      "0.86\n",
      "# MarcoPolo: a method to discover differentially expressed genes in single-cell RNA-seq data without depending on prior clustering.\n",
      "The standard analysis pipeline for single-cell RNA-seq data consists of sequential steps initiated by clustering the cells. An innate limitation of this pipeline is that an imperfect clustering result can irreversibly affect the succeeding steps. For example, there can be cell types not well distinguished by clustering because they largely share the global structure, such as the anterior primitive streak and mid primitive streak cells. If one searches differentially expressed genes (DEGs) solely based on clustering, marker genes for distinguishing these types will be missed. Moreover, clustering depends on many parameters and can often be subjective to manual decisions. To overcome these limitations, we propose MarcoPolo, a method that identifies informative DEGs independently of prior clustering. MarcoPolo sorts out genes by evaluating if the distributions are bimodal, if similar expression patterns are observed in other genes, and if the expressing cells are proximal in a low-dimensional space. Using real datasets with FACS-purified cell labels, we demonstrate that MarcoPolo recovers marker genes better than competing methods. Notably, MarcoPolo finds key genes that can distinguish cell types that are not distinguishable by the standard clustering. MarcoPolo is built in a convenient software package that provides analysis results in an HTML file.\n",
      "\n",
      "0.859\n",
      "# Bias, robustness and scalability in single-cell differential expression analysis.\n",
      "Many methods have been used to determine differential gene expression from single-cell RNA (scRNA)-seq data. We evaluated 36 approaches using experimental and synthetic data and found considerable differences in the number and characteristics of the genes that are called differentially expressed. Prefiltering of lowly expressed genes has important effects, particularly for some of the methods developed for bulk RNA-seq data analysis. However, we found that bulk RNA-seq analysis methods do not generally perform worse than those developed specifically for scRNA-seq. We also present conquer, a repository of consistently processed, analysis-ready public scRNA-seq data sets that is aimed at simplifying method evaluation and reanalysis of published results. Each data set provides abundance estimates for both genes and transcripts, as well as quality control and exploratory analysis reports.\n",
      "\n",
      "0.859\n",
      "# Differential Expression Analysis in Single-Cell Transcriptomics.\n",
      "Differential expression analysis is an important aspect of bulk RNA sequencing (RNAseq). A lot of tools are available, and among them DESeq2 and edgeR are widely used. Since single-cell RNA sequencing (scRNAseq) expression data are zero inflated, single-cell data are quite different from those generated by conventional bulk RNA sequencing. Comparative analysis of tools used to detect differentially expressed genes between two groups of single cells showed that edgeR with quasi-likelihood F-test (QLF) outperforms other methods.In bulk RNAseq, differential expression is mainly used to compare limited number of replicates of two or more biological conditions. However, scRNAseq differential expression analysis might be also instrumental to identify the main players of cells subpopulation organization, thus requiring the use of multiple comparisons tools. Nowadays, edgeR is one of the few tools that are able to handle both zero inflated matrices and multiple comparisons. Here, we provide a guide to the use of edgeR as a tool to detect differential expression in single-cell data.\n",
      "\n",
      "0.858\n",
      "# Multi-Objective Optimized Fuzzy Clustering for Detecting Cell Clusters from Single-Cell Expression Profiles.\n",
      "Rapid advance in single-cell RNA sequencing (scRNA-seq) allows measurement of the expression of genes at single-cell resolution in complex disease or tissue. While many methods have been developed to detect cell clusters from the scRNA-seq data, this task currently remains a main challenge. We proposed a multi-objective optimization-based fuzzy clustering approach for detecting cell clusters from scRNA-seq data. First, we conducted initial filtering and SCnorm normalization. We considered various case studies by selecting different cluster numbers ( c l = 2 to a user-defined number), and applied fuzzy c-means clustering algorithm individually. From each case, we evaluated the scores of four cluster validity index measures, Partition Entropy ( P E ), Partition Coefficient ( P C ), Modified Partition Coefficient ( M P C ), and Fuzzy Silhouette Index ( F S I ). Next, we set the first measure as minimization objective (↓) and the remaining three as maximization objectives (↑), and then applied a multi-objective decision-making technique, TOPSIS, to identify the best optimal solution. The best optimal solution (case study) that had the highest TOPSIS score was selected as the final optimal clustering. Finally, we obtained differentially expressed genes (DEGs) using Limma through the comparison of expression of the samples between each resultant cluster and the remaining clusters. We applied our approach to a scRNA-seq dataset for the rare intestinal cell type in mice [GEO ID: GSE62270, 23,630 features (genes) and 288 cells]. The optimal cluster result (TOPSIS optimal score= 0.858) comprised two clusters, one with 115 cells and the other 91 cells. The evaluated scores of the four cluster validity indices, F S I , P E , P C , and M P C for the optimized fuzzy clustering were 0.482, 0.578, 0.607, and 0.215, respectively. The Limma analysis identified 1240 DEGs (cluster 1 vs. cluster 2). The top ten gene markers were \n",
      "\n",
      "0.858\n",
      "# Single-cell RNA sequencing data analysis based on non-uniform ε-neighborhood network.\n",
      "Single-cell RNA sequencing (scRNA-seq) technology provides the possibility to study cell heterogeneity and cell development on the resolution of individual cells. Arguably, three of the most important computational targets on scRNA-seq data analysis are data visualization, cell clustering and trajectory inference. Although a substantial number of algorithms have been developed, most of them do not treat the three targets in a systematic or consistent manner.\n",
      "In this article, we propose an efficient scRNA-seq analysis framework, which accomplishes the three targets consistently by non-uniform ε-neighborhood (NEN) network. First, a network is generated by our NEN method, which combines the advantages of both k-nearest neighbors (KNN) and ε-neighborhood (EN) to represent the manifold that data points reside in gene space. Then from such a network, we use its layout, its community and further its shortest path to achieve the purpose of scRNA-seq data visualization, clustering and trajectory inference. The results on both synthetic and real datasets indicate that our NEN method not only can visually provide the global topological structure of a dataset accurately compared with t-SNE (t-Distributed Stochastic Neighbor Embedding) and UMAP (Uniform Manifold Approximation and Projection), but also has superior performances on clustering and pseudotime ordering of cells over the existing approaches.\n",
      "This analysis method has been made into a python package called ccnet and is freely available at https://github.com/Just-Jia/ccNet.\n",
      "Supplementary data are available at Bioinformatics online.\n",
      "\n",
      "0.858\n",
      "# Latent cellular analysis robustly reveals subtle diversity in large-scale single-cell RNA-seq data.\n",
      "Single-cell RNA sequencing (scRNA-seq) is a powerful tool for characterizing the cell-to-cell variation and cellular dynamics in populations which appear homogeneous otherwise in basic and translational biological research. However, significant challenges arise in the analysis of scRNA-seq data, including the low signal-to-noise ratio with high data sparsity, potential batch effects, scalability problems when hundreds of thousands of cells are to be analyzed among others. The inherent complexities of scRNA-seq data and dynamic nature of cellular processes lead to suboptimal performance of many currently available algorithms, even for basic tasks such as identifying biologically meaningful heterogeneous subpopulations. In this study, we developed the Latent Cellular Analysis (LCA), a machine learning-based analytical pipeline that combines cosine-similarity measurement by latent cellular states with a graph-based clustering algorithm. LCA provides heuristic solutions for population number inference, dimension reduction, feature selection, and control of technical variations without explicit gene filtering. We show that LCA is robust, accurate, and powerful by comparison with multiple state-of-the-art computational methods when applied to large-scale real and simulated scRNA-seq data. Importantly, the ability of LCA to learn from representative subsets of the data provides scalability, thereby addressing a significant challenge posed by growing sample sizes in scRNA-seq data analysis.\n",
      "\n",
      "0.858\n",
      "# Machine learning and statistical methods for clustering single-cell RNA-sequencing data.\n",
      "Single-cell RNAsequencing (scRNA-seq) technologies have enabled the large-scale whole-transcriptome profiling of each individual single cell in a cell population. A core analysis of the scRNA-seq transcriptome profiles is to cluster the single cells to reveal cell subtypes and infer cell lineages based on the relations among the cells. This article reviews the machine learning and statistical methods for clustering scRNA-seq transcriptomes developed in the past few years. The review focuses on how conventional clustering techniques such as hierarchical clustering, graph-based clustering, mixture models, $k$-means, ensemble learning, neural networks and density-based clustering are modified or customized to tackle the unique challenges in scRNA-seq data analysis, such as the dropout of low-expression genes, low and uneven read coverage of transcripts, highly variable total mRNAs from single cells and ambiguous cell markers in the presence of technical biases and irrelevant confounding biological variations. We review how cell-specific normalization, the imputation of dropouts and dimension reduction methods can be applied with new statistical or optimization strategies to improve the clustering of single cells. We will also introduce those more advanced approaches to cluster scRNA-seq transcriptomes in time series data and multiple cell populations and to detect rare cell types. Several software packages developed to support the cluster analysis of scRNA-seq data are also reviewed and experimentally compared to evaluate their performance and efficiency. Finally, we conclude with useful observations and possible future directions in scRNA-seq data analytics.\n",
      "All the source code and data are available at https://github.com/kuanglab/single-cell-review.\n",
      "\n",
      "0.856\n",
      "# Riemannian geometry and statistical modeling correct for batch effects and control false discoveries in single-cell surface protein count data.\n",
      "Recent advances in next generation sequencing-based single-cell technologies have allowed high-throughput quantitative detection of cell-surface proteins along with the transcriptome in individual cells, extending our understanding of the heterogeneity of cell populations in diverse tissues that are in different diseased states or under different experimental conditions. Count data of surface proteins from the cellular indexing of transcriptomes and epitopes by sequencing (CITE-seq) technology pose new computational challenges, and there is currently a dearth of rigorous mathematical tools for analyzing the data. This work utilizes concepts and ideas from Riemannian geometry to remove batch effects between samples and develops a statistical framework for distinguishing positive signals from background noise. The strengths of these approaches are demonstrated on two independent CITE-seq data sets in mouse and human.\n",
      "\n",
      "0.856\n",
      "# SCell: integrated analysis of single-cell RNA-seq data.\n",
      "Analysis of the composition of heterogeneous tissue has been greatly enabled by recent developments in single-cell transcriptomics. We present SCell, an integrated software tool for quality filtering, normalization, feature selection, iterative dimensionality reduction, clustering and the estimation of gene-expression gradients from large ensembles of single-cell RNA-seq datasets. SCell is open source, and implemented with an intuitive graphical interface. Scripts and protocols for the high-throughput pre-processing of large ensembles of single-cell, RNA-seq datasets are provided as an additional resource.\n",
      "Binary executables for Windows, MacOS and Linux are available at http://sourceforge.net/projects/scell, source code and pre-processing scripts are available from https://github.com/diazlab/SCellSupplementary information: Supplementary data are available at Bioinformatics online.\n",
      "aaron.diaz@ucsf.edu.\n",
      "\n",
      "0.856\n",
      "# DENDRO: genetic heterogeneity profiling and subclone detection by single-cell RNA sequencing.\n",
      "Although scRNA-seq is now ubiquitously adopted in studies of intratumor heterogeneity, detection of somatic mutations and inference of clonal membership from scRNA-seq is currently unreliable. We propose DENDRO, an analysis method for scRNA-seq data that clusters single cells into genetically distinct subclones and reconstructs the phylogenetic tree relating the subclones. DENDRO utilizes transcribed point mutations and accounts for technical noise and expression stochasticity. We benchmark DENDRO and demonstrate its application on simulation data and real data from three cancer types. In particular, on a mouse melanoma model in response to immunotherapy, DENDRO delineates the role of neoantigens in treatment response.\n",
      "\n",
      "0.856\n",
      "# Single-cell mRNA quantification and differential analysis with Census.\n",
      "Single-cell gene expression studies promise to reveal rare cell types and cryptic states, but the high variability of single-cell RNA-seq measurements frustrates efforts to assay transcriptional differences between cells. We introduce the Census algorithm to convert relative RNA-seq expression levels into relative transcript counts without the need for experimental spike-in controls. Analyzing changes in relative transcript counts led to dramatic improvements in accuracy compared to normalized read counts and enabled new statistical tests for identifying developmentally regulated genes. Census counts can be analyzed with widely used regression techniques to reveal changes in cell-fate-dependent gene expression, splicing patterns and allelic imbalances. We reanalyzed single-cell data from several developmental and disease studies, and demonstrate that Census enabled robust analysis at multiple layers of gene regulation. Census is freely available through our updated single-cell analysis toolkit, Monocle 2.\n",
      "\n",
      "0.856\n",
      "# Effective single-cell clustering through ensemble feature selection and similarity measurements.\n",
      "Single-cell RNA sequencing technologies have revolutionized biomedical research by providing an effective means to profile gene expressions in individual cells. One of the first fundamental steps to perform the in-depth analysis of single-cell sequencing data is cell type classification and identification. Computational methods such as clustering algorithms have been utilized and gaining in popularity because they can save considerable resources and time for experimental validations. Although selecting the optimal features (i.e., genes) is an essential process to obtain accurate and reliable single-cell clustering results, the computational complexity and dropout events that can introduce zero-inflated noise make this process very challenging. In this paper, we propose an effective single-cell clustering algorithm based on the ensemble feature selection and similarity measurements. We initially identify the set of potential features, then measure the cell-to-cell similarity based on the subset of the potentials through multiple feature sampling approaches. We construct the ensemble network based on cell-to-cell similarity. Finally, we apply a network-based clustering algorithm to obtain single-cell clusters. We evaluate the performance of our proposed algorithm through multiple assessments in real-world single-cell RNA sequencing datasets with known cell types. The results show that our proposed algorithm can identify accurate and consistent single-cell clustering. Moreover, the proposed algorithm takes relative expression as input, so it can easily be adopted by existing analysis pipelines. The source code has been made publicly available at https://github.com/jeonglab/scCLUE.\n",
      "\n",
      "0.856\n",
      "# A Novel Method to Identify the Differences Between Two Single Cell Groups at Single Gene, Gene Pair, and Gene Module Levels.\n",
      "Single-cell sequencing technology can not only view the heterogeneity of cells from a molecular perspective, but also discover new cell types. Although there are many effective methods on dropout imputation, cell clustering, and lineage reconstruction based on single cell RNA sequencing (RNA-seq) data, there is no systemic pipeline on how to compare two single cell clusters at the molecular level. In the study, we present a novel pipeline on comparing two single cell clusters, including calling differential gene expression, coexpression network modules, and so on. The pipeline could reveal mechanisms behind the biological difference between cell clusters and cell types, and identify cell type specific molecular mechanisms. We applied the pipeline to two famous single-cell databases, Usoskin from mouse brain and Xin from human pancreas, which contained 622 and 1,600 cells, respectively, both of which were composed of four types of cells. As a result, we identified many significant differential genes, differential gene coexpression and network modules among the cell clusters, which confirmed that different cell clusters might perform different functions.\n",
      "\n",
      "0.856\n",
      "# Evaluating the performance of dropout imputation and clustering methods for single-cell RNA sequencing data.\n",
      "Recent advances in single-cell RNA sequencing (scRNA-seq) provide exciting opportunities for transcriptome analysis at single-cell resolution. Clustering individual cells is a key step to reveal cell subtypes and infer cell lineage in scRNA-seq analysis. Although many dedicated algorithms have been proposed, clustering quality remains a computational challenge for scRNA-seq data, which is exacerbated by inflated zero counts due to various technical noise. To address this challenge, we assess the combinations of nine popular dropout imputation methods and eight clustering methods on a collection of 10 well-annotated scRNA-seq datasets with different sample sizes. Our results show that (i) imputation algorithms do typically improve the performance of clustering methods, and the quality of data visualization using t-Distributed Stochastic Neighbor Embedding; and (ii) the performance of a particular combination of imputation and clustering methods varies with dataset size. For example, the combination of single-cell analysis via expression recovery and Sparse Subspace Clustering (SSC) methods usually works well on smaller datasets, while the combination of adaptively-thresholded low-rank approximation and single-cell interpretation via multikernel learning (SIMLR) usually achieves the best performance on larger datasets.\n",
      "\n",
      "0.855\n",
      "# Impact of similarity metrics on single-cell RNA-seq data clustering.\n",
      "Advances in high-throughput sequencing on single-cell gene expressions [single-cell RNA sequencing (scRNA-seq)] have enabled transcriptome profiling on individual cells from complex samples. A common goal in scRNA-seq data analysis is to discover and characterise cell types, typically through clustering methods. The quality of the clustering therefore plays a critical role in biological discovery. While numerous clustering algorithms have been proposed for scRNA-seq data, fundamentally they all rely on a similarity metric for categorising individual cells. Although several studies have compared the performance of various clustering algorithms for scRNA-seq data, currently there is no benchmark of different similarity metrics and their influence on scRNA-seq data clustering. Here, we compared a panel of similarity metrics on clustering a collection of annotated scRNA-seq datasets. Within each dataset, a stratified subsampling procedure was applied and an array of evaluation measures was employed to assess the similarity metrics. This produced a highly reliable and reproducible consensus on their performance assessment. Overall, we found that correlation-based metrics (e.g. Pearson's correlation) outperformed distance-based metrics (e.g. Euclidean distance). To test if the use of correlation-based metrics can benefit the recently published clustering techniques for scRNA-seq data, we modified a state-of-the-art kernel-based clustering algorithm (SIMLR) using Pearson's correlation as a similarity measure and found significant performance improvement over Euclidean distance on scRNA-seq data clustering. These findings demonstrate the importance of similarity metrics in clustering scRNA-seq data and highlight Pearson's correlation as a favourable choice. Further comparison on different scRNA-seq library preparation protocols suggests that they may also affect clustering performance. Finally, the benchmarking framework is available at http://www.maths.usyd.edu.au/u/SMS/bioinformatics/software.html.\n",
      "\n",
      "0.855\n",
      "# scdNet: a computational tool for single-cell differential network analysis.\n",
      "Single-cell RNA sequencing (scRNA-Seq) is an emerging technology that has revolutionized the research of the tumor heterogeneity. However, the highly sparse data matrices generated by the technology have posed an obstacle to the analysis of differential gene regulatory networks.\n",
      "Addressing the challenges, this study presents, as far as we know, the first bioinformatics tool for scRNA-Seq-based differential network analysis (scdNet). The tool features a sample size adjustment of gene-gene correlation, comparison of inter-state correlations, and construction of differential networks. A simulation analysis demonstrated the power of scdNet in the analyses of sparse scRNA-Seq data matrices, with low requirement on the sample size, high computation efficiency, and tolerance of sequencing noises. Applying the tool to analyze two datasets of single circulating tumor cells (CTCs) of prostate cancer and early mouse embryos, our data demonstrated that differential gene regulation plays crucial roles in anti-androgen resistance and early embryonic development.\n",
      "Overall, the tool is widely applicable to datasets generated by the emerging technology to bring biological insights into tumor heterogeneity and other studies. MATLAB implementation of scdNet is available at https://github.com/ChenLabGCCRI/scdNet .\n",
      "\n",
      "0.854\n",
      "# Resolving single-cell heterogeneity from hundreds of thousands of cells through sequential hybrid clustering and NMF.\n",
      "The rapid proliferation of single-cell RNA-sequencing (scRNA-Seq) technologies has spurred the development of diverse computational approaches to detect transcriptionally coherent populations. While the complexity of the algorithms for detecting heterogeneity has increased, most require significant user-tuning, are heavily reliant on dimension reduction techniques and are not scalable to ultra-large datasets. We previously described a multi-step algorithm, Iterative Clustering and Guide-gene Selection (ICGS), which applies intra-gene correlation and hybrid clustering to uniquely resolve novel transcriptionally coherent cell populations from an intuitive graphical user interface.\n",
      "We describe a new iteration of ICGS that outperforms state-of-the-art scRNA-Seq detection workflows when applied to well-established benchmarks. This approach combines multiple complementary subtype detection methods (HOPACH, sparse non-negative matrix factorization, cluster 'fitness', support vector machine) to resolve rare and common cell-states, while minimizing differences due to donor or batch effects. Using data from multiple cell atlases, we show that the PageRank algorithm effectively downsamples ultra-large scRNA-Seq datasets, without losing extremely rare or transcriptionally similar yet distinct cell types and while recovering novel transcriptionally distinct cell populations. We believe this new approach holds tremendous promise in reproducibly resolving hidden cell populations in complex datasets.\n",
      "ICGS2 is implemented in Python. The source code and documentation are available at http://altanalyze.org.\n",
      "Supplementary data are available at Bioinformatics online.\n",
      "\n",
      "0.854\n",
      "# Detection of differentially abundant cell subpopulations in scRNA-seq data.\n",
      "Comprehensive and accurate comparisons of transcriptomic distributions of cells from samples taken from two different biological states, such as healthy versus diseased individuals, are an emerging challenge in single-cell RNA sequencing (scRNA-seq) analysis. Current methods for detecting differentially abundant (DA) subpopulations between samples rely heavily on initial clustering of all cells in both samples. Often, this clustering step is inadequate since the DA subpopulations may not align with a clear cluster structure, and important differences between the two biological states can be missed. Here, we introduce DA-seq, a targeted approach for identifying DA subpopulations not restricted to clusters. DA-seq is a multiscale method that quantifies a local DA measure for each cell, which is computed from its \n",
      "\n",
      "0.853\n",
      "# Alignment of single-cell RNA-seq samples without overcorrection using kernel density matching.\n",
      "Single-cell RNA sequencing (scRNA-seq) technology is poised to replace bulk cell RNA sequencing for many biological and medical applications as it allows users to measure gene expression levels in a cell type-specific manner. However, data produced by scRNA-seq often exhibit batch effects that can be specific to a cell type, to a sample, or to an experiment, which prevent integration or comparisons across multiple experiments. Here, we present Dmatch, a method that leverages an external expression atlas of human primary cells and kernel density matching to align multiple scRNA-seq experiments for downstream biological analysis. Dmatch facilitates alignment of scRNA-seq data sets with cell types that may overlap only partially and thus allows integration of multiple distinct scRNA-seq experiments to extract biological insights. In simulation, Dmatch compares favorably to other alignment methods, both in terms of reducing sample-specific clustering and in terms of avoiding overcorrection. When applied to scRNA-seq data collected from clinical samples in a healthy individual and five autoimmune disease patients, Dmatch enabled cell type-specific differential gene expression comparisons across biopsy sites and disease conditions and uncovered a shared population of pro-inflammatory monocytes across biopsy sites in RA patients. We further show that Dmatch increases the number of eQTLs mapped from population scRNA-seq data. Dmatch is fast, scalable, and improves the utility of scRNA-seq for several important applications. Dmatch is freely available online.\n",
      "\n",
      "0.853\n",
      "# Analysis of aggregated cell-cell statistical distances within pathways unveils therapeutic-resistance mechanisms in circulating tumor cells.\n",
      "As 'omics' biotechnologies accelerate the capability to contrast a myriad of molecular measurements from a single cell, they also exacerbate current analytical limitations for detecting meaningful single-cell dysregulations. Moreover, mRNA expression alone lacks functional interpretation, limiting opportunities for translation of single-cell transcriptomic insights to precision medicine. Lastly, most single-cell RNA-sequencing analytic approaches are not designed to investigate small populations of cells such as circulating tumor cells shed from solid tumors and isolated from patient blood samples.\n",
      "In response to these characteristics and limitations in current single-cell RNA-sequencing methodology, we introduce an analytic framework that models transcriptome dynamics through the analysis of aggregated cell-cell statistical distances within biomolecular pathways. Cell-cell statistical distances are calculated from pathway mRNA fold changes between two cells. Within an elaborate case study of circulating tumor cells derived from prostate cancer patients, we develop analytic methods of aggregated distances to identify five differentially expressed pathways associated to therapeutic resistance. Our aggregation analyses perform comparably with Gene Set Enrichment Analysis and better than differentially expressed genes followed by gene set enrichment. However, these methods were not designed to inform on differential pathway expression for a single cell. As such, our framework culminates with the novel aggregation method, cell-centric statistics (CCS). CCS quantifies the effect size and significance of differentially expressed pathways for a single cell of interest. Improved rose plots of differentially expressed pathways in each cell highlight the utility of CCS for therapeutic decision-making.\n",
      "http://www.lussierlab.org/publications/CCS/ CONTACT: yves@email.arizona.edu or piegorsch@math.arizona.edu\n",
      "Supplementary data are available at Bioinformatics online.\n",
      "\n",
      "0.853\n",
      "# Characterizing the replicability of cell types defined by single cell RNA-sequencing data using MetaNeighbor.\n",
      "Single-cell RNA-sequencing (scRNA-seq) technology provides a new avenue to discover and characterize cell types; however, the experiment-specific technical biases and analytic variability inherent to current pipelines may undermine its replicability. Meta-analysis is further hampered by the use of ad hoc naming conventions. Here we demonstrate our replication framework, MetaNeighbor, that quantifies the degree to which cell types replicate across datasets, and enables rapid identification of clusters with high similarity. We first measure the replicability of neuronal identity, comparing results across eight technically and biologically diverse datasets to define best practices for more complex assessments. We then apply this to novel interneuron subtypes, finding that 24/45 subtypes have evidence of replication, which enables the identification of robust candidate marker genes. Across tasks we find that large sets of variably expressed genes can identify replicable cell types with high accuracy, suggesting a general route forward for large-scale evaluation of scRNA-seq data.\n",
      "\n",
      "0.853\n",
      "# Digitaldlsorter: Deep-Learning on scRNA-Seq to Deconvolute Gene Expression Data.\n",
      "The development of single cell transcriptome sequencing has allowed researchers the possibility to dig inside the role of the individual cell types in a plethora of disease scenarios. It also expands to the whole transcriptome what before was only possible for a few tenths of antibodies in cell population analysis. More importantly, it allows resolving the permanent question of whether the changes observed in a particular bulk experiment are a consequence of changes in cell type proportions or an aberrant behavior of a particular cell type. However, single cell experiments are still complex to perform and expensive to sequence making bulk RNA-Seq experiments yet more common. scRNA-Seq data is proving highly relevant information for the characterization of the immune cell repertoire in different diseases ranging from cancer to atherosclerosis. In particular, as scRNA-Seq becomes more widely used, new types of immune cell populations emerge and their role in the genesis and evolution of the disease opens new avenues for personalized immune therapies. Immunotherapy have already proven successful in a variety of tumors such as breast, colon and melanoma and its value in other types of disease is being currently explored. From a statistical perspective, single-cell data are particularly interesting due to its high dimensionality, overcoming the limitations of the \"skinny matrix\" that traditional bulk RNA-Seq experiments yield. With the technological advances that enable sequencing hundreds of thousands of cells, scRNA-Seq data have become especially suitable for the application of Machine Learning algorithms such as Deep Learning (DL). We present here a DL based method to enumerate and quantify the immune infiltration in colorectal and breast cancer bulk RNA-Seq samples starting from scRNA-Seq. Our method makes use of a Deep Neural Network (DNN) model that allows quantification not only of lymphocytes as a general population but also of specific CD8+, CD4Tmem, CD4Th and CD4Tregs subpopulations, as well as B-cells and Stromal content. Moreover, the signatures are built from scRNA-Seq data from the tumor, preserving the specific characteristics of the tumor microenvironment as opposite to other approaches in which cells were isolated from blood. Our method was applied to synthetic bulk RNA-Seq and to samples from the TCGA project yielding very accurate results in terms of quantification and survival prediction.\n",
      "\n",
      "0.852\n",
      "# Tracking intratumoral heterogeneity in glioblastoma via regularized classification of single-cell RNA-Seq data.\n",
      "Understanding cellular and molecular heterogeneity in glioblastoma (GBM), the most common and aggressive primary brain malignancy, is a crucial step towards the development of effective therapies. Besides the inter-patient variability, the presence of multiple cell populations within tumors calls for the need to develop modeling strategies able to extract the molecular signatures driving tumor evolution and treatment failure. With the advances in single-cell RNA Sequencing (scRNA-Seq), tumors can now be dissected at the cell level, unveiling information from their life history to their clinical implications.\n",
      "We propose a classification setting based on GBM scRNA-Seq data, through sparse logistic regression, where different cell populations (neoplastic and normal cells) are taken as classes. The goal is to identify gene features discriminating between the classes, but also those shared by different neoplastic clones. The latter will be approached via the network-based twiner regularizer to identify gene signatures shared by neoplastic cells from the tumor core and infiltrating neoplastic cells originated from the tumor periphery, as putative disease biomarkers to target multiple neoplastic clones. Our analysis is supported by the literature through the identification of several known molecular players in GBM. Moreover, the relevance of the selected genes was confirmed by their significance in the survival outcomes in bulk GBM RNA-Seq data, as well as their association with several Gene Ontology (GO) biological process terms.\n",
      "We presented a methodology intended to identify genes discriminating between GBM clones, but also those playing a similar role in different GBM neoplastic clones (including migrating cells), therefore potential targets for therapy research. Our results contribute to a deeper understanding on the genetic features behind GBM, by disclosing novel therapeutic directions accounting for GBM heterogeneity.\n",
      "\n",
      "0.852\n",
      "# Improvements Achieved by Multiple Imputation for Single-Cell RNA-Seq Data in Clustering Analysis and Differential Expression Analysis.\n",
      "In a single-cell RNA-seq (scRNA-seq) data set, a high proportion of missing values (or an excessive number of zeroes) are frequently observed. For the related follow-up tasks, such as clustering analysis and differential expression analysis, a data set without missing values is generally required. Many imputation approaches have been proposed for this purpose. Multiple imputation (MI) is a well-established approach to address possible biases in a follow-up analysis result based on one-time imputed data. There is a lack of investigation on this in the analysis of scRNA-seq data. In this study, we have investigated how to efficiently apply the MI approach to the clustering analysis and the differential expression analysis of scRNA-seq data. We proposed an MI procedure for clustering analysis and an MI procedure for differential expression analysis. To demonstrate the improvements achieved by MI in clustering analysis and differential expression analysis of scRNA-seq data, we analyzed three well-known scRNA-seq data sets. scIGANs, an scRNA-seq imputation method based on the generative adversarial networks (GANs), has been recently proposed for scRNA-seq data imputation. Multiple randomly imputed data sets can be conveniently generated by this method. We implemented our MI procedures based on scIGANs. We demonstrated that MI yielded improved performances on the clustering analysis and differential expression analysis results. Our applications to experimental scRNA-seq data illustrated the advantages of MI over one-time imputation of missing values in scRNA-seq data.\n",
      "\n",
      "0.852\n",
      "# Leveraging Single-Cell RNA Sequencing Experiments to Model Intratumor Heterogeneity.\n",
      "Many cancers can be treated with targeted therapy. Almost inevitably, tumors develop resistance to targeted therapy, either from pre-existence or by evolving new genotypes and traits. Intratumor heterogeneity serves as a reservoir for resistance, which often occurs as a result of the selection of minor cellular subclones. On the level of gene expression, clonal heterogeneity can only be revealed using high-dimensional single-cell methods. We propose using a general diversity index (GDI) to quantify heterogeneity on multiple scales and relate it to disease evolution.\n",
      "We focused on individual patient samples that were probed with single-cell RNA (scRNA) sequencing to describe heterogeneity. We developed a pipeline to analyze single-cell data via sample normalization, clustering, and mathematical interpretation using a generalized diversity measure, as well as to exemplify the utility of this platform using single-cell data.\n",
      "We focused on three sources of patient scRNA sequencing data: two healthy bone marrow (BM) donors, two patients with acute myeloid leukemia-each sampled before and after BM transplantation, four samples of presorted lineages-and six patients with lung carcinoma with multiregion sampling. While healthy/normal samples scored low in diversity overall, GDI further quantified the ways in which these samples differed. Whereas a widely used Shannon diversity index sometimes reveals fewer differences, GDI exhibits differences in the number of potential key drivers or clonal richness. Comparison of pre- and post-BM transplantation acute myeloid leukemia samples did not reveal differences in heterogeneity, although biological differences can exist.\n",
      "GDI can quantify cellular heterogeneity changes across a wide spectrum, even when standard measures, such as the Shannon index, do not. Our approach can be widely applied to quantify heterogeneity across samples and conditions.\n",
      "\n",
      "0.852\n",
      "# Detection of differentially expressed genes in discrete single-cell RNA sequencing data using a hurdle model with correlated random effects.\n",
      "Single-cell RNA sequencing (scRNA-seq) technologies are revolutionary tools allowing researchers to examine gene expression at the level of a single cell. Traditionally, transcriptomic data have been analyzed from bulk samples, masking the heterogeneity now seen across individual cells. Even within the same cellular population, genes can be highly expressed in some cells but not expressed (or lowly expressed) in others. Therefore, the computational approaches used to analyze bulk RNA sequencing data are not appropriate for the analysis of scRNA-seq data. Here, we present a novel statistical model for high dimensional and zero-inflated scRNA-seq count data to identify differentially expressed (DE) genes across cell types. Correlated random effects are employed based on an initial clustering of cells to capture the cell-to-cell variability within treatment groups. Moreover, this model is flexible and can be easily adapted to an independent random effect structure if needed. We apply our proposed methodology to both simulated and real data and compare results to other popular methods designed for detecting DE genes. Due to the hurdle model's ability to detect differences in the proportion of cells expressed and the average expression level (among the expressed cells), our methods naturally identify some genes as DE that other methods do not, and we demonstrate with real data that these uniquely detected genes are associated with similar biological processes and functions.\n",
      "\n",
      "0.852\n",
      "# Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors.\n",
      "Large-scale single-cell RNA sequencing (scRNA-seq) data sets that are produced in different laboratories and at different times contain batch effects that may compromise the integration and interpretation of the data. Existing scRNA-seq analysis methods incorrectly assume that the composition of cell populations is either known or identical across batches. We present a strategy for batch correction based on the detection of mutual nearest neighbors (MNNs) in the high-dimensional expression space. Our approach does not rely on predefined or equal population compositions across batches; instead, it requires only that a subset of the population be shared between batches. We demonstrate the superiority of our approach compared with existing methods by using both simulated and real scRNA-seq data sets. Using multiple droplet-based scRNA-seq data sets, we demonstrate that our MNN batch-effect-correction method can be scaled to large numbers of cells.\n",
      "\n",
      "0.852\n",
      "# A multitask clustering approach for single-cell RNA-seq analysis in Recessive Dystrophic Epidermolysis Bullosa.\n",
      "Single-cell RNA sequencing (scRNA-seq) has been widely applied to discover new cell types by detecting sub-populations in a heterogeneous group of cells. Since scRNA-seq experiments have lower read coverage/tag counts and introduce more technical biases compared to bulk RNA-seq experiments, the limited number of sampled cells combined with the experimental biases and other dataset specific variations presents a challenge to cross-dataset analysis and discovery of relevant biological variations across multiple cell populations. In this paper, we introduce a method of variance-driven multitask clustering of single-cell RNA-seq data (scVDMC) that utilizes multiple single-cell populations from biological replicates or different samples. scVDMC clusters single cells in multiple scRNA-seq experiments of similar cell types and markers but varying expression patterns such that the scRNA-seq data are better integrated than typical pooled analyses which only increase the sample size. By controlling the variance among the cell clusters within each dataset and across all the datasets, scVDMC detects cell sub-populations in each individual experiment with shared cell-type markers but varying cluster centers among all the experiments. Applied to two real scRNA-seq datasets with several replicates and one large-scale droplet-based dataset on three patient samples, scVDMC more accurately detected cell populations and known cell markers than pooled clustering and other recently proposed scRNA-seq clustering methods. In the case study applied to in-house Recessive Dystrophic Epidermolysis Bullosa (RDEB) scRNA-seq data, scVDMC revealed several new cell types and unknown markers validated by flow cytometry. MATLAB/Octave code available at https://github.com/kuanglab/scVDMC.\n",
      "\n",
      "0.852\n",
      "# Souporcell: robust clustering of single-cell RNA-seq data by genotype without reference genotypes.\n",
      "Methods to deconvolve single-cell RNA-sequencing (scRNA-seq) data are necessary for samples containing a mixture of genotypes, whether they are natural or experimentally combined. Multiplexing across donors is a popular experimental design that can avoid batch effects, reduce costs and improve doublet detection. By using variants detected in scRNA-seq reads, it is possible to assign cells to their donor of origin and identify cross-genotype doublets that may have highly similar transcriptional profiles, precluding detection by transcriptional profile. More subtle cross-genotype variant contamination can be used to estimate the amount of ambient RNA. Ambient RNA is caused by cell lysis before droplet partitioning and is an important confounder of scRNA-seq analysis. Here we develop souporcell, a method to cluster cells using the genetic variants detected within the scRNA-seq reads. We show that it achieves high accuracy on genotype clustering, doublet detection and ambient RNA estimation, as demonstrated across a range of challenging scenarios.\n",
      "\n",
      "0.851\n",
      "# Choice of pre-processing pipeline influences clustering quality of scRNA-seq datasets.\n",
      "Single-cell RNA sequencing (scRNA-seq) has quickly become one of the most dominant techniques in modern transcriptome assessment. In particular, 10X Genomics' Chromium system, with its high throughput approach, turn key and thorough user guide made this cutting-edge technique accessible to many laboratories using diverse animal models. However, standard pre-processing, including the alignment and cell filtering pipelines might not be ideal for every organism or tissue. Here we applied an alternative strategy, based on the pseudoaligner kallisto, on twenty-two publicly available single cell sequencing datasets from a wide range of tissues of eight organisms and compared the results with the standard 10X Genomics' Cell Ranger pipeline.\n",
      "In most of the tested samples, kallisto produced higher sequencing read alignment rates and total gene detection rates in comparison to Cell Ranger. Although datasets processed with Cell Ranger had higher cell counts, outside of human and mouse datasets, these additional cells were routinely of low quality, containing low gene detection rates. Thorough downstream analysis of one kallisto processed dataset, obtained from the zebrafish pineal gland, revealed clearer clustering, allowing the identification of an additional photoreceptor cell type that previously went undetected. The finding of the new cluster suggests that the photoreceptive pineal gland is essentially a bi-chromatic tissue containing both green and red cone-like photoreceptors and implies that the alignment and pre-processing pipeline can affect the discovery of biologically-relevant cell types.\n",
      "While Cell Ranger favors higher cell numbers, using kallisto results in datasets with higher median gene detection per cell. We could demonstrate that cell type identification was not hampered by the lower cell count, but in fact improved as a result of the high gene detection rate and the more stringent filtering. Depending on the acquired dataset, it can be beneficial to favor high quality cells and accept a lower cell count, leading to an improved classification of cell types.\n",
      "\n",
      "0.851\n",
      "# R code and downstream analysis objects for the scRNA-seq atlas of normal and tumorigenic human breast tissue.\n",
      "Breast cancer is a common and highly heterogeneous disease. Understanding cellular diversity in the mammary gland and its surrounding micro-environment across different states can provide insight into cancer development in the human breast. Recently, we published a large-scale single-cell RNA expression atlas of the human breast spanning normal, preneoplastic and tumorigenic states. Single-cell expression profiles of nearly 430,000 cells were obtained from 69 distinct surgical tissue specimens from 55 patients. This article extends the study by providing quality filtering thresholds, downstream processed R data objects, complete cell annotation and R code to reproduce all the analyses. Data quality assessment measures are presented and details are provided for all the bioinformatic analyses that produced results described in the study.\n",
      "\n",
      "0.851\n",
      "# Sierra: discovery of differential transcript usage from polyA-captured single-cell RNA-seq data.\n",
      "High-throughput single-cell RNA-seq (scRNA-seq) is a powerful tool for studying gene expression in single cells. Most current scRNA-seq bioinformatics tools focus on analysing overall expression levels, largely ignoring alternative mRNA isoform expression. We present a computational pipeline, Sierra, that readily detects differential transcript usage from data generated by commonly used polyA-captured scRNA-seq technology. We validate Sierra by comparing cardiac scRNA-seq cell types to bulk RNA-seq of matched populations, finding significant overlap in differential transcripts. Sierra detects differential transcript usage across human peripheral blood mononuclear cells and the Tabula Muris, and 3 \n",
      "\n",
      "0.851\n",
      "# An entropy-based metric for assessing the purity of single cell populations.\n",
      "Single-cell RNA sequencing (scRNA-seq) is a versatile tool for discovering and annotating cell types and states, but the determination and annotation of cell subtypes is often subjective and arbitrary. Often, it is not even clear whether a given cluster is uniform. Here we present an entropy-based statistic, ROGUE, to accurately quantify the purity of identified cell clusters. We demonstrate that our ROGUE metric is broadly applicable, and enables accurate, sensitive and robust assessment of cluster purity on a wide range of simulated and real datasets. Applying this metric to fibroblast, B cell and brain data, we identify additional subtypes and demonstrate the application of ROGUE-guided analyses to detect precise signals in specific subpopulations. ROGUE can be applied to all tested scRNA-seq datasets, and has important implications for evaluating the quality of putative clusters, discovering pure cell subtypes and constructing comprehensive, detailed and standardized single cell atlas.\n",
      "\n",
      "0.851\n",
      "# Cardelino: computational integration of somatic clonal substructure and single-cell transcriptomes.\n",
      "Bulk and single-cell DNA sequencing has enabled reconstructing clonal substructures of somatic tissues from frequency and cooccurrence patterns of somatic variants. However, approaches to characterize phenotypic variations between clones are not established. Here we present cardelino (https://github.com/single-cell-genetics/cardelino), a computational method for inferring the clonal tree configuration and the clone of origin of individual cells assayed using single-cell RNA-seq (scRNA-seq). Cardelino flexibly integrates information from imperfect clonal trees inferred based on bulk exome-seq data, and sparse variant alleles expressed in scRNA-seq data. We apply cardelino to a published cancer dataset and to newly generated matched scRNA-seq and exome-seq data from 32 human dermal fibroblast lines, identifying hundreds of differentially expressed genes between cells from different somatic clones. These genes are frequently enriched for cell cycle and proliferation pathways, indicating a role for cell division genes in somatic evolution in healthy skin.\n",
      "\n",
      "0.851\n",
      "# Identification of Differential Gene Groups From Single-Cell Transcriptomes Using Network Entropy.\n",
      "A complex tissue contains a variety of cells with distinct molecular signatures. Single-cell RNA sequencing has characterized the transcriptomes of different cell types and enables researchers to discover the underlying mechanisms of cellular heterogeneity. A critical task in single-cell transcriptome studies is to uncover transcriptional differences among specific cell types. However, the intercellular transcriptional variation is usually confounded with high level of technical noise, which masks the important biological signals. Here, we propose a new computational method DiffGE for differential analysis, adopting network entropy to measure the expression dynamics of gene groups among different cell types and to identify the highly differential gene groups. To evaluate the effectiveness of our proposed method, DiffGE is applied to three independent single-cell RNA-seq datasets and to identify the highly dynamic gene groups that exhibit distinctive expression patterns in different cell types. We compare the results of our method with those of three widely applied algorithms. Further, the gene function analysis indicates that these detected differential gene groups are significantly related to cellular regulation processes. The results demonstrate the power of our method in evaluating the transcriptional dynamics and identifying highly differential gene groups among different cell types.\n",
      "\n",
      "0.851\n",
      "# Single-cell RNA-seq data clustering: A survey with performance comparison study.\n",
      "Clustering analysis has been widely applied to single-cell RNA-sequencing (scRNA-seq) data to discover cell types and cell states. Algorithms developed in recent years have greatly helped the understanding of cellular heterogeneity and the underlying mechanisms of biological processes. However, these algorithms often use different techniques, were evaluated on different datasets and compared with some of their counterparts usually using different performance metrics. Consequently, there lacks an accurate and complete picture of their merits and demerits, which makes it difficult for users to select proper algorithms for analyzing their data. To fill this gap, we first do a review on the major existing scRNA-seq data clustering methods, and then conduct a comprehensive performance comparison among them from multiple perspectives. We consider 13 state of the art scRNA-seq data clustering algorithms, and collect 12 publicly available real scRNA-seq datasets from the existing works to evaluate and compare these algorithms. Our comparative study shows that the existing methods are very diverse in performance. Even the top-performance algorithms do not perform well on all datasets, especially those with complex structures. This suggests that further research is required to explore more stable, accurate, and efficient clustering algorithms for scRNA-seq data.\n",
      "\n",
      "0.851\n",
      "# MultiK: an automated tool to determine optimal cluster numbers in single-cell RNA sequencing data.\n",
      "Single-cell RNA sequencing (scRNA-seq) provides new opportunities to characterize cell populations, typically accomplished through some type of clustering analysis. Estimation of the optimal cluster number (K) is a crucial step but often ignored. Our approach improves most current scRNA-seq cluster methods by providing an objective estimation of the number of groups using a multi-resolution perspective. MultiK is a tool for objective selection of insightful Ks and achieves high robustness through a consensus clustering approach. We demonstrate that MultiK identifies reproducible groups in scRNA-seq data, thus providing an objective means to estimating the number of possible groups or cell-type populations present.\n",
      "\n",
      "0.851\n",
      "# Benchmarking single cell RNA-sequencing analysis pipelines using mixture control experiments.\n",
      "Single cell RNA-sequencing (scRNA-seq) technology has undergone rapid development in recent years, leading to an explosion in the number of tailored data analysis methods. However, the current lack of gold-standard benchmark datasets makes it difficult for researchers to systematically compare the performance of the many methods available. Here, we generated a realistic benchmark experiment that included single cells and admixtures of cells or RNA to create 'pseudo cells' from up to five distinct cancer cell lines. In total, 14 datasets were generated using both droplet and plate-based scRNA-seq protocols. We compared 3,913 combinations of data analysis methods for tasks ranging from normalization and imputation to clustering, trajectory analysis and data integration. Evaluation revealed pipelines suited to different types of data for different tasks. Our data and analysis provide a comprehensive framework for benchmarking most common scRNA-seq analysis steps.\n",
      "\n",
      "0.85\n",
      "# Linear-time cluster ensembles of large-scale single-cell RNA-seq and multimodal data.\n",
      "A fundamental task in single-cell RNA-seq (scRNA-seq) analysis is the identification of transcriptionally distinct groups of cells. Numerous methods have been proposed for this problem, with a recent focus on methods for the cluster analysis of ultralarge scRNA-seq data sets produced by droplet-based sequencing technologies. Most existing methods rely on a sampling step to bridge the gap between algorithm scalability and volume of the data. Ignoring large parts of the data, however, often yields inaccurate groupings of cells and risks overlooking rare cell types. We propose method Specter that adopts and extends recent algorithmic advances in (fast) spectral clustering. In contrast to methods that cluster a (random) subsample of the data, we adopt the idea of landmarks that are used to create a sparse representation of the full data from which a spectral embedding can then be computed in linear time. We exploit Specter's speed in a cluster ensemble scheme that achieves a substantial improvement in accuracy over existing methods and identifies rare cell types with high sensitivity. Its linear-time complexity allows Specter to scale to millions of cells and leads to fast computation times in practice. Furthermore, on CITE-seq data that simultaneously measures gene and protein marker expression, we show that Specter is able to use multimodal omics measurements to resolve subtle transcriptomic differences between subpopulations of cells.\n",
      "\n",
      "0.85\n",
      "# ClusterMap: compare multiple single cell RNA-Seq datasets across different experimental conditions.\n",
      "Single cell RNA-Seq (scRNA-Seq) facilitates the characterization of cell type heterogeneity and developmental processes. Further study of single cell profiles across different conditions enables the understanding of biological processes and underlying mechanisms at the sub-population level. However, developing proper methodology to compare multiple scRNA-Seq datasets remains challenging.\n",
      "We have developed ClusterMap, a systematic method and workflow to facilitate the comparison of scRNA-seq profiles across distinct biological contexts. Using hierarchical clustering of the marker genes of each sub-group, ClusterMap matches the sub-types of cells across different samples and provides 'similarity' as a metric to quantify the quality of the match. We introduce a purity tree cut method designed specifically for this matching problem. We use Circos plot and regrouping method to visualize the results concisely. Furthermore, we propose a new metric 'separability' to summarize sub-population changes among all sample pairs. In the case studies, we demonstrate that ClusterMap has the ability to provide us further insight into the different molecular mechanisms of cellular sub-populations across different conditions.\n",
      "ClusterMap is implemented in R and available at https://github.com/xgaoo/ClusterMap.\n",
      "Supplementary data are available at Bioinformatics online.\n",
      "\n",
      "0.85\n",
      "# Comparative analysis of differential gene expression analysis tools for single-cell RNA sequencing data.\n",
      "The analysis of single-cell RNA sequencing (scRNAseq) data plays an important role in understanding the intrinsic and extrinsic cellular processes in biological and biomedical research. One significant effort in this area is the detection of differentially expressed (DE) genes. scRNAseq data, however, are highly heterogeneous and have a large number of zero counts, which introduces challenges in detecting DE genes. Addressing these challenges requires employing new approaches beyond the conventional ones, which are based on a nonzero difference in average expression. Several methods have been developed for differential gene expression analysis of scRNAseq data. To provide guidance on choosing an appropriate tool or developing a new one, it is necessary to evaluate and compare the performance of differential gene expression analysis methods for scRNAseq data.\n",
      "In this study, we conducted a comprehensive evaluation of the performance of eleven differential gene expression analysis software tools, which are designed for scRNAseq data or can be applied to them. We used simulated and real data to evaluate the accuracy and precision of detection. Using simulated data, we investigated the effect of sample size on the detection accuracy of the tools. Using real data, we examined the agreement among the tools in identifying DE genes, the run time of the tools, and the biological relevance of the detected DE genes.\n",
      "In general, agreement among the tools in calling DE genes is not high. There is a trade-off between true-positive rates and the precision of calling DE genes. Methods with higher true positive rates tend to show low precision due to their introducing false positives, whereas methods with high precision show low true positive rates due to identifying few DE genes. We observed that current methods designed for scRNAseq data do not tend to show better performance compared to methods designed for bulk RNAseq data. Data multimodality and abundance of zero read counts are the main characteristics of scRNAseq data, which play important roles in the performance of differential gene expression analysis methods and need to be considered in terms of the development of new methods.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sdf = df[[\"title\", \"abstract\"]].reset_index().join(DataFrame(dict(sim=query_sims[:,0]))).sort_values('sim', ascending=False)\n",
    "for i in range(50):\n",
    "    print(round(sdf.sim.values[i], 3))\n",
    "    print(\"# \" + sdf.title.values[i])\n",
    "    print(sdf.abstract.values[i])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## Embed UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from umap import UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 13s, sys: 2.46 s, total: 3min 16s\n",
      "Wall time: 22.4 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "ump = UMAP(n_components=20, metric=\"cosine\")\n",
    "emb_umaps = ump.fit_transform(embs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAi4AAAGhCAYAAABGRD9PAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAAApsElEQVR4nO3df3RU9Z3/8dckzCQTksAGgSy/JAlCGUz5tYXQcohLxXYhUWxktQcWEYiFUKW4uFJleyyNCAdMqvKrWHTLD3c58VC2ssSzdnXTtRVOj6GtFXBlIAt4NAIWJpKQCcn9/uF3RsaEMJPMr8/k+TjHE3Pv5w6fy5vJvPK5n/u5NsuyLAEAABggKdYdAAAACBbBBQAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguAADAGL1i3YFwOnLkiCzLkt1uj3VXAABAkFpaWmSz2TR+/Pgbtk2oERfLstST19OzLEter7dH/x3EK2oTn6hL/KI28SsStQnl8zuhRlx8Iy35+fkx7klsNDY26tixYxoxYoTS0tJi3R1cg9rEJ+oSv6hN/IpEbd59992g2ybUiAsAAEhsBBcAAGAMggsAADAGwQUAABiD4AIAAIxBcAEAAMYguAAAAGMQXAAAgDEILgAAwBgEFwAAYAyCCwAAMAbBBQAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguAAD0AG1tVljaxFqvWHcAAABEXlKSTRv3vKOz9Q0d7h8yMEMr506Mcq9CR3ABAKCHOFvfIPeHl2LdjW7hUhEAADAGwQUAABiD4AIAAIxBcAEAAMYguAAAAGMQXAAAgDEILgAAwBgEFwAAYAyCCwAAMAbBBQAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguAADAGAQXAABgDIILAAAwBsEFAAAYg+ACAACMQXABAADGILgAAABjhBxcfv3rX2vOnDmaMGGCvv71r+v73/++Tp482a5dTU2NZs+erfz8fM2YMUN79uzp8PV27Nih6dOnKz8/XyUlJTp8+HDoZwEAAHqEkILL7373O33/+99XTk6Onn/+ef3zP/+zTp06pQceeECfffaZv92RI0dUVlYml8ulF154QXfffbfKy8tVVVUV8Ho7duxQZWWl5s6dq+3bt+vmm29WaWmp3n///fCcHQAASCi9Qmn8H//xHxo0aJDWr18vm80mSRo8eLDmzJmjd955R4WFhZKkzZs3y+Vyae3atZKkgoICffTRR3r22WdVUlKipKQkeb1ebd26VfPnz9eiRYskSZMmTVJxcbG2bdumysrKcJ4nAABIACGNuFy9elW9e/f2hxZJysjICGjj9Xp16NAhzZo1K2B7cXGxzp07p6NHj0qSamtr1dDQoKKiIn+b5ORkzZw5UzU1NbIsK+STAQAAiS2kEZd77rlHCxYs0K5du3TXXXfJ4/Fo/fr1ysvL05QpUyRJp0+fVktLi3JzcwOOHTFihCTJ7Xbr1ltvldvtlqR27fLy8nT58mXV19crOzs75BOyLEuNjY0hH5cImpqaAr4iflCb+ERd4he1CS+bzSan0xlU26ampk4HDyJRG8uyAgZFOhNScPna176mTZs26R//8R9VXl4u6fNA8uKLL8rhcEiSLl26JEnKzMwMONb3vW+/x+ORw+FQampqQLs+ffpIki5evNil4NLS0qJjx46FfFwiqauri3UXcB3UJj5Rl/hFbcLD6XTK5XIF1fbUqVNBhZJw18aXI24kpOBSW1urRx99VCUlJZo+fbo+++wzbdu2TaWlpfrXf/1Xpaen+9teLzldu72jNr6UF2zy+jK73e4f3elpmpqaVFdXp+HDhwedrBEd1CY+UZf4RW3CK5TP1JycnBuOuIS7NidOnAi6bUjBpby8XAUFBXriiSf82yZOnKhp06apqqpKDzzwgH/ExDey4uPxeCR9MfKSmZmp5uZmNTc3KyUlpV073+uEymazKS0trUvHJgqn09nj/w7iFbWJT9QlflGb6As2jISzNqEEq5Am57rdbn3lK18J2JaVlaUBAwbo9OnTkqRhw4bJbre3W9vFl6by8vICvvrmulz7Z/Tu3VsDBw4MpWsAAKAHCCm4DBo0SO+9917AtnPnzumTTz7R4MGDJX1+jaqgoEDV1dUB7Q4cOKD+/fv7r7FNmDBBGRkZOnjwoL9Na2urqqurVVhY2OVLRQAAIHGFdKlo7ty5+slPfqI1a9bom9/8pjwej372s58pLS1Nd955p7/dsmXLNG/ePK1evVrFxcWqra1VVVWV1qxZo6Skz7OSw+HQ0qVLVVlZqaysLLlcLlVVVenMmTOqqKgI71kCAICEEHJwsdvtevnll/XLX/5SaWlpys/P1/r16zVgwAB/u/Hjx2vLli2qqKjQ/v37lZ2drdWrV2vOnDkBr7dw4UJZlqVdu3bp/PnzGjlypLZv365Ro0aF5+wAAEBCCSm42Gw23Xvvvbr33ntv2LawsNC/km5nr7d48WItXrw4lG4AAIAeiqdDAwAAYxBcAACAMQguAADAGAQXAABgDIILAAAwBsEFAAAYg+ACAACMQXABAADGILgAAABjEFwAAIAxCC4AAMAYBBcAAGAMggsAADAGwQUAABiD4AIAAIxBcAEAAMYguAAAAGMQXAAAgDEILgAAwBgEFwAAYAyCCwAAMAbBBQAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguAADAGAQXAABgDIILAAAwBsEFAAAYg+ACAACMQXABAADGILgAAABjEFwAAIAxCC4AAMAYBBcAAGAMggsAADAGwQUAABiD4AIAAIxBcAEAAMYguAAAAGMQXAAAgDEILgAAwBgEFwAAYAyCCwAAMAbBBQAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguAADAGAQXAABgDIILAAAwBsEFAAAYg+ACAACMQXABAADGILgAAABjEFwAAIAxCC4AAMAYBBcAAGAMggsAADAGwQUAABiD4AIAAIxBcAEAAMYguAAAAGMQXAAAgDEILgAAwBgEFwAAYIwuBZeqqirdeeedys/P15QpU7RkyZKA/TU1NZo9e7by8/M1Y8YM7dmzp8PX2bFjh6ZPn678/HyVlJTo8OHDXekOAAA9WlubFesuRE2vUA94/vnn9S//8i9asmSJxo4dq0uXLul//ud//PuPHDmisrIy3XXXXVq1apVqa2tVXl4uh8OhOXPm+Nvt2LFDlZWVWrFihVwul6qqqlRaWqqqqiqNGjUqPGcHAEAPkJRk08Y97+hsfUOH+yd8ZYDmz3RFuVeREVJwcbvd2rp1q7Zv366pU6f6t8+YMcP//5s3b5bL5dLatWslSQUFBfroo4/07LPPqqSkRElJSfJ6vdq6davmz5+vRYsWSZImTZqk4uJibdu2TZWVleE4NwAAeoyz9Q1yf3ipw31DBqRHuTeRE9Klon379mno0KEBoeVaXq9Xhw4d0qxZswK2FxcX69y5czp69Kgkqba2Vg0NDSoqKvK3SU5O1syZM1VTUyPL6jlDXgAAIHghjbj88Y9/1MiRI7V582bt3r1bDQ0NGjdunJ544gmNHj1ap0+fVktLi3JzcwOOGzFihKTPR2xuvfVWud1uSWrXLi8vT5cvX1Z9fb2ys7O7dEKWZamxsbFLx5quqakp4CviB7WJT9QlflGb4NlsNjmdzrC9XlNTU6cDCJGojWVZstlsQbUNKbicO3dO7733nj744AP9+Mc/lt1u16ZNm/TAAw/oP//zP3Xp0udDVJmZmQHH+b737fd4PHI4HEpNTQ1o16dPH0nSxYsXuxxcWlpadOzYsS4dmyjq6upi3QVcB7WJT9QlflGbG3M6nXK5wjd/5dSpU0GFknDXxuFwBNUupODiG814/vnndcstt0iSxowZo29+85vau3evJkyYIEnXTU3Xbu+ojS/hBZu6OmK32/0jPD1NU1OT6urqNHz48LCmb3QftYlP1CV+UZvgdeczsyM5OTk3HHEJd21OnDgRdNuQgkufPn100003+UOLJA0YMEC5ubk6ceKE/vZv/1bSFyMrPh6PR9IXIy+ZmZlqbm5Wc3OzUlJS2rXzjbx0hc1mU1paWpePTwROp7PH/x3EK2oTn6hL/KI20RdsGAlnbUIJXyFNzs3Ly+twu2VZSkpK0rBhw2S323Xy5MmA/b4k5Tve99U318XH7Xard+/eGjhwYCjdAgAAPURIweW2227T+fPn9b//+7/+bfX19Tp58qRGjRolh8OhgoICVVdXBxx34MAB9e/f338NbsKECcrIyNDBgwf9bVpbW1VdXa3CwsKwD3sBAIDEENKlohkzZmjMmDF66KGHtHz5cjkcDm3evFlZWVn6+7//e0nSsmXLNG/ePK1evVrFxcWqra1VVVWV1qxZo6Skz3OSw+HQ0qVLVVlZqaysLP8CdGfOnFFFRUX4zxIAAIRFrAcXQgouycnJeuGFF7R27Vr96Ec/0tWrV/W1r31NzzzzjP861/jx47VlyxZVVFRo//79ys7O1urVqwNWzZWkhQsXyrIs7dq1S+fPn9fIkSO1fft2Vs0FACAG+makqK3NUlJS58HEkZIiu90epV61F/KS//369dMzzzzTaZvCwkIVFhZ22sZms2nx4sVavHhxqF0AAABhlu603/DRAUMGZmjl3Inq1Svk+BA2sfuTAQBA3Ons0QHxoEtPhwYAAIgFggsAADAGwQUAABiD4AIAAIxBcAEAAMYguAAAAGMQXAAAgDEILgAAwBgEFwAAYAyCCwAAMAbBBQAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguAADAGAQXAABgDIILAAAwBsEFAAAYg+ACAACMQXABAADGILgAAABjEFwAAIAxCC4AAMAYBBcAAGAMggsAADAGwQUAABiD4AIAAIxBcAEAAMYguAAAAGMQXAAAgDEILgAAwBgEFwAAYAyCCwAAMAbBBQAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguAADAGAQXAABgDIILAAAwBsEFAAAYg+ACAACMQXABAADGILgAAABjEFwAAIAxCC4AAMAYBBcAAGAMggsAADAGwQUAABiD4AIAAIxBcAEAAMYguAAAAGMQXAAAgDEILgAAwBgEFwAAYAyCCwAAMAbBBQAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguAADAGAQXAABgDIILAAAwRreCy+XLlzVt2jSNGjVK7777bsC+mpoazZ49W/n5+ZoxY4b27NnT4Wvs2LFD06dPV35+vkpKSnT48OHudAkAACSwbgWXLVu2qLW1td32I0eOqKysTC6XSy+88ILuvvtulZeXq6qqKqDdjh07VFlZqblz52r79u26+eabVVpaqvfff7873QIAAAmqy8HF7Xbr5Zdf1kMPPdRu3+bNm+VyubR27VoVFBSorKxM99xzj5599lm1tbVJkrxer7Zu3ar58+dr0aJFmjJlijZs2KAhQ4Zo27ZtXT8jAACQsLocXJ566indd999ysnJCdju9Xp16NAhzZo1K2B7cXGxzp07p6NHj0qSamtr1dDQoKKiIn+b5ORkzZw5UzU1NbIsq6tdAwAACapLweW1117T8ePHtWzZsnb7Tp8+rZaWFuXm5gZsHzFihKTPR2qu/frldnl5ebp8+bLq6+u70jUAAJDAeoV6QFNTk9atW6dHHnlE6enp7fZfunRJkpSZmRmw3fe9b7/H45HD4VBqampAuz59+kiSLl68qOzs7FC7J8uy1NjYGPJxiaCpqSngK+IHtYlP0aqLzWYLqh0jzV/gPRM8m80mp9MZ9T83nLWxLCvo90nIwWXr1q3q16+fvvOd73Ta7noduHZ7R218b9xgT+DLWlpadOzYsS4dmyjq6upi3QVcB7WJT5Gsi91ul8s1Rr16JXfa7urVVh09+p5aWloi1hcT8Z65MafTKZfLFfU/N9y1cTgcQbULKbh8+OGHevHFF7V582Z99tlnkuQf3WhsbNTly5f9Iya+kRUfj8cj6YuRl8zMTDU3N6u5uVkpKSnt2vleJ1R2u91/WaqnaWpqUl1dnYYPHx6T9I3rozbxKRp1sdls6tUrWRv3vKOz9Q0dthkyMEMr507ULbfcwqjL/8d7Jnhd/UW/u8JZmxMnTgTdNqTgcvbsWbW0tOjBBx9st2/+/PkaO3asdu/eLbvdrpMnT2ratGntOpWXlxfw1e12ByRFt9ut3r17a+DAgaF0zc9msyktLa1LxyYKp9PZ4/8O4hW1iU/RqMvZ+ga5P7zUaRs+oNvjPRO/wlmbUMJXSMFl9OjR2rlzZ8C2Y8eO6emnn9aPf/xj5efny+FwqKCgQNXV1VqwYIG/3YEDB9S/f39/SJkwYYIyMjJ08OBB/7bW1lZVV1ersLAwZgkSAADEr5CCS2ZmpiZPntzhvjFjxmjMmDGSpGXLlmnevHlavXq1iouLVVtbq6qqKq1Zs0ZJSZ/fyORwOLR06VJVVlYqKytLLpdLVVVVOnPmjCoqKrp5WgAAIBGFPDk3GOPHj9eWLVtUUVGh/fv3Kzs7W6tXr9acOXMC2i1cuFCWZWnXrl06f/68Ro4cqe3bt2vUqFGR6BYAADBct4PL5MmTO1yiv7CwUIWFhZ0ea7PZtHjxYi1evLi73QAAAD0AT4cGAADGILgAAABjEFwAAIAxCC4AAMAYBBcAAGAMggsAdKCtLbil94NtByA8IrKOCwCYLinJ1unzhaQvnjEEIHoILgBwHcE8XwhAdHGpCAAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguAADEMdYKCsTt0AB6LJvNJqfTKZvNFuuuANd1ozWFJnxlgObPdEW5V7FDcAHQ47S1WUpK+jy0uFw95wc+zNXZmkJDBqRHuTexRXAB0OPwGyxgLoILgB6J32ABMzE5FwAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAKCL+maksDgYEGXcVQQAXZTutHNrNRBlBBcACcW3uFw0cWs1ED0EFwAJhREQILERXAAkHEZAgMTF5FwAiLFgJ/kyERhgxAUAYi6YSb5DBmZo5dyJN3ytYOb4xGIeEBAuBBcAiBOdXeIKVrgCEBCvCC4AkGDCEYCAeMUcFwDGYI4HAEZcABijJ9/q7JvAy9wU9HQEFwBG6am3OrNKL/A5ggsAGKSnBrdExSha6AguAADECKNooSO4AAAQQ4yihYa7igAAgDEILgAAwBgEFwAAYAyCCwAAMAbBBQDQJTzRGrHAXUUAgC7hgY6IBYILAKDLeKAjoo1LRQAAwBgEFwAAYAyCCwAAMAbBBQAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguAADAGAQXAABgDIILAAAwBsEFMAxP5EV39M1I4d8QjMZDFgHD8ERedEe6086/IRiN4AIYiCfyorv4NwRTcakIAAAYg+ACAAgQ7DwYIBa4VAQACBDMPJgJXxmg+TNdUe4ZQHABAFxHZ/NghgxIj3JvgM9xqQgAABiD4AIAAIxBcAEAAMYguAAAAGMQXAAAgDEILgAAwBghBZfq6mqVlZWpsLBQ48aNU3FxsV5++WW1tbUFtKupqdHs2bOVn5+vGTNmaM+ePR2+3o4dOzR9+nTl5+erpKREhw8f7vqZAIhbPNQPQLiEtI7LSy+9pEGDBumf/umf1K9fPx0+fFhPPfWUzpw5o8cee0ySdOTIEZWVlemuu+7SqlWrVFtbq/LycjkcDs2ZM8f/Wjt27FBlZaVWrFghl8ulqqoqlZaWqqqqSqNGjQrvWQKImLY2S0lJtk7b8FA/AOESUnDZtm2bsrKy/N8XFBSosbFRe/bs0YoVK+RwOLR582a5XC6tXbvW3+ajjz7Ss88+q5KSEiUlJcnr9Wrr1q2aP3++Fi1aJEmaNGmSiouLtW3bNlVWVobxFAFEUrArrPJQPwDhENKlomtDi8/o0aPV3Nysixcvyuv16tChQ5o1a1ZAm+LiYp07d05Hjx6VJNXW1qqhoUFFRUX+NsnJyZo5c6ZqampkWQwZAybxhZKO/vvk08ZYdw9AAun2kv/vvPOO+vbtq379+unUqVNqaWlRbm5uQJsRI0ZIktxut2699Va53W5JatcuLy9Ply9fVn19vbKzs7vUH8uy1NjYM39QNjU1BXxF/AhXbWw2m5xOZ9B/ZqR/CQilP8G4cuXKdftss9mUmpoatj8L0dOVf4s94edZuN8/0RbO2liWJZut80vOPt0KLu+++6727dunZcuWKTk5WZcufT4MnJmZGdDO971vv8fjkcPhaPdDqE+fPpKkixcvdjm4tLS06NixY106NlHU1dXFugu4ju7Wxul0yuUK7sF2p06divgP/VD60xnf04gJJompO/8WE/nnWbjeP7ES7to4HI6g2nU5uJw7d04PP/yw8vPzVVpaGrDveqnp2u0dtfEl8mBTV0fsdrt/hKenaWpqUl1dnYYPH250ik9E4apNKO+N3NzcG/6W290Rme68V6/F04gTW05OTpdGXBL951m43j+xEs7anDhxIui2XQouDQ0NKi0tVWpqqrZu3Sq73S7pixET38iKj8fjkfTFyEtmZqaam5vV3NyslJSUdu18r9MVNptNaWlpXT4+ETidzh7/dxCvolGbYEcvgrkbKNp4GnFi6s6HGz/P4lc4axNKiAs5uDQ3N2vp0qU6f/689u7dq7/6q7/y7xs2bJjsdrtOnjypadOm+bf7klReXl7AV7fbHTBM5na71bt3bw0cODDUbgH4/4IZveD2YwCmCim4XL16VcuXL9fx48e1e/duDR48OGC/w+FQQUGBqqurtWDBAv/2AwcOqH///v6QMmHCBGVkZOjgwYP+ba2traqurlZhYaHxw2dAPOD2YwCJKKTgsmbNGr355pt69NFHdeXKFf3hD3/w7xsxYoTS09O1bNkyzZs3T6tXr1ZxcbFqa2tVVVWlNWvWKCnp87uvHQ6Hli5dqsrKSmVlZfkXoDtz5owqKirCeoIAACBxhBRc3nrrLUnShg0b2u3buXOnJk+erPHjx2vLli2qqKjQ/v37lZ2drdWrVwesmitJCxculGVZ2rVrl86fP6+RI0dq+/btrJoLAAnCN9/qRnOp4nG+FeJXSMHljTfeCKpdYWGhCgsLO21js9m0ePFiLV68OJQuAAAMwXwrREK3F6ADAKAzzLdCOIW05D8AACbjSeXmY8QFANBjcOnKfAQXoAdi0iR6Mi5dmY3gAvRATJoEOhbOUE/wjwyCC9CD8ZsnECicoZ7nb0UGwQUAgC8JV6jn+Vvhx11FAADAGAQXAABgDIILAAAwBsEFAAAYg+ACAIg7NptNTqdTNhu3EyMQdxUBAGLmeuumOJ1OuVxf3CrMuinwIbgAAGKGdVMQKoILACDmWDcFwWKOCwAAMAbBBQAAGIPgAgAAjEFwAQAAxiC4AAAAYxBcAACAMQguQBxpa7Ni3QUAiGus4wLEERbQAoDOEVyAOMMCWgBwfVwqAtAh3zNkAATivRFbjLgA6FAwz5Dh0hV6It4bsUVwAdApLl0BHeO9ERtcKgIAAMYguAAA4hpzSnAtLhUBAOIac0pwLYILAMAIzCmBxKUiAABgEIILAAAwBsEFAAAYg+ACAACMQXABAADGILgAAABjEFyAKLHb7bLZbLHuBgAYjXVcgCiw2WxyucaoV6/kWHcFAIxGcAGipFevZFb+BIBuIrgAUcTKnwDQPcxxAQAAxiC4AAAAYxBcAACAMQguAADAGAQXAABgDIILEAUsPAcA4UFwAcKgrc3qdH9qamqUegIAiY11XIAwSEqysbgcAEQBwQUIExaXA4DI41JRDNzoskKwbQAA6GkYcYmBG11WGDIwQyvnToxyrwAAiH8Elxjp7LICAADoGJeKAACAMQguAADAGAQXAABgDIILAAAwBsEFAAAYg+ACAACMQXAJMxaOSzzUFADiB+u4hBnPrEk81BQA4gfBJQJ4Zk18aGuzlJRk63YbiZoCQLwguMShvhkpYf3Q7aluNFIyOidLpXflR7lXAIDuILjEoXSnnecZhcmNRkq4DAQAZiG4xDGeZxQdXAYCAHNwVxGiKpg7dLiLBwBwPYy4IKrCdQmM+T0A0DMRXBKIzWaT0+mUzRabD/Rw3KETLOamAEDPFNPgcurUKZWXl+udd96R0+nUrFmztHLlSqWmpsayW0bo6M4jp9MplyvwwzqaIxPRDhPMTQGAnidmwcXj8ej+++/XoEGD9Nxzz+nTTz/V008/rYsXL2rjxo2x6pYxwnnnUTytdxLsreAAgJ4pZsHl3/7t3+TxeLR//35lZWVJkpKTk7Vy5UotXbpUeXl5seradcXjB2o0LrtEc72TYAIZl4EAoOeKWXD5zW9+oylTpvhDiyR961vf0uOPP66ampq4DC6mfaCGMnoRb+udcBkIANARm2VZMbn3dMqUKSopKdHKlSsDts+aNUvjxo3TU089FfJr1tbWyrIs2e32cHUzgM1m06XPvLra2tbh/hR7stLT7HHX5nJTi1qvc4uxvVeSnCm94q7PtKFNvLeJxz7RhjaRbtMrOUl90h1qaWlRcnJy2G4GaWlpkc1m04QJE27YNmbBZcyYMVq+fLkefPDBgO3f/e531a9fP23atCnk1zxy5EhEgwsAAAg/X3AZP378DdvG3e3QlmV1OcEFc8IAAMBcMVs5NzMzUx6Pp932hoYGZWZmxqBHAAAg3sUsuOTl5cntdgds83q9On36dFxOzAUAALEXs+Aybdo0HTp0SH/5y1/8215//XV5vV4VFhbGqlsAACCOxWxyrsfjUVFRkQYPHqyysjJduHBB69at09SpU1mADgAAdChmwUUKXPI/NTVVRUVFLPkPAACuK6bBBQAAIBQxm+MCAAAQKoILAAAwBsEFAAAYg+ACAACMQXABAADGILgAAABjEFzi0KlTp7Ro0SKNGzdOU6ZMUXl5ua5cuXLD47xerzZs2KCpU6fqq1/9qu655x69/fbbHbb94IMPtGTJEk2cOFHjx4/Xd77zHdXW1ob7VBJOJGuzb98+jRo1qsP/Fi1aFKlTSgiRfs+cPXtWjzzyiKZOnarx48fr7rvv1q9+9atInErCiXRt6uvr9YMf/MD/s2zJkiU6c+ZMJE4lofzf//2ffvSjH+muu+6Sy+VSUVFR0Mf+8pe/1Le//W3l5+erqKhI1dXV7dq0tLTomWee0dSpUzV27Fj9wz/8g44fPx6Wvsfd06F7Oo/Ho/vvv1+DBg3Sc889p08//VRPP/20Ll68eMMVhdeuXat///d/1w9+8APl5uZq3759Ki0t1d69ezVmzBh/u+PHj2vu3Lm67bbbVFFRoV69eum9994L6odJTxbp2tx2223au3dvwHF1dXV67LHHNG3atIidl+kiXZfm5mZ/cHz88cfVt29fHThwQI8++qhSU1N1xx13RPwcTRXp2rS2tmrx4sVqamrSmjVrlJKSok2bNun+++/Xq6++qt69e0fjNI30wQcfqKamRmPHjlVbW5uCXdLttdde06pVq/Tggw/qG9/4hn79619rxYoVysjI0NSpU/3tnn76ae3fv1+rVq3S4MGD9fOf/1wLFizQq6++qv79+3ev8xbiys9+9jNr7Nix1oULF/zbfvWrX1kjR460Tpw4cd3jPv74Y2v06NHWzp07/dva2tqsoqIia8mSJQFt7733XuuRRx4Jf+cTXDRq82XPPfecNXr0aOuTTz7p/gkkqEjX5fe//701cuRI6+233w44fubMmdby5cvDdyIJKNK1OXDggDVy5Ejr+PHjAcfeeuut1ksvvRTek0kwra2t/v9/7LHHrFmzZgV13Le//W3r4YcfDti2cOFCa86cOf7vffXbvXu3f1tDQ4M1adIka8OGDd3suWVxqSjO/OY3v9GUKVOUlZXl3/atb31LDodDNTU11z3u+PHjam1tDUi8NptNU6dO1VtvvSWv1ytJcrvdOnLkiObNmxe5k0hQka5NRw4cOKCCgoLu/4aSwCJdl6tXr0qSMjIyAo7PyMgI+rfUnirStTl69Kj69++vUaNG+dsNHDhQt9xyi954440InFHiSEoK/eP/zJkzOnnyZLvLSkVFRfrTn/6kTz/9VJL01ltvqbW1VbNmzfK3SU9P1/Tp0zute7AILnHG7XYrLy8vYJvD4dCwYcPkdruve5zvjWy329sd6/V6dfbsWUnSH/7wB0lSQ0OD/9rm9OnTtWvXrjCeRWKKdG2+7N1331VdXV1I1557okjXZeLEiRoxYoQqKip05swZNTQ0aO/evfrzn/+s++67L8xnk1giXZvm5mY5HI52xzscDp08ebK73ceX+P5Oc3NzA7bn5eXJsiz/frfbrZtuukl9+/Zt1+7UqVNqa2vrVj8ILnHG4/EoMzOz3fbMzExdunTpuscNHz5ckvSnP/0pYLsvqPiOPX/+vCTp0Ucf1cyZM/Xiiy/q9ttvV3l5OZMNbyDStfmyAwcOKCUlhTkUNxDputjtdu3cuVMNDQ26/fbb9Td/8zf6yU9+onXr1mnKlCnhOYkEFena5OTk6OOPP1Z9fb2/zeXLl3XixAldvHixe51HO76/9y/XtE+fPgH7PR5PuxFKX7uWlhY1NjZ2qx9MzjWEZVmy2WzX3X/LLbdo0qRJ2rhxo7Kzs5WTk6N9+/bp97//vaQvhgV9SbekpETf+973JEkFBQU6ffq0tm3bpjvvvDPCZ5J4wlWba7W1tengwYO67bbblJ6eHrG+J7Jw1eXKlSt6+OGH1draqk2bNik9PV1vvPGGfvjDHyozM5OJ010QrtoUFRXpueee0w9/+EM9+eSTSklJ0fr169XY2Khevfh4i5Qv1853yfTa7R3VN1yXVhlxiTOZmZnyeDzttjc0NHT4m8u11q1bp6ysLH33u99VQUGB9uzZo7KyMknSTTfdJOmLZFxQUBBwbEFBgerq6tTS0hKO00hIka7NtQ4fPqxPPvlExcXF4el8Aot0XV555RX98Y9/1AsvvKAZM2ZoypQpeuKJJzRt2jRt2LAh/CeUQKLx86yiokInTpzQjBkzNG3aNH3yySeaPXt2h+8rdM+XR1Z8fDX21fR6dfd4PLLb7UpLS+tWP4ikcSYvL6/dtV+v16vTp0+rpKSk02MHDx6sV155RWfPntWVK1eUk5Ojl156Sf3799fgwYP9r389SUlJnf4W1NNFujbXevXVV5WRkaHCwsKwnkMiinRdTpw4oYEDBwZMMJWk0aNH67e//W14TybBROM9841vfENvvvmm6urq5HA4NHToUD344IMaN25cJE6pR/PNbTl58mTAZ4nb7ZbNZvPvz8vL04ULF3Tx4sWAeS5ut1s5OTldmhh8LUZc4sy0adN06NAh/eUvf/Fve/311+X1eoP+EBsyZIhGjBihlpYWvfLKK5ozZ45/3/jx49WnT592Czm9/fbbysvLY3i1E5GujY/X69Xrr7+uO+64o8OJhwgU6boMGjRI9fX1unDhQsAxf/7znzsMnfhCtN4zycnJysvL09ChQ+V2u/W73/2uw3bonqFDhyo3N1cHDx4M2H7gwAF99atf9Yf7qVOnKikpKWBhusuXL+uNN94Iyy9jfErFmfvuu0+7d+9WWVmZysrKdOHCBa1bt07FxcUBCffxxx/X/v37dfToUf+23bt3Kz09XX/913+tDz/8UC+99JJSUlJUWlrqb+NwOFRWVqaNGzcqIyNDY8eO1Ztvvqn//u//1ubNm6N6rqaJdG18ampq5PF4uEwUpEjX5c4779T27dtVWlqq0tJSZWZm6vXXX9ebb76pJ598MpqnapxovGc2bNigcePGKT09Xe+//762bt2q2bNnM3H6Bpqamvy3Jn/44Yf67LPP9Nprr0mSJk2apKysrA7r8vDDD2vFihUaNmyYvv71r+u//uu/9Nvf/lY///nP/W0GDhyo++67Txs3blSvXr00aNAgvfjii5Kk+++/v9t9J7jEmczMTP3iF79QeXm5HnroIaWmpqqoqEgrV64MaNfW1qbW1taAbV6vV5s2bdLHH3+svn376o477tDy5cvbXU9csGCBbDabdu7cqS1btmjo0KFav369br/99oifn8miURtJ/pUlJ0+eHNHzSRSRrkt2drZ27dqln/70pyovL1djY6NuvvlmlZeX65577onKOZoqGu+Zjz/+WE8++aQuXbqkwYMH63vf+15YPhwT3YULF7R8+fKAbb7vd+7cqcmTJ3dYl7/7u7/TlStXtG3bNu3YsUM333yzKisrA9bckaRVq1YpLS1NP/3pT9XQ0KCxY8fqF7/4RVjWpLJZrKAEAAAMwRwXAABgDIILAAAwBsEFAAAYg+ACAACMQXABAADGILgAAABjEFwAAIAxCC4AAMAYBBcAAGAMggsAADAGwQUAABjj/wERvT1coWNnyAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query_sims_umap = cosine_similarity(emb_umaps, ump.transform(np.atleast_2d(query_emb)))\n",
    "plt.hist(query_sims_umap, bins=50);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.0\n",
      "Recent technological advances, such as single-cell RNA sequencing (scRNA-seq), allow the measurement of gene expression profiles of individual cells. These expression profiles typically exhibit substantial variations even across seemingly homogeneous populations of cells. Two main different sources contribute to this measured variability: actual differences between the biological activity of the cells and technical measurement errors. Analysis of the biological variability may provide information about the underlying gene regulation of the cells, yet distinguishing it from the technical variability is a challenge. Here, we apply a recently developed computational method for measuring the global gene coordination level (GCL) to systematically study the cell-to-cell variability in numerical models of gene regulation. We simulate 'biological variability' by introducing heterogeneity in the underlying regulatory dynamic of different cells, while 'technical variability' is represented by stochastic measurement noise. We show that the GCL decreases for cohorts of cells with increased 'biological variability' only when it is originated from the interactions between the genes. Moreover, we find that the GCL can evaluate and compare-for cohorts with the same cell-to-cell variability-the ratio between the introduced biological and technical variability. Finally, we show that the GCL is robust against spurious correlations that originate from a small sample size or from the compositionality of the data. The presented methodology can be useful for future analysis of high-dimensional ecological and biochemical dynamics.\n",
      "\n",
      "1.0\n",
      "In high-throughput studies, hundreds to millions of hypotheses are typically tested. Statistical methods that control the false discovery rate (FDR) have emerged as popular and powerful tools for error rate control. While classic FDR methods use only p values as input, more modern FDR methods have been shown to increase power by incorporating complementary information as informative covariates to prioritize, weight, and group hypotheses. However, there is currently no consensus on how the modern methods compare to one another. We investigate the accuracy, applicability, and ease of use of two classic and six modern FDR-controlling methods by performing a systematic benchmark comparison using simulation studies as well as six case studies in computational biology.\n",
      "Methods that incorporate informative covariates are modestly more powerful than classic approaches, and do not underperform classic approaches, even when the covariate is completely uninformative. The majority of methods are successful at controlling the FDR, with the exception of two modern methods under certain settings. Furthermore, we find that the improvement of the modern FDR methods over the classic methods increases with the informativeness of the covariate, total number of hypothesis tests, and proportion of truly non-null hypotheses.\n",
      "Modern FDR methods that use an informative covariate provide advantages over classic FDR-controlling procedures, with the relative gain dependent on the application and informativeness of available covariates. We present our findings as a practical guide and provide recommendations to aid researchers in their choice of methods to correct for false discoveries.\n",
      "\n",
      "1.0\n",
      "Perturbations in the environment lead to distinctive gene expression changes within a cell. Observed over time, those variations can be characterized by single impulse-like progression patterns. ImpulseDE is an R package suited to capture these patterns in high throughput time series datasets. By fitting a representative impulse model to each gene, it reports differentially expressed genes across time points from a single or between two time courses from two experiments. To optimize running time, the code uses clustering and multi-threading. By applying ImpulseDE , we demonstrate its power to represent underlying biology of gene expression in microarray and RNA-Seq data.\n",
      "ImpulseDE is available on Bioconductor ( https://bioconductor.org/packages/ImpulseDE/ ).\n",
      "niryosef@berkeley.edu.\n",
      "Supplementary data are available at Bioinformatics online.\n",
      "\n",
      "1.0\n",
      "Finding a causal gene is a fundamental problem in genomic medicine. We present a causal inference framework, CoCoA-diff, that prioritizes disease genes by adjusting confounders without prior knowledge of control variables in single-cell RNA-seq data. We demonstrate that our method substantially improves statistical power in simulations and real-world data analysis of 70k brain cells collected for dissecting Alzheimer's disease. We identify 215 differentially regulated causal genes in various cell types, including highly relevant genes with a proper cell type context. Genes found in different types enrich distinctive pathways, implicating the importance of cell types in understanding multifaceted disease mechanisms.\n",
      "\n",
      "1.0\n",
      "Analysis of gene expression data is widely used in transcriptomic studies to understand functions of molecules inside a cell and interactions among molecules. Differential co-expression analysis studies diseases and phenotypic variations by finding modules of genes whose co-expression patterns vary across conditions. We review the best practices in gene expression data analysis in terms of analysis of (differential) co-expression, co-expression network, differential networking, and differential connectivity considering both microarray and RNA-seq data along with comparisons. We highlight hurdles in RNA-seq data analysis using methods developed for microarrays. We include discussion of necessary tools for gene expression analysis throughout the paper. In addition, we shed light on scRNA-seq data analysis by including preprocessing and scRNA-seq in co-expression analysis along with useful tools specific to scRNA-seq. To get insights, biological interpretation and functional profiling is included. Finally, we provide guidelines for the analyst, along with research issues and challenges which should be addressed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sdf = df[[\"abstract\"]].reset_index(drop=True).join(DataFrame(dict(sim=query_sims_umap[:,0]))).sort_values('sim', ascending=False)\n",
    "for i in range(5):\n",
    "    print(round(sdf.sim.values[i], 3))\n",
    "    print(sdf.abstract.values[i])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9652, 20)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "emb_umaps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Embed PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 24.2 s, sys: 709 ms, total: 24.9 s\n",
      "Wall time: 4.55 s\n"
     ]
    }
   ],
   "source": [
    "pca = PCA()\n",
    "%time emb_pcs = pca.fit_transform(embs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjkAAAG2CAYAAACd5Zf9AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAABGG0lEQVR4nO3deXhU5fnG8XuyTDYSIKyyhCUYIOxghACCijugolapQqmiVanaqlgrVeuCQq3WFqxVAddqUaulP63YYlVwAdsCCsgiCWFHCNkmyyQzmXl/fyQZCEkgk0wy2/dzXbkmc+acM88TIHPznvecYzHGGAEAAISYCH8XAAAA0BIIOQAAICQRcgAAQEgi5AAAgJBEyAEAACGJkAMAAEISIQcAAIQkQg4AAAhJUf4uoDk2btwoY4yio6P9XQoAAGgkp9Mpi8WiESNGtOj7BPVIjjFGLXHBZmOMHA5Hi+w70NBr6AmXPiV6DVXh0mu49CnV7bWlPr9PFNQjOTUjOEOGDPHpfsvKyrRt2zb169dP8fHxPt13oKHX0BMufUr0GqrCpddw6VOq2+vmzZtb5X2DeiQHAACgIYQcAAAQkgg5AAAgJBFyAABASCLkAACAkETIAQAAIYmQAwAAQhIhBwAAhCRCDgAACEleh5w9e/bowQcf1GWXXab09HRNmTKl0dv+7W9/00UXXaQhQ4ZoypQpWrlypbdvDwAA0Cheh5ydO3dq9erV6tWrl1JTUxu93Ycffqhf/vKXOv/887VkyRKNGTNGd955pz7//HNvSwAAADglr+9dde655+q8886TJP3yl7/Uli1bGrXdH/7wB1100UW6++67JUljxoxRTk6OFi1apPHjx3tbBgAAwEl5PZITEeH9NJ59+/Zp165ddQ5tTZkyRZs2bVJ+fr7X+wQAADiZVrkL+a5duyRJffv2rbU8NTVVxhjt2rVLycnJTdq3MUZlZWXNrvF4dru91mMoo9fQEy59SvQaqgK1V2OMjJHcbiO3Mcc96oTnRq7jX2tgfbu9XAcPl6vUHFJUdLTMKdav2a8xRm5TXY8kU/29u7q+mufGqHpZ/cslHbeNOW4/x57379VOZw07rdk/uxP/TI0xslgszd7vqbRKyCkqKpIkJSUl1Vretm3bWq83hdPp1LZt25pe3Ens3r27RfYbiOg19IRLnxK9BpOqD2qp0m3kcpmqR7eOfX/csh37t1Uvk1zumuVVH/qex+NCgMsTEI69fvxylycs1GxX872OCybHrXv8vqpDQss42lI7brZPNx5QclSBIiJ8E0iO//trtVp9ss+TaZWQU+PE1Gaq/8Y0J81FR0erX79+zarrRHa7Xbt371bv3r0VFxfn030HGnoNPeHSp0SvTeV2Gzkq3XI4XccenW45KqsendWPxy87tm7VMmfN98ftw1npVqXLLWelqX50y+mqWlZZ6ZbTVRUsQpXFIkVYLIqIqP6yWBQRcdwyz6MUEWGRRVJlpVOxMVZFRUZWrVtrvdrr1zy3VO/XIovnPS2Wqs/SmseIWs9rllU/17HXqpaduN6xz+X+Ke00qF+HZv9sTvz7m5WV1ex9NkarhJzjR2w6duzoWW6z2STVHeHxhsViUXx8fPMKbEBcXFyL7TvQ0GvoCZc+pdDttdLllr2iUvaKShUWu3QgzyF3rF3G4lR5RaXKHa5jj46q9SocLtmPW1Z3PZcqXW5/t+YRFRmh6Kiqr5rvIyMsclU61CYhTjHWqFqvRUVWfUVGWqoeIyyKrH6sszwiQlGRx79e9X1URNV6NdvWXufYY8Rx+6zZ37EAUxU8Ik8IH94oKyvTtm3bNHDgwJD8+1ufmn+rrXGoSmqlkFMzF2fXrl21TjvPzs6WxWKpM1cHAIKZ221kr6hUWXmlysqdKiuvVGm5U/bqR8/yikqV2p2yVz+WVVSq7LhHR2V9YeSIT2uNirTIGh3p+YqJjqj6PipSMZ7lEdWvHXseU8820VGRVYElMkJRUXXDS+3vIxUVWX8wCMcPf7SMVgk5PXv2VN++ffXBBx/o/PPP9yx///33NXTo0CZPOgaAllThdKmkzKGSMqdK7E4Ve753qLjM2eBrpXanfHlUJioyQrHWSEVY3EpMiFVcbLTirFGKjYlUrDVKsdZIxcVEKeb4R2uUYmOqXouNiVKctWp5jDWyVkCJ9NFcCyAQeR1y7Ha7Vq9eLUk6cOCASkpK9OGHH0qSzjzzTCUnJ2vevHlasWKFtm7d6tnujjvu0J133qmUlBSNHTtW//73v/XFF19o6dKlPmoFABrmdhuVljtVVFKhohJH1WOpQ7bqx6LiChWVVr1mK3WopMzRwEhK40VFWhQfG62E2GjFxUYpITZa8bFR1V/RnseE2CjFVT9POG55fGyU4mKiFBUZwegG0AReh5y8vDz97Gc/q7Ws5vmrr76q0aNHy+12y+Vy1Vrn4osvVnl5uZ577jktW7ZMvXr10tNPP82FAAE0mTFGZRUu7f2+WGXOEhXYypVvK1dBcYUKiytUVFIhW+mxQNOUSa8RFqlNvFVt4qKVGG9Vm/hotYmzKjE+WgnxVcsSq5e1qX7eJi5aCXHRskZHtkDXABrL65DTo0cP7dix46TrLFy4UAsXLqyzfNq0aZo2bZq3bwkgzBhjVGJ36mihXUcL7cq3VaiguDrA2MpVYKtQfnHV95UuI+lQo/cdHxultm1i1DbBqrZtYpSUYFW7xBglJcSobRur2iZULUtMqAorcTFRPjt9FkDratVTyAFAksorKpVbHWBqHmu+zy2w62iRXRUO16l3VC0xPlrJSbFqnxRb9ZgYo/ZJsWqbYFVSmxi1a1MVYJISrIqOYnQFCBeEHAA+V+6o1JH8Mn2fX6bDeWX6Pr9Uh/PKdDi/TEcL7SqxOxu1n7ZtrOrQNs4TXJKTYpXcNlbtE2OVnBSjuGijQ/tzNGRwOvNUANRByAHgNWOM8m3lOni0VIfzSo+FmbxSHc4vU0FxxSn3ER8bpY7t4tSxXZw6VX8d/7xDuzjFnGJOS1lZmXIPcSgJQP0IOQAaVFbu1IHcEh3ILdWBIyU6mFui/blVj+WnOJyUEBulLh0S1CU5Xl07JKhrh3h1bh+vTu2rQkx8bHQrdQEgXBFygDBnjFFBcYX2fm/Tnu+Lte9wcVWwOVJy0hGZiAiLurSPV9cO8erSIUFdk+PVpUO8uiZXBZo28S1/XxoAOBlCDhBGbKUOT5jZ871Ne78v1t7vbSoua3iOTLvEGHXv1Oa4rwR179xGXZITFB0V0YrVA4B3CDlACHIbo0N5pTq4o0DZBwq160CRdh+yNTgyE2GRTuuYoJSuSUrpkqjunasCTbdObdQmjsNKAIITIQcIcpUut/YdLlb2/iJlHyhU1r4C7TpQJEflgXrX75wcr5QuierVNVG9TqsKNT26JJ5yki8ABBtCDhBEjDE6dLRUO/YW6Ls9Bdqxt0A5B2313lU6OipCfbolqW/3durbva36dKsKNEz4BRAuCDlAACsuc+i74wLNd3sL6p0/kxAb5QkzPTrFytiPavyZg5WY2MYPVQNAYCDkAAHkSH6ZtuzK09acPG3JztOB3JI660RHRSi1e1ul9WqvASnJOj2lnbokx8tiqbpeTNWNHIsUGcmkYADhjZAD+IkxRvuPlFSFml152rIrT0cL7XXW69YxQWm92qt/Snv179VevU9ry1lNANAIhByglZjqM56+/i5XX3+Xq2935clW6qi1TmSERf16tFN63w4a3LeDBvROVlIC15sBgKYg5AAtyFbq0Dc7q0LN1ztzdSS/rNbr1qgIDeidrPQ+VaGmf6/2io3hnyUA+AK/TQEfcrnc2ro7X+u3HdbXO3O160CRjDn2elSkRQN6J2t4WicNTe2kfj3bcegJAFoIIQdopuIyh9ZvP6L/bv1e67cfUekJd9ju1TVRw9M6a3haJw3q20FxjNQAQKvgty3gpZoJw//d+r3+s/Wwtu3Ol9t9bLgmMd6qUQM7a2T/zhp2eiclJ8X6sVoACF+EHKARjDHadaBIX2w6qC83HdSB3NJar/fqmqiM9K46M72r0nq1V2SExU+VAgBqEHKABhhjtHNfob7cdFBfbDqo7/OOTRqOiozQ0H4dlZHeRRnpXdUlOd6PlQIA6kPIAU6Qc7BIn67fr8+/OaAjBceuW2ONjtQZAztr3NBuOmNgF26PAAABjpADSMorsmv1hgP6ZP0+7T5k8yyPsUYqY2AXjRvWTaMGdGHSMAAEEX5jI2zZKyq1dvNBffK//fomK9dzqndUpEUZ6V01cWQPjRrQWbFW/pkAQDDitzfCijFGWfuL9OnGHfrs6wMqd7g8rw3snaxzzuip8cO6KTGeqwwDQLAj5CAslNqd+tdX+/SPz4/ocOEBz/LTOibonFE9dc6oHuraIcGPFQIAfI2Qg5C2Y0++Ply7R2u+PiCHs2rUJjoqQuOHddOFY3orvU+y5+7dAIDQQshByHFWuvXFpoN677Nsfbe30LO8R+cEDe4ZpR9cMEKdO7b1X4EAgFZByEHIKCqp0Idrd+uDL3OUb6uQVHU9mwkjuuvizN7q2SlG27dvV5t4Tv0GgHBAyEHQ23e4WH/7NEufbtgvZ6VbktQ+MUaXjOuji8b0VrvEGElSWVnZyXYDAAgxhBwErax9hXr74++0dvMhz+nf/Xq206Vn9dX4Yd25uzcAhDlCDoKKMUZbsvP09r+/08bvcj3LRw/qqivO6aeBvZlIDACoQshBUDDGaON3ufrLP7dr+54CSVJEhEUTRnTXVeecrl6nJfm5QgBAoCHkIOB9uytPr63cpm935UmqOgX8vDNTdMXZ/bi2DQCgQYQcBKzv9hbo9Q+3a8OOI5Kqws3FY3vrqnNOV/ukWD9XBwAIdIQcBJyDR0v08vtbtXbzIUlSZIRF54/upWvOS1PHdnF+rg4AECwIOQgYJXan3ly1Q+9/vkuVLqMIi3T2qJ764QX9OSwFAPAaIQd+V+ly68O1u/XGP3eouMwhSRo1oLOunzpIvboyoRgA0DSEHPjVNztz9fzfNmnf4RJJUs8uibrx0sEaOaCznysDAAQ7Qg78osBWrhff+1afbtgvSWrbxqrrLhygC0b3UmQkF/EDADQfIQetyuU2+vDLHL22cptKyytlsUiTx/bRdRcPVJs47ikFAPAdQg5aze5DNi16c6N27iuUVHULhp9eOUz9erbza10AgNBEyEGLq3S59c7HO7V81Q5VuowSYqM085J0XZTZW5ER3IIBANAyCDloUTkHi/SHNzcqe3+RpKp7TM25apiSuZgfAKCFEXLQIlxuo79+/J2W/6tq9KZNXLRunjZEE0f24AaaAIBWQciBz+UW2PXUG+s995oaPairfnrVMG7FAABoVYQc+NTazQe16M2vVWJ3Ki4mUrdcMVTnjOrJ6A0AoNURcuATFU6Xlv19i1au3S2p6sype2aMUreObfxbGAAgbBFy0GyHjpZqwSv/Uc5BmyTpynP66bqLBio6iov6AQD8h5CDZvnftsN68vX1KrU71a5NjO66dqRG9OeWDAAA/yPkoEncbqM3V+3QX1btkDFS/17t9csfZahjuzh/lwYAgCRCDpqgrNypJ19fr/9uPSxJunhsb9102WBFR0X6uTIAAI4h5MArRwrK9Oiyr7T7kE3WqAjNuWqYJmWk+LssAADqIOSg0XbuK9Cjy75SQXGF2ifG6P4bRistpb2/ywIAoF6EHDTK2s0H9eTrG+RwutT7tCQ9MHu0OreP93dZAAA0iJCDU/q/z7K19O9bZIw0akBn/WLmGYqPjfZ3WQAAnBQhBw0yxuiNf+7Q8lU7JEmTx/XRTZcNVmQk178BAAQ+Qg7q5XYbLVmxWe9/kSNJmnHxAF09KY3bMwAAggYhB3VUutz6w/KN+nTDflks0s3ThmryuD7+LgsAAK8QclBLpcutJ177n9ZuPqTICIvu/OFITRzZw99lAQDgNUIOPCpdbv32z1UBJzoqQvfNylBGeld/lwUAQJN4PYM0JydHs2fP1vDhw5WZman58+ervLz8lNuVlZXpySef1Hnnnadhw4bpggsu0OLFi+VwOJpUOHzL5XLrydfX68tNhxQVGaF5Pz6TgAMACGpejeTYbDbNmjVL3bp106JFi5Sfn68FCxaosLBQTz755Em3feihh/TRRx/pzjvv1Omnn65NmzZp0aJFKioq0v3339+sJtA8LrfR4r9u0dothxUVadG8H2fojIFd/F0WAADN4lXIWb58uWw2m1asWKHk5GRJUmRkpObOnatbb71Vqamp9W5XWVmpDz/8UDfeeKNmzpwpSRozZowOHjyoDz74gJDjR2630Yq1+dq8x66oSIvum8UIDgAgNHh1uGrNmjXKzMz0BBxJuvDCC2W1WrV69eoGtzPGyOVyKTExsdbypKQkGWO8LBm+YozRKyt3aPMeuyIjLLr3Rxk6cxABBwAQGrwKOdnZ2XVGa6xWq1JSUpSdnd3gdtHR0briiiv02muv6ZtvvlFpaanWrVunt956S9ddd13TKkez/fXjnfpw3T5J0k+vHKwxg0/zc0UAAPiO13NykpKS6ixPSkpSUVHRSbd96KGH9Otf/1pXX321Z9nMmTN12223eVNCHcYYlZWVNWsfJ7Lb7bUeQ9En6w/o1Q+2SZIuGtlWI09v6/OfY6AJhz9XKXz6lOg1VIVLr+HSp1S3V2NMq1xc1ienkDem2CeffFKffvqpHn30UfXp00fffvutFi1apKSkJN1xxx1Nfm+n06lt27Y1efuT2b17d4vs1992HLBr+Zo8SdL49ESNGZAYsr3WJ1x6DZc+JXoNVeHSa7j0KdXu1Wq1tvj7eRVykpKSZLPZ6iwvLi5ucNKxJH333Xd68cUX9eyzz2rSpEmSpIyMDFksFj3xxBO67rrr1KFDBy9LrxIdHa1+/fo1aduG2O127d69W71791ZcXJxP9+1vuw8V6923/yNjpIkjuunHF/XRnj17QrLXE4Xyn+vxwqVPiV5DVbj0Gi59SnV7zcrKapX39SrkpKam1pl743A4tHfvXl155ZUNblfTzMCBA2stHzhwoCorK3XgwIEmhxyLxaL4+PgmbXsqcXFxLbZvfygoLtdv3/hGFU63hp/eST//4Sg5KqqucRRqvZ5MuPQaLn1K9BqqwqXXcOlTOtZra90H0auJxxMmTNC6detUUFDgWbZq1So5HA5NnDixwe26d+8uSfr2229rLd+yZYskqUcPbhvQ0pyVLj3+0n90tNCu7p0SdO+PzlAUdxMHAIQwr0Zypk+frj//+c+aM2eO5syZo7y8PC1cuFBTp06tdbhq3rx5WrFihbZu3SpJGjx4sIYOHapf//rXOnr0qPr06aPNmzfr2Wef1SWXXFLrlHT4njFGz7z9jbbvKVBCXLQemD1GbeJb/lgoAAD+5PWcnFdeeUXz58/X7bffrtjYWE2ZMkVz586ttZ7b7ZbL5fI8j4yM1HPPPac//OEPWrJkiY4eParTTjtNM2bM0C233OKbTtCg9z7bpY//t08RERbdO/MMde/Uxt8lAQDQ4rw+u6pPnz5atmzZSddZuHChFi5cWGtZhw4d9Mgjj3j7dmimbTn5evG9qsOEN0wdpBH9O/u5IgAAWgeTMkJYYXGFfvPaf+VyG40f1k2XntXX3yUBANBqCDkhyuU2evL1/ymvqFzdO7XR7VcPb7XZ7AAABAJCToh6a9UOfbPzqGKskbrvxxmKj432d0kAALQqQk4I2r47X8tX7ZAkzblymHp1rXsrDgAAQh0hJ8SUlTv11Bvr5TbSxBE9dO4ZPf1dEgAAfkHICTFLVmzR93ll6tQ+TrdcOdTf5QAA4DeEnBDyxaaD+ui/e2WxSHf+cKTaxDEPBwAQvgg5IaKopELP/vUbSdKV55yuIakd/VwRAAD+RcgJES+s2CxbqUO9T0vStRcO8Hc5AAD4HSEnBPzn2++1ZuMBRVik268erugo/lgBAODTMMiV2p169p2qw1SXTeyntJT2fq4IAIDAQMgJci+9/63yisp1WscEXXthf3+XAwBAwCDkBLHtu/P1z3V7JEm3/2C4Yq1e328VAICQRcgJUi630Z/e2SRJOi8jRUP6cTYVAADHI+QEqQ+/zNGug0VKiIvWrMnp/i4HAICAQ8gJQkUlFXrtw+2SpJkXDVC7xBg/VwQAQOAh5AShV/6xVaV2p/p2b6uLxvbxdzkAAAQkQk6Qyd5fqI/+u1eSdOsVQxUZYfFzRQAABCZCTpB5+R9bZYw0YUR3Deid7O9yAAAIWIScILJhxxF9/V2uoiItmnnxQH+XAwBAQCPkBAmX2+jl97+VJE0e11ddOyT4uSIAAAIbISdIrN6wTzkHbUqIjdLV56X5uxwAAAIeIScIOCvd+nP1KeM/mJSmpASrnysCACDwEXKCwEf/3avcAruSk2I05ay+/i4HAICgQMgJcM5Kt9766DtJ0lXnpikmOtLPFQEAEBwIOQHuo//u1dHCqlGcC8f08nc5AAAEDUJOADtxFMfKKA4AAI1GyAlgjOIAANB0hJwA5XIbvfvJTknSleeczigOAABeIuQEqLWbD+r7vDIlxkfrgtGM4gAA4C1CTgAyxujdT7IkVV3dODYmys8VAQAQfAg5AWhLdp527iuUNSpCU8b38Xc5AAAEJUJOAHqnei7OeWemqG2bGD9XAwBAcCLkBJg939u0fvsRRVikyyf283c5AAAELUJOgPnH5zmSpNGDT9NpHbnTOAAATUXICSCldqc+Wb9PkjR1PPeoAgCgOQg5AeTf/9urcodLKV0TNTi1g7/LAQAgqBFyAoTbbfTBF1WHqiaP6yOLxeLnigAACG6EnADxzc5cHcgtVXxslM4Z1dPf5QAAEPQIOQHiH9WjOOee0VNxXPwPAIBmI+QEgHxbuf677bAk6ZKxXPwPAABfIOQEgE/+t09ut9HA3snq2SXR3+UAABASCDl+ZozRR//dK0malJHi52oAAAgdhBw/27G3QPuPlMgaHamzhnfzdzkAAIQMQo6fffSfqlGccUNPU3xstJ+rAQAgdBBy/KjcUanPvj4gqepmnAAAwHcIOX60bvMhlZVXqnNyvAb37ejvcgAACCmEHD/6dMN+SdKkM3oqIoIrHAMA4EuEHD8pKqnQxu9yJUkTR/bwczUAAIQeQo6ffP7NQbndRv16tFX3Tm38XQ4AACGHkOMnazZWHapiFAcAgJZByPGDIwVl2pqTL4tFOmt4d3+XAwBASCLk+MHn1aeND+7bUR3axvm5GgAAQhMhxw++3HRIkrjCMQAALYiQ08ryiuzasbdAFos0evBp/i4HAICQRchpZeu2fC9J6p/SXslJsX6uBgCA0EXIaWXrtlQdqsocwigOAAAtiZDTikrKHNqcdVSSNIZDVQAAtChCTiv677bDcrmNUromqhsXAAQAoEV5HXJycnI0e/ZsDR8+XJmZmZo/f77Ky8sbtW1hYaEeeughjR8/XkOGDNGFF16o5cuXe110sFq7ufpQFaM4AAC0uChvVrbZbJo1a5a6deumRYsWKT8/XwsWLFBhYaGefPLJk25bWlqqmTNnKiYmRvPmzVOHDh20Z88eOZ3OZjUQLCqcLm3YcUSSNIb5OAAAtDivQs7y5ctls9m0YsUKJScnS5IiIyM1d+5c3XrrrUpNTW1w2+eff17l5eV6++23FRtbdVbR6NGjm1F6cNmcdVQVDpc6totTave2/i4HAICQ59XhqjVr1igzM9MTcCTpwgsvlNVq1erVq0+67TvvvKOrrrrKE3DCzfrthyVJZwzsIovF4udqAAAIfV6FnOzs7DqjNVarVSkpKcrOzm5wu3379uno0aNKSkrSzTffrMGDB2v06NF6+OGHGz2fJ9it3151qGpk/85+rgQAgPDg9ZycpKSkOsuTkpJUVFTU4HZHj1adNv3EE0/ooosu0pIlS5SVlaXf/e53cjqdmj9/vpdlH2OMUVlZWZO3r4/dbq/12Fzf55Xp0NFSRUZYlNYjwef1Noevew1k4dJruPQp0WuoCpdew6VPqW6vxphWOarhVchpyKmKdbvdkqTU1FQtWLBAkpSZmanKyko98cQT+tnPfqZOnTo16b2dTqe2bdvWpG1PZffu3T7Zz1c7SiRJPTtatXvXTp/s09d81WswCJdew6VPiV5DVbj0Gi59SrV7tVqtLf5+XoWcpKQk2Wy2OsuLi4tPOum4Xbt2kqQxY8bUWj5mzBi53W5lZ2c3OeRER0erX79+Tdq2IXa7Xbt371bv3r0VF9f8u4T/3/82SpLGDk/RwIG9m70/X/J1r4EsXHoNlz4leg1V4dJruPQp1e01KyurVd7Xq5CTmppaZ+6Nw+HQ3r17deWVVza4Xc+ePRUdHV1nuTFGkhQR0fRrElosFsXHxzd5+5OJi4tr9r4dTpe25BRIksYM6d5itTaXL3oNFuHSa7j0KdFrqAqXXsOlT+lYr611Ao5X6WLChAlat26dCgoKPMtWrVolh8OhiRMnNrid1WrVuHHjtHbt2lrL165dq6ioKJ+PxASSrTl5cjhdSk6KVe/T6s5nAgAALcOrkDN9+nQlJiZqzpw5+uyzz7RixQo9+uijmjp1aq3DVfPmzVN6enqtbX/6059qx44d+sUvfqHPP/9cL7/8shYvXqzrrruu1inpoWZT9b2qhp3ekVPHAQBoRV7PyXnllVc0f/583X777YqNjdWUKVM0d+7cWuu53W65XK5ay4YOHarnn39eTz31lG655Ra1a9dOM2bM0M9+9rPmdxHAam7IObRfRz9XAgBAePH67Ko+ffpo2bJlJ11n4cKFWrhwYZ3l48aN07hx47x9y6Blr6jUzn2FkqQh/Zo2sRoAADQNdyFvQdty8uVyG3VuH6cuyeExqQwAgEBByGlBm7JyJUlDOFQFAECrI+S0oM3ZzMcBAMBfCDktpKzcqaz9Vbe6GJLKfBwAAFobIaeFbM3Jl9tt1LVDvDq1D+0rWQIAEIgIOS1k2+58SdKgvh38XAkAAOGJkNNCtleHnIG9Q/dChwAABDJCTgtwudzasbfq1hcDCDkAAPgFIacF5ByyqcLhUkJslHp2TvR3OQAAhCVCTguoOVTVv3eyIiK4XxUAAP5AyGkB25iPAwCA3xFyWoAn5PQi5AAA4C+EHB87WmhXboFdERbp9JR2/i4HAICwRcjxsR17qs6q6n1aW8XHRvu5GgAAwhchx8d27qsKOYziAADgX4QcH8uuvl9Vvx7t/FsIAABhjpDjQ8YYZe0vlETIAQDA3wg5PnQ4v0wldqeiIi3qdRoXAQQAwJ8IOT5Uc6iq12lJio6K9HM1AACEN0KOD3GoCgCAwEHI8aGakJNKyAEAwO8IOT5ijFG2ZySnrX+LAQAAhBxfyS2wq7isatJx79OS/F0OAABhj5DjI7u/t0mSenROZNIxAAABgJDjI3sOVYWcXl0ZxQEAIBAQcnxkd03I4fo4AAAEBEKOj3hGcpiPAwBAQCDk+ICz0q39R0okSb05XAUAQEAg5PjAgdwSudxG8bFR6tQ+zt/lAAAAEXJ84vhJxxaLxc/VAAAAiZDjE3u+Zz4OAACBhpDjA54zq7pyZhUAAIGCkOMDnFkFAEDgIeQ0U1m5U0cK7JLE7RwAAAgghJxm2nu4WJKUnBSjxHirn6sBAAA1CDnNtP9w1fVxenRmPg4AAIGEkNNMB3JrQk4bP1cCAACOR8hppv1Hqg5XdSfkAAAQUAg5zVRzOwcOVwEAEFgIOc1Q6XLr+7xSSRyuAgAg0BBymuFwfpkqXUbW6Eh1bMs9qwAACCSEnGbYX336eI9ObRQRwT2rAAAIJIScZuDMKgAAAhchpxmOTTom5AAAEGgIOc1QE3I4fRwAgMBDyGkGT8jpRMgBACDQEHKaqNTuVHGZQ5J0WscEP1cDAABORMhpoprr47RtY1V8bLSfqwEAACci5DTR9/llkqSuHRjFAQAgEBFymuhw9UhO12RCDgAAgYiQ00Tf59WM5MT7uRIAAFAfQk4T1czJIeQAABCYCDlNVDOS04U5OQAABCRCThO4XG4dKag+XMWcHAAAAhIhpwmOFpXL5TaKioxQcttYf5cDAADqQchpgpr5OF2S4xTJ3ccBAAhIhJwmOHZmFYeqAAAIVIScJjh2ZhUhBwCAQEXIaQJOHwcAIPB5HXJycnI0e/ZsDR8+XJmZmZo/f77Ky8u92seqVavUv39/TZkyxdu3Dwi5BXZJUqf2hBwAAAJVlDcr22w2zZo1S926ddOiRYuUn5+vBQsWqLCwUE8++WSj9lFeXq4FCxaoY8eOTSo4EOQWVoecdnF+rgQAADTEq5CzfPly2Ww2rVixQsnJyZKkyMhIzZ07V7feeqtSU1NPuY/nn39e3bp1U48ePbRly5amVe1HlS63CoqrRq4IOQAABC6vDletWbNGmZmZnoAjSRdeeKGsVqtWr159yu337t2rl156Sffff7/3lQaIvKJyGSNFRUaobZsYf5cDAAAa4FXIyc7OrjNaY7ValZKSouzs7FNu/9hjj+myyy7TgAEDvKsygBytPlTVsV2sIrhGDgAAAcvrOTlJSUl1liclJamoqOik23788cfauHGjPvzwQ+8qPAVjjMrKyny6T7vdXuvxePsPF0qSkpNifP6+/nCyXkNNuPQaLn1K9BqqwqXXcOlTqturMUYWS8sPFHgVchpyqmIrKir0+OOP6/bbb691qMsXnE6ntm3b5tN91ti9e3edZdt22iRJUaaixd7XH+rrNVSFS6/h0qdEr6EqXHoNlz6l2r1ardYWfz+vQk5SUpJsNlud5cXFxSeddPzKK68oIiJCkydP9mzvdDrldrtls9kUGxvb5Gajo6PVr1+/Jm3bELvdrt27d6t3796Ki6s9ufjLrG2SbErt1UUDB/r2ff3hZL2GmnDpNVz6lOg1VIVLr+HSp1S316ysrFZ5X69CTmpqap25Nw6HQ3v37tWVV17Z4Ha7du3Snj17lJmZWee1jIwMPfTQQ/rhD3/oTSkeFotF8fEtc72auLi4OvsuKHFKkk7rlNRi7+sP9fUaqsKl13DpU6LXUBUuvYZLn9KxXlvjUJXkZciZMGGC/vSnP6mgoEDt27eXVHVhP4fDoYkTJza43U033aRp06bVWvbCCy8oJydHCxYsUO/evb2v3E+Oco0cAACCgldnV02fPl2JiYmaM2eOPvvsM61YsUKPPvqopk6dWutw1bx585Senu55npqaqtGjR9f66tSpk+Lj4zV69Gh16dLFdx21MEIOAADBwes5Oa+88ormz5+v22+/XbGxsZoyZYrmzp1baz232y2Xy+XTQgNBeUWlisuqDld1JOQAABDQvD67qk+fPlq2bNlJ11m4cKEWLlx4ynWCTc3tHOJjo5QQF+3nagAAwMlwF3IvHLsQIKM4AAAEOkKOF7gxJwAAwYOQ44U8RnIAAAgahBwv5BdXSJKSk2L9XAkAADgVQo4X8ovKJRFyAAAIBoQcL+Tbqg5XJbcl5AAAEOgIOV7It1Ufrkok5AAAEOgIOY3kchsVFlcfrmIkBwCAgEfIaaSikgq5jRRhkdq2ifF3OQAA4BQIOY2Ub6saxWmXGKPIiNa5eyoAAGg6Qk4j1YQczqwCACA4EHIa6djp41wIEACAYEDIaaSC6pGc9knMxwEAIBgQchoprzrkdOBwFQAAQYGQ00ieOTmcPg4AQFAg5DTSscNVhBwAAIIBIaeROLsKAIDgQshpBLfbqLDEIUlqn8jEYwAAggEhpxFKy51yu40kKSmBkAMAQDAg5DRCYXHVjTkTYqMUHcWPDACAYMAndiPYSqsOVXHPKgAAggchpxEKS6pGcgg5AAAED0JOI9iqQ05SgtXPlQAAgMYi5DRCUfXhqnacWQUAQNAg5DRCUTEjOQAABBtCTiMUMfEYAICgQ8hphCImHgMAEHQIOY3gCTkcrgIAIGgQchqBiccAAAQfQs4puN3GczFAJh4DABA8CDmnUGLnvlUAAAQjQs4p1MzHSYiL5r5VAAAEET61T4FJxwAABCdCzilwjRwAAIITIecUuG8VAADBiZBzCsVlTkmEHAAAgg0h5xSKy6oOV7WJJ+QAABBMCDmnUGqvGslpExft50oAAIA3CDmnUDOSkxhPyAEAIJgQck6hZk4Oh6sAAAguhJxT4HAVAADBiZBzCscOVzGSAwBAMCHknMKxw1WM5AAAEEwIOSfhcLrkcLokMZIDAECwIeScRGl5pSQpwiLFxUT5uRoAAOANQs5JlFQfqkqIsyoiwuLnagAAgDcIOSdRWl4VcrhGDgAAwYeQcxIlTDoGACBoEXJOosReNSeHCwECABB8CDkn4TlcFUfIAQAg2BByTqLmcBVzcgAACD6EnJOoOYU8gZADAEDQIeScRFl1yImPIeQAABBsCDknYa+oDjmxXAgQAIBgQ8g5CUIOAADBi5BzEjUhh1s6AAAQfAg5J2GvqLo5Z3wsc3IAAAg2hJyT4HAVAADBi5BzEhyuAgAgeHn96Z2Tk6P58+dr/fr1iouL0+TJkzV37lzFxsY2uE1JSYleeuklrVmzRjk5OYqKitKgQYN01113adCgQc1qoKW43EYVTrckQg4AAMHIq5Ecm82mWbNmqbS0VIsWLdK9996r9957T/fff/9Jtzt48KDefPNNjR07Vk8//bQWLFggt9ut6dOn69tvv21WAy3FUWk833O4CgCA4OPVp/fy5ctls9m0YsUKJScnS5IiIyM1d+5c3XrrrUpNTa13ux49emjVqlWKi4vzLBs7dqwmTZqkP//5z1qwYEEzWmgZNaM40VERio6K9HM1AADAW16N5KxZs0aZmZmegCNJF154oaxWq1avXt3gdvHx8bUCjiTFxMQoNTVVR44c8bLk1lHhrBrJYRQHAIDg5FXIyc7OrjNaY7ValZKSouzsbK/euKysTNu2bVPfvn292q61MB8HAIDg5tUnuM1mU1JSUp3lSUlJKioq8uqNf//738tut2vGjBlebXciY4zKysqatY8T2e12z0hObHSkz/cfSOx2e63HUBYuvYZLnxK9hqpw6TVc+pTq9mqMkcViafH39ckwhbfFvvfee3rllVf04IMPqlevXs16b6fTqW3btjVrH/VxVFaN5Bh3RYvsP9Ds3r3b3yW0mnDpNVz6lOg1VIVLr+HSp1S7V6vV2uLv51XISUpKks1mq7O8uLi4wUnHJ/riiy903333afbs2bruuuu8eft6RUdHq1+/fs3ez/Hsdrs2ZFed9dWhfZIGDhzo0/0HErvdrt27d6t379515k2FmnDpNVz6lOg1VIVLr+HSp1S316ysrFZ5X69CTmpqap25Nw6HQ3v37tWVV155yu03bdqk2267TRdddJHuuece7yptgMViUXx8vE/2dbyaOTlt4mNaZP+BJi4uLiz6lMKn13DpU6LXUBUuvYZLn9KxXlvjUJXk5cTjCRMmaN26dSooKPAsW7VqlRwOhyZOnHjSbbOzs3XTTTdp5MiRWrBgQas12FSO6jk5TDwGACA4eRVypk+frsTERM2ZM0efffaZVqxYoUcffVRTp06tdbhq3rx5Sk9P9zzPy8vT7NmzFR0drRtvvFHffvutvv76a3399dfaunWr77rxIYerKuTEWLlGDgAAwcjrOTmvvPKK5s+fr9tvv12xsbGaMmWK5s6dW2s9t9stl8vleZ6VlaVDhw5Jkn784x/XWrd79+76+OOPm1h+y3FWTzyOtTKSAwBAMPL6E7xPnz5atmzZSddZuHChFi5c6Hk+evRo7dixw/vq/Kjmtg6xjOQAABCUuAt5A5yVHK4CACCYEXIa4Ak50RyuAgAgGBFyGlAz8ZjDVQAABCdCTgOOTTwm5AAAEIwIOQ04NvGYw1UAAAQjQk4DPHNyYhjJAQAgGBFyGuB0MZIDAEAwI+Q0gOvkAAAQ3Ag59TDGHHcKOSEHAIBgRMiph6P6DuQSFwMEACBYEXLqUeE8dt+tGObkAAAQlAg59Sh3VIWc6KgIRUZY/FwNAABoCkJOPSqqQw6TjgEACF6EnHrUHK5i0jEAAMGLkFMPT8hhJAcAgKBFyKlHzeGqmGh+PAAABCs+xetRUX0KOSM5AAAEL0JOPZyVNWdXEXIAAAhWhJx6VFbftyoqktPHAQAIVoScelS6qg5XRUfy4wEAIFjxKV4PZ2VVyIki5AAAELT4FK+H53BVFIerAAAIVoScetQcrmIkBwCA4MWneD0qOVwFAEDQ41O8HpxdBQBA8CPk1MNzdlUUPx4AAIIVn+L1cDInBwCAoMeneD2YeAwAQPDjU7welZWcQg4AQLAj5NSDkRwAAIIfn+L1IOQAABD8+BSvR80p5Ny7CgCA4MWneD2O3buKOTkAAAQrQk49OFwFAEDw41O8Hp7DVVwMEACAoMWneD1q7l0VyeEqAACCFiGnHhyuAgAg+PEpXg/uXQUAQPDjU7we3IUcAIDgR8ipBzfoBAAg+PEpXo/Y6EhJUmJ8tJ8rAQAATRXl7wIC0R1XD9GmrdlKTor1dykAAKCJCDn16NejrZzFcf4uAwAANAOHqwAAQEgi5AAAgJBEyAEAACGJkAMAAEISIQcAAIQkQg4AAAhJhBwAABCSCDkAACAkEXIAAEBIIuQAAICQRMgBAAAhiZADAABCEiEHAACEJIsxxvi7iKbasGGDjDGyWq0+3a8xRk6nU9HR0bJYLD7dd6Ch19ATLn1K9BqqwqXXcOlTqturw+GQxWLRyJEjW/R9o1p07y2spf5SWCwWnwenQEWvoSdc+pToNVSFS6/h0qdUt1eLxdIqwS6oR3IAAAAawpwcAAAQkgg5AAAgJBFyAABASCLkAACAkETIAQAAIYmQAwAAQhIhBwAAhCRCDgAACEmEHAAAEJIIOQAAICQRcgAAQEgi5AAAgJBEyDlOTk6OZs+ereHDhyszM1Pz589XeXm5v8tqtJUrV2rOnDmaOHGihg8frqlTp+qNN96Q2+2utd7q1at1+eWXa8iQITr//PP1+uuv17u/ZcuW6dxzz9WQIUN05ZVX6quvvmqNNpqktLRUEyZMUP/+/bV58+Zar4VKv2+//bYuvfRSDRkyRJmZmbrllltqvR4KfX700Uf6wQ9+oJEjR2rs2LG67bbbtGvXrjrrBVuve/bs0YMPPqjLLrtM6enpmjJlSr3r+bKvkpISPfjggxo9erRGjBihW265RQcOHPBpX/U5Va8ul0tLlizRjBkzNGbMGGVkZOi6667T2rVr691fMPd6oi1btmjgwIEaMWJEva8Haq+N7bO8vFy/+93vdM4552jw4ME699xz9cwzz9RZrzX7JORUs9lsmjVrlkpLS7Vo0SLde++9eu+993T//ff7u7RGe+mll2S1WvWLX/xCzz33nM477zw99thj+u1vf+tZZ+PGjZozZ47S09O1ZMkSTZs2TfPnz9fbb79da1/Lli3T008/reuuu04vvPCCevXqpZtuukk7duxo7bYa5dlnn5XL5aqzPFT6Xbx4sRYuXKipU6dq2bJleuSRR9S5c2fP66HQ55dffqnbbrtNffr00eLFi/XAAw8oJydH119/vUpKSjzrBWOvO3fu1OrVq9WrVy+lpqbWu46v+7r77rv18ccf64EHHtDTTz+tI0eO6Prrr2/x/7idqtfy8nI9//zzGjBggBYsWKDf/e536tKli66//np98sknIdXr8YwxevTRR5WcnFzv64Hca2P6dLlcuvnmm/XPf/5Td9xxh1588UXdcccdioyM9G+fBsYYY55//nkzbNgwk5eX51n2f//3fyYtLc1kZWX5sbLGO772Go8//rgZMmSIqaioMMYYM3v2bHPVVVfVWuf+++8348aNMy6XyxhjTEVFhRk1apT5zW9+41mnsrLSXHzxxebnP/95C3bQNFlZWWb48OHmL3/5i0lLSzObNm3yvBYK/WZlZZmBAweazz77rMF1QqHPefPmmXPOOce43W7Psm+++cakpaWZTz/91LMsGHutqcsYY+69914zefLkOuv4sq+vv/66zs/twIEDJj093bzxxhs+66s+p+q1srLSFBYW1lrmdrvNtGnTzIwZMzzLQqHX47399tvm/PPPN0899ZQZPnx4rdcCvdfG9Ll8+XJzxhlnmNzc3Ab3448+GcmptmbNGmVmZtZK2RdeeKGsVqtWr17tx8oar77/IQwcOFAVFRUqLCyUw+HQunXrNHny5FrrTJ06Vbm5udq6daskacOGDSouLq41JBkZGalLLrlEq1evljGmZRvx0mOPPabp06erT58+tZaHSr/vvvuuevbsqfHjx9f7eqj0WVlZqYSEBFksFs+yxMTEWusEa68RESf/VevrvlavXq2kpCRNmDDBs163bt00cuTIFv99dqpeIyMj1bZt21rLLBaLBgwYoCNHjniWhUKvNWw2m5566indd999io6OrvN6oPfamD7/+te/6uKLL1bHjh0bXMcffRJyqmVnZ9cZhrNarUpJSVF2drafqmq+9evXq127durQoYP27t0rp9Opvn371lqnX79+kuTps+bxxPVSU1NVWlqqw4cPt0LljfPhhx9q+/bt+ulPf1rntVDp95tvvlFaWpr++Mc/KjMzU4MHD9aMGTO0bds2SaHT51VXXaVdu3bptddek81m0/79+/Wb3/xGqampyszMlBQ6vZ7I131lZ2erT58+tQJjzf4C8feZ2+3Wxo0ba/0ODqVef//732vQoEE655xz6n092Ht1OBzaunWrunbtqnvuuUfDhg3TyJEjdffdd6ugoMCznj/6JORUs9lsSkpKqrM8KSlJRUVFfqio+TZv3qx3331Xs2bNUmRkpKePE/useV7zus1mk9VqVWxsbK31av73VVhY2MKVN47dbtfChQt11113qU2bNnVeD5V+c3Nz9fnnn+u9997Tww8/rMWLF8tut+v666+XzWYLmT4zMjL0zDPP6Omnn1ZGRoYmTZqkffv26cUXX5TVapUUOn+mJ/J1Xzabrc4oWM3+AvH32WuvveaZf1UjVHrdtm2b/vrXv+q+++5rcJ1g77WwsFCVlZVasmSJiouL9cc//lEPPPCAvvzyS911112e9fzRZ5SXvYQdY0ydNBkMcnNzdccdd2jIkCG66aabar3WUD/HL69vnZqhxED5efzpT39Shw4ddMUVV5x0vWDv1xijsrIyLV68WKeffrokadCgQZo0aZLefPNNjRw5UlLw97lhwwbdc889uvLKK3XuueeqpKREzz33nG666Sb95S9/qRVkg73Xhviyr8bsKxD85z//0W9/+1vdcMMNysjIqPVasPdqjNEjjzyia6+99pQTk4O515o6k5KStGjRIs9/ShISEnT77bdr06ZNGjp0aIN1tmSfjORUS0pKks1mq7O8uLi43hGeQFZcXKybbrpJsbGx+tOf/uQ5BlyTlk9MwjV91/SZlJSkiooKVVRU1LveicfT/eHAgQOe2fslJSWy2WwqKyuTJJWVlam0tDRk+m3btq06duzoCTiS1LlzZ/Xt21dZWVkh0+f8+fM1ZswY/epXv1JmZqbOP/98LVmyRDk5OZ4zjEKl1xP5uq+Gfp81NGLtL9u3b9ecOXN03nnn6Z577qn1Wij0+sEHHyg7O1szZ86UzWaTzWbz9HP898Hea817jxw50hNwJGnMmDGSqs7Oqlmvtfsk5FRLTU2tc6zP4XBo7969p0zggaSiokK33nqrjh49qqVLl6p9+/ae11JSUhQdHV3nuiNZWVmS5Omz5vHEn0d2drYSEhLUpUuXlmyhUfbv3y+n06mf/OQnysjIUEZGhue6MT/60Y90/fXXh0y/Df39M8YoIiIiZPrMzs7WgAEDai1LTk5W586dtXfvXkmh9Xf4eL7uKzU1VTk5OXUmWGdlZQXM77O9e/fqxhtvVHp6up544ok6/0MPhV537dqloqIinXvuuZ7fU0uWLFFZWZkyMjK0ePFiScHfa1xcnLp3797g6zUTl/3RJyGn2oQJE7Ru3bpak6RWrVolh8OhiRMn+rGyxqusrNTPfvYzbd++XUuXLq3zl85qtWrMmDFauXJlreXvv/++OnXqpPT0dElVaTwxMVEffPCBZx2Xy6WVK1dq4sSJfh8ClqrOGnv11VdrfdUc83744Yf161//OmT6Pfvss3X06FF99913nmWHDx/Wrl271L9//5Dps1u3bvr2229rLcvNzdWRI0c8f5dDpdcT+bqviRMnymaz6bPPPvOsd+jQIW3YsCEgfp/l5ubqhhtuUMeOHfXss8/W+t9/jVDoddq0aXV+T02bNk0xMTF69dVXdc0110gKjV7PPvtsrV+/Xg6Hw7Psyy+/lCTPf1780qdXJ5yHsKKiInPWWWeZ6dOnmzVr1pi//e1vZvTo0ebuu+/2d2mN9sADD5i0tDSzZMkSs3HjxlpfxcXFxhhjNmzYYNLT082vfvUrs27dOvPss8+aAQMGmLfeeqvWvpYuXWoGDRpkli1bZtauXWvuuusuM2TIELN9+3Z/tNYo69atq3OdnFDot7Ky0kybNs1ccMEF5h//+IdZtWqVufzyy81ZZ51lSktLjTGh0edrr71m0tLSzMMPP2w+//xz88EHH5jLLrvMZGRkmMOHD3vWC8Zey8rKzMqVK83KlSvNjBkzzMSJEz3Pa65v5eu+fvKTn5jx48eb999/33z66adm2rRp5vzzzzd2u92vvdrtdnPppZea4cOHm1WrVtX5XRVKvdZn0aJFda6TY0xg99qYPvfv329GjRplbrzxRvPpp5+at956y4wePdrMmTPHr30Sco6za9cuc8MNN5hhw4aZ0aNHm0cffbTF/5H40jnnnGPS0tLq/Vq3bp1nvU8//dRceumlZtCgQWbSpEnmz3/+c519ud1us2TJEnP22WebwYMHmyuuuMKsXbu2NdvxWn0hx5jQ6Pfo0aPmrrvuMqNGjTLDhg0zN954o8nOzq61TrD36Xa7zfLlyz0fgGPHjjU333xzvaEk2Hrdt29fq//bLC4uNvfff7/JyMgww4cPNzfffLPZv39/i/ZpzKl7PdnraWlpIdVrfRoKOYHca2P73Lx5s7nuuuvMkCFDzJlnnmkeeOABz3+w/dWnxZgAu7IbAACADzAnBwAAhCRCDgAACEmEHAAAEJIIOQAAICQRcgAAQEgi5AAAgJBEyAEAACGJkAO0onfffVf9+/fXkCFDdODAgTqvz5w5U1OmTPFDZQgUhw8f1uLFi7Vt2zZ/lwIEPUIO4AcOh0O///3v/V0GAtCRI0f0zDPPEHIAHyDkAH5w1lln6f3339f27dv9XQoAhCxCDuAHN954o9q1a6ff/va3J11v//796t+/v9599906r/Xv31+LFy/2PF+8eLH69++v7du364477tCoUaN05plnasGCBaqsrNSuXbs0e/ZsjRgxQueee66WLFnSrB7WrFmjWbNmadSoURo2bJguvvhiPf/887XW+fe//61rrrlGw4YN04gRI3T99ddr48aNtdZpbt1fffWV+vfvr7///e9asGCBxo0bp6FDh2rGjBnaunVrnbq9qWnnzp266667NGrUKI0dO1b33XefiouLa61rjNHrr7+uyy67TEOHDlVGRobuuOMO7du3r9Z6NYciN23apGuvvVbDhg3TpEmT9MILL8jtdnt6ueqqqyRJ9913n/r371/rz3nfvn268847NX78eA0ePFhjx47VrFmzGPUBGhDl7wKAcJSQkKBbb71Vjz32mNauXavMzEyf7fvnP/+5Lr30Uk2fPl1ffPGFli5dqsrKSn355Ze69tprNXv2bL333nt68skn1atXL11wwQVev8fbb7+tBx54QBkZGXr44YfVoUMH5eTkaOfOnZ513nvvPc2dO1fjx4/XU089JYfDoaVLl2rmzJl6+eWXdcYZZ/i07qefflrp6emaP3++iouL9cwzz2jmzJlasWKFevbs2aSabr/9dl1yySW66qqr9N133+mpp56SJC1YsMCzzoMPPqi//e1vmjlzpubOnauioiL98Y9/1PTp0/X3v/9dHTt29Kybm5ure+65R9dff71uu+02rVq1Sk899ZQ6d+6syy+/XIMGDdKCBQt033336dZbb9XZZ58tSeratask6aabbpLb7dY999yjbt26qaCgQBs3bpTNZvP6zxAIC17f0hNAk73zzjueO6VXVFSYSZMmmSuuuMK43W5jjDEzZswwkydP9qxfc/ffd955p86+0tLSzKJFizzPFy1aZNLS0syLL75Ya73LLrvMpKWlmX/961+eZU6n04wZM8bcdtttXvdQUlJiRo4caX74wx966j6Ry+Uy48ePN1OmTDEul6vWtpmZmeaaa67xWd01d5+fNm1arXr2799vBg0aZH71q181uaYlS5bUqumhhx4yQ4YM8bzPxo0b66390KFDZujQoeaJJ57wLJsxY4ZJS0sz33zzTa11L7nkEnPDDTd4nm/atKneP/P8/HyTlpZmXn75ZQOgcThcBfiJ1WrVz3/+c23ZskUrV6702X5r/vdfIzU1VRaLRRMmTPAsi4qKUq9eveo9w+tUNm7cqJKSEl177bWyWCz1rpOTk6MjR47osssuU0TEsV8zCQkJuuCCC/TNN9/Ibrf7tO4pU6bUqqd79+4aMWKEvvrqqybXdO6559Z63r9/f1VUVCgvL0+S9Mknn8hisejSSy9VZWWl56tjx44aMGCA/vOf/9TavlOnTho6dGidfR48eLBOPydq166dUlJStGzZMr300kvaunWr5zAXgPoRcgA/mjx5sgYNGqSnn35aTqfTJ/ts27ZtrefR0dGKi4tTTExMneUOh8Pr/efn50s6dgilPgUFBZKqPtRP1LlzZ7nd7jqHWJpb9/GHhY5fVlhY2OSa2rVrV+u51WqVJJWXl0uS8vLyZIzR2LFjNWjQoFpfX3/9tec9G9pfzT4rKirqLD+RxWLRyy+/rPHjx2vp0qWaNm2aMjMzNX/+fJWUlJxyeyAcMScH8COLxaK5c+fq+uuv11tvvVXn9ZoP+BM/1E/88GxNycnJkqTvv/++wXXat28vqWoOyomOHDmiiIgIJSUl+bSuo0eP1rusJli0RE3t27eXxWLR66+/7glAx6tvWXN0795djz/+uKSqkamVK1fqmWeekcPh0COPPOLT9wJCASM5gJ+NHTtW48aN0x//+EeVlpbWeq1jx46KiYnRjh07ai3/97//3Zol1jJixAglJiZq+fLlMsbUu06fPn3UpUsXvf/++7XWKSsr07/+9S8NHz5ccXFxPq3rxPc6cOCANm7cqDPPPLPFajr77LNljNHhw4c1ZMiQOl/9+/f3uo8TR4sa0qdPH82ZM0dpaWn1nkUGgJEcICDMnTtXV1xxhfLy8nT66ad7ltfM93jnnXeUkpKiAQMGaNOmTXr//fdbpI709HRlZGTolVdeaXCdhIQE3Xvvvbr//vv14x//WFdffbU6dOigvXv3avv27XrwwQcVERGhe+65R3PnztXNN9+sa665Rg6HQ8uWLZPNZtPdd9/t89rz8/P105/+VFdffbWKi4u1ePFiWa1W3XzzzZLUIjWNGjVK11xzjebNm6ctW7YoIyNDcXFxys3N1fr165WWlqZrr73Wq32mpKQoNjZW7733nlJTUxUfH6/OnTuroKBAjz76qC666CL16tVL0dHRWrdunXbs2KGf/OQnXtcOhANCDhAA0tPTNXny5HrDyy9/+UtJ0tKlS1VWVqbRo0frueeeqzMp1hdcLlejJrP+4Ac/UOfOnbV06VLdf//9Msaoe/fuuvzyyz3rTJ06VXFxcXrhhRd05513KjIyUsOGDdOrr76qkSNH+rz2O++8U5s3b9Z9992nkpISDR06VL/73e+UkpLSojU98sgjGjZsmN5880395S9/kdvtVufOnTVy5Mg6k4wbIy4uTo8//rieeeYZzZ49W06nU7fddpuuvfZapaSk6I033vAcKuzZs6fuvfdezZw5s0m1A6HOYhoabwaAIPDVV1/pRz/6kf7whz/ooosu8nc5AAIIc3IAAEBIIuQAAICQxOEqAAAQkhjJAQAAIYmQAwAAQhIhBwAAhCRCDgAACEmEHAAAEJIIOQAAICQRcgAAQEgi5AAAgJBEyAEAACHp/wFs6TYvFaAm0AAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.plot(np.cumsum(pca.explained_variance_ratio_))\n",
    "plt.xlabel(\"Num. components\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjYAAAGhCAYAAAByPf5TAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA0oklEQVR4nO3df1DU953H8dcu7sKqoCVVmWgSBQ90lSaSa8TLDxoTp6nAaUJN09HSRrATtWnO1ra2pbmrMSYzsVBrUUtCe4nh5lo6KXOxkmmuuSNpetgb4a5ppL24SNWcx2FagcAGCHzvD283riDyXZbd9cPzMZOhfL+f7/b9fQvLaz/fXw7LsiwBAAAYwBnrAgAAACKFYAMAAIxBsAEAAMYg2AAAAGMQbAAAgDEINgAAwBgEGwAAYIwpsS4gkpqbm2VZllwuV6xLAQAAYzQwMCCHw6Fly5aN+7WMmrGxLEvxdL9By7LU398fVzXFM/plD/2yh37ZQ7/soV/2XNqvSP79NmrGJjBTk52dHeNKLujt7VVLS4sWLlyoqVOnxrqcuEe/7KFf9tAve+iXPfTLnkv79cYbb0TstY2asQEAAJMbwQYAABiDYAMAAIxBsAEAAMYg2AAAAGMQbAAAgDEINgAAwBgEGwAAYAyCDQAAMAbBBgAAGINgAwAAjEGwAQAAxiDYAAAAYxBsAACAMQg2AADAGAQbAABgDIINECVDQ1ZExgAALm9KrAsAJgun06E9Ncd0pr17xPXz5iRr+/qbo1wVAJiFYANE0Zn2bvne7ox1GQBgLA5FAQAAYxBsAACAMQg2AADAGAQbAABgDIINAAAwhq1g85nPfEZZWVkj/vfzn/88OK6hoUFr165Vdna2Vq1apZqamhFfr7q6WitXrlR2draKiop09OjR8e0NAACY1Gxd7v23f/u3evfdd0OWPfvss/rFL36hFStWSJKam5u1ZcsWrVmzRjt27FBTU5N27dolt9utdevWBberrq5WRUWFtm3bJq/Xq9raWm3atEm1tbXKysqKwK4BAIDJxlawWbhw4bBlX/7yl3XrrbcqNTVVklRZWSmv16vdu3dLknJzc3X27Fnt3btXRUVFcjqd6u/v14EDB1RcXKySkhJJ0i233KLCwkIdPHhQFRUV490vAAAwCY3rHJumpiadOXNGhYWFkqT+/n41NjYqPz8/ZFxhYaE6Ojp0/Pjx4Hbd3d0qKCgIjklISNDq1avV0NAgy+K28gAAwL5x3Xn48OHD8ng8uuuuuyRJp06d0sDAgNLT00PGBWZ6fD6fli5dKp/PJ0nDxmVkZKinp0ft7e1KS0sLqybLstTb2xvWtpHm9/tDvmJ0JvfL4XDI4/GMaazf7x9TuDe5XxOBftlDv+yhX/Zc2i/LsuRwOCLy2mEHm/fff18vvfSS7rrrLk2dOlWS1Nl54VbxKSkpIWMD3wfWd3V1ye12KykpKWTcjBkzJEnnz58PO9gMDAyopaUlrG0nSltbW6xLuKqY2C+PxyOv1zumsSdPnrT15mhivyYS/bKHftlDv+y5uF9utzsirxl2sHn99df1zjvvhBxOCrhc6rp4+UhjAp9Sx5PaXC7XiOcCxYLf71dbW5vmz58/5k/rk5nJ/bLzM71gwYIxz9iY2q+JQL/soV/20C97Lu3XiRMnIvbaYQebw4cPa+bMmbrtttuCywIzLoGZmYCuri5JH8zcpKSkqK+vT319fUpMTBw2LvA64XA4HMEZpHjh8XjirqZ4Ntn7ZfdNcbL3yy76ZQ/9sod+2RPoV6QOQ0lhnjz83nvv6Ze//KXuueceuVyu4PLrr79eLpdLra2tIeMDSSwjIyPka+BcmwCfz6dp06Zpzpw54ZQFAAAmubCCzSuvvKKenp7g1VABbrdbubm5qq+vD1l++PBhzZo1K3iOQU5OjpKTk3XkyJHgmMHBQdXX1ysvLy+iyQ0AAEweYR2KevHFF3Xttdfq5ptvHrZu69at2rBhg8rKylRYWKimpibV1tZq586dcjov5Ci3263NmzeroqJCqampwRv0nT59WuXl5ePbIwAAMGnZDjadnZ167bXX9NnPfnbEmZVly5Zp//79Ki8vV11dndLS0lRWVhZy12FJ2rhxoyzL0qFDh3Tu3DllZmaqqqqKuw4DAICw2Q42M2bM0O9+97tRx+Tl5SkvL2/UMQ6HQ6WlpSotLbVbAgAAwIh4ujcAADAGwQYAABiDYAMAAIxBsAEAAMYg2AAAAGMQbAAAgDEINgAAwBgEGwAAYAyCDQAAMAbBBogTM5MTNTRkXXFcYIzD4ZDH4+GhsQBwkbAeggkg8qZ7XHI6HdpTc0xn2rtHHLN4Qao2rcmWJHk8Hnm93hHHDQ1ZcjoJPAAmH4INEGfOtHfL93bniOvmzZ5+xfAzb06ytq+/eSJLBIC4RbABrkKjhR8AmMw4xwYwjN1zdQDAJMzYAIYZy7k6HK4CYCqCDWAoDlcBmIw4FAVEAId1ACA+MGMDRMCVDv3kLJqt4tUjX5oNAIgcgg0QIVe6TBsAMPE4FAUAAIxBsAEAAMYg2AAAAGMQbAAAgDEINgAAwBgEGwAAYAyCDQAAMAbBBgAAGINgAwAAjEGwAQAAxiDYAAAAYxBsAACAMQg2AADAGAQbAABgDIINAAAwBsEGAAAYI6xgU1tbq7/+679Wdna2VqxYoYceeihkfUNDg9auXavs7GytWrVKNTU1I75OdXW1Vq5cqezsbBUVFeno0aPhlANMqKEhK9YlAADGaIrdDfbt26e///u/10MPPaQbb7xRnZ2deu2114Lrm5ubtWXLFq1Zs0Y7duxQU1OTdu3aJbfbrXXr1gXHVVdXq6KiQtu2bZPX61Vtba02bdqk2tpaZWVlRWbvgAhwOh3aU3NMZ9q7R1yfs2i2ild7o1wVAGAktoKNz+fTgQMHVFVVpdtuuy24fNWqVcH/XVlZKa/Xq927d0uScnNzdfbsWe3du1dFRUVyOp3q7+/XgQMHVFxcrJKSEknSLbfcosLCQh08eFAVFRWR2DcgYs60d8v3dueI6+bNnh7lagAAl2PrUNQLL7yg6667LiTUXKy/v1+NjY3Kz88PWV5YWKiOjg4dP35cktTU1KTu7m4VFBQExyQkJGj16tVqaGiQZTH1DwAA7LM1Y/Of//mfyszMVGVlpZ5//nl1d3frpptu0je/+U0tXrxYp06d0sDAgNLT00O2W7hwoaQLMz5Lly6Vz+eTpGHjMjIy1NPTo/b2dqWlpYW1Q5Zlqbe3N6xtI83v94d8xejisV8Oh0MejyfWZUTczOREDQ1Zcjodo44bHBpSf1+fER824vHnK57RL3volz2X9suyLDkco78fjZWtYNPR0aE333xTb731lr797W/L5XLp+9//vh588EH94he/UGfnhan6lJSUkO0C3wfWd3V1ye12KykpKWTcjBkzJEnnz58PO9gMDAyopaUlrG0nSltbW6xLuKrEU788Ho+8XvPOn5nucV3x3KF5c5K1ff3NOnnypFFv1vH083U1oF/20C97Lu6X2+2OyGvaCjaB2ZB9+/bpL/7iLyRJS5Ys0V133aUf//jHysnJkaTLpq6Ll480JvCpcDypzeVyBWeIYs3v96utrU3z58838lN/pMVjvyL1CSJejXbuUMCCBQuMmbGJt5+veEa/7KFf9lzarxMnTkTstW0FmxkzZujDH/5wMNRI0uzZs5Wenq4TJ07ozjvvlPTBzExAV1eXpA9mblJSUtTX16e+vj4lJiYOGxeYuQmHw+HQ1KlTw95+Ing8nrirKZ7Rr/hi2ps0P1/20C976Jc9gX5F8kOkrZOHMzIyRlxuWZacTqeuv/56uVwutba2hqwPJLHA9oGvgXNtAnw+n6ZNm6Y5c+bYKQsAAECSzWDzsY99TOfOndN//dd/BZe1t7ertbVVWVlZcrvdys3NVX19fch2hw8f1qxZs4LnKuTk5Cg5OVlHjhwJjhkcHFR9fb3y8vKMn/4HAAATw9ahqFWrVmnJkiV6+OGH9cgjj8jtdquyslKpqam6//77JUlbt27Vhg0bVFZWpsLCQjU1Nam2tlY7d+6U03khR7ndbm3evFkVFRVKTU0N3qDv9OnTKi8vj/xeAgCAScFWsElISNDTTz+t3bt369FHH9X777+vj370o/rOd74TPKa4bNky7d+/X+Xl5aqrq1NaWprKyspC7josSRs3bpRlWTp06JDOnTunzMxMVVVVcddhAAAQNtuPVLjmmmv0ne98Z9QxeXl5ysvLG3WMw+FQaWmpSktL7ZYAAAAwIp7uDQAAjEGwAQAAxiDYAAAAYxBsAACAMQg2AADAGAQbAABgDIINAAAwBsEGAAAYg2ADAACMQbABAADGINgAAABjEGwAAIAxCDYAAMAYBBsAAGAMgg0AADAGwQYAABiDYAMAAIxBsAEAAMYg2AAAAGMQbAAAgDEINgAAwBgEGwAAYAyCDQAAMAbBBgAAGINgAwAAjEGwAQAAxiDYAAAAYxBsAACAMQg2mNSGhqxYlwAAiKApsS4AiCWn06E9Ncd0pr17xPU5i2areLU3ylUBAMJFsMGkd6a9W763O0dcN2/29ChXAwAYDw5FAQAAYxBsAACAMQg2AADAGAQbAABgDIINAAAwhq1g88ILLygrK2vYf3v27AkZ19DQoLVr1yo7O1urVq1STU3NiK9XXV2tlStXKjs7W0VFRTp69Gj4ewIAACa9sC73fuaZZ5ScnBz8fs6cOcH/3dzcrC1btmjNmjXasWOHmpqatGvXLrndbq1bty44rrq6WhUVFdq2bZu8Xq9qa2u1adMm1dbWKisraxy7BAAAJquwgs2SJUuUmpo64rrKykp5vV7t3r1bkpSbm6uzZ89q7969KioqktPpVH9/vw4cOKDi4mKVlJRIkm655RYVFhbq4MGDqqioCHN3AADAZBbRc2z6+/vV2Nio/Pz8kOWFhYXq6OjQ8ePHJUlNTU3q7u5WQUFBcExCQoJWr16thoYGWRa3uQcAAPaFNWNTUFCgP//5z7r22mt1//33q7S0VAkJCTp16pQGBgaUnp4eMn7hwoWSJJ/Pp6VLl8rn80nSsHEZGRnq6elRe3u70tLSwilNlmWpt7c3rG0jze/3h3zF6KLdL4fDIY/HE5X/r6uZ3+834sMGv4/20C976Jc9l/bLsiw5HI6IvLatYDNr1iw9/PDDuvHGG+VwOPTKK6/ou9/9rtrb2/Xoo4+qs/PCbelTUlJCtgt8H1jf1dUlt9utpKSkkHEzZsyQJJ0/fz7sYDMwMKCWlpawtp0obW1tsS7hqhKtfnk8Hnm9PAfqSk6ePGnUmzW/j/bQL3volz0X98vtdkfkNW0Fm9tvv12333578PvbbrtNiYmJevbZZ/XQQw8Fl18udV28fKQxgU+F40ltLpcrOEMUa36/X21tbZo/fz4zA2MQ7X5F6tOB6RYsWGDMjA2/j2NHv+yhX/Zc2q8TJ05E7LXH/RDMT3ziE/rhD3+olpYWzZ07V9IHMzMBXV1dkj6YuUlJSVFfX5/6+vqUmJg4bFxg5iYcDodDU6dODXv7ieDxeOKupnhGv+KLaW/S/HzZQ7/soV/2BPoVyQ+aET15+Prrr5fL5VJra2vI8kASy8jICPkaONcmwOfzadq0aSGXjwMAAIzVuIPNkSNHlJCQIK/XK7fbrdzcXNXX14eMOXz4sGbNmhU8nyEnJ0fJyck6cuRIcMzg4KDq6+uVl5fHIQIAABAWW4eiSkpKlJubq8zMTEnSL3/5S/3kJz9RcXGxZs2aJUnaunWrNmzYoLKyMhUWFqqpqUm1tbXauXOnnM4LOcrtdmvz5s2qqKhQampq8AZ9p0+fVnl5eYR3EQAATBa2gs2CBQv005/+VP/zP/+joaEhzZ8/X9/4xjf0mc98Jjhm2bJl2r9/v8rLy1VXV6e0tDSVlZWF3HVYkjZu3CjLsnTo0CGdO3dOmZmZqqqq4q7DAAAgbLaCTVlZ2ZjG5eXlKS8vb9QxDodDpaWlKi0ttVMCgCiZmZyooSFLTufoh4bHMgYAomXcV0UBMNN0j0tOp0N7ao7pTHv3iGPmzUnW9vU3R7kyALg8gg2AUZ1p75bv7c4rDwSAOBDRy70BAABiiWADAACMQbABAADGINgAAABjEGwAAIAxCDYAAMAYBBsAAGAMgg0AADAGwQYAABiDYAMAAIxBsAEAAMYg2AAAAGMQbAAAgDEINgAAwBgEGwAAYAyCDQAAMAbBBgAAGINgAwAAjEGwAQAAxiDYwFhDQ1asSwAARNmUWBcATBSn06E9Ncd0pr17xPU5i2areLU3ylUBACYSwQZGO9PeLd/bnSOumzd7epSrAQBMNA5FAQAAYxBsAACAMQg2AADAGAQbAABgDIINgLDNTE4c02X1XHoPIFq4KgpA2KZ7XFe8rH7enGRtX39zlCsDMFkRbACM22iX1QNANHEoCgAAGINgAwAAjEGwAQAAxiDYAAAAY4wr2PT09OiOO+5QVlaW3njjjZB1DQ0NWrt2rbKzs7Vq1SrV1NSM+BrV1dVauXKlsrOzVVRUpKNHj46nJAAAMImNK9js379fg4ODw5Y3Nzdry5Yt8nq9evrpp3Xvvfdq165dqq2tDRlXXV2tiooKrV+/XlVVVbrhhhu0adMm/eEPfxhPWQAAYJIKO9j4fD79wz/8gx5++OFh6yorK+X1erV7927l5uZqy5Yt+uQnP6m9e/dqaGhIktTf368DBw6ouLhYJSUlWrFihZ566inNmzdPBw8eDH+PAADApBV2sHn88cf1wAMPaMGCBSHL+/v71djYqPz8/JDlhYWF6ujo0PHjxyVJTU1N6u7uVkFBQXBMQkKCVq9erYaGBlkWdyoFAAD2hHWDvpdeekm///3v9b3vfU9vvvlmyLpTp05pYGBA6enpIcsXLlwo6cJMz9KlS+Xz+SRp2LiMjAz19PSovb1daWlptmuzLEu9vb22t5sIfr8/5CtGF8l+ORwOeTyecb8OIsfv98f0Awu/j/bQL3volz2X9suyLDkcjoi8tu1g4/f79eSTT+pLX/qSpk+fPmx9Z+eFu4+mpKSELA98H1jf1dUlt9utpKSkkHEzZsyQJJ0/fz6sYDMwMKCWlhbb202ktra2WJdwVYlEvzwej7xe7/iLQcScPHkyLt70+X20h37ZQ7/subhfbrc7Iq9pO9gcOHBA11xzje67775Rx10ueV28fKQxgU904SY3l8sVnB2KNb/fr7a2Ns2fP5/ZgzGIZL8ilfwROenp6VecsZnIGR1+H+2hX/bQL3su7deJEyci9tq2gs3bb7+tH/7wh6qsrNS7774rScHDPr29verp6QnOuARmZgK6urokfTBzk5KSor6+PvX19SkxMXHYuMDr2OVwODR16tSwtp0oHo8n7mqKZ/TLLIEngF86O3upoSFLTufEB1J+vuyhX/bQL3sC/Yrkh1FbwebMmTMaGBjQ5z//+WHriouLdeONN+r555+Xy+VSa2ur7rjjjuD6QBrLyMgI+erz+UIOGfh8Pk2bNk1z5syxvzcA4g5PAAcQTbaCzeLFi/Xcc8+FLGtpadETTzyhb3/728rOzpbb7VZubq7q6+v1uc99Ljju8OHDmjVrVjDE5OTkKDk5WUeOHAkuGxwcVH19vfLy8jiUABiGJ4ADiAZbwSYlJUXLly8fcd2SJUu0ZMkSSdLWrVu1YcMGlZWVqbCwUE1NTaqtrdXOnTvldF64wtztdmvz5s2qqKhQamqqvF6vamtrdfr0aZWXl49ztwAAwGQU1uXeV7Js2TLt379f5eXlqqurU1pamsrKyrRu3bqQcRs3bpRlWTp06JDOnTunzMxMVVVVKSsrayLKAgAAhht3sFm+fPmIj0DIy8tTXl7eqNs6HA6VlpaqtLR0vGUAAADwdG8AAGAOgg0AADAGwQYAABiDYAMAAIxBsAEAAMYg2OCqNDQUu6dEAwDi14TcxwaYaFe6RX/OotkqXs3TvQFgsiHY4Ko12i36582eHuVqAADxgENRAADAGAQbAABgDIINAAAwBsEGAAAYg2ADAACMQbABAADGINgAAABjEGwAAIAxCDYAAMAYBBsAAGAMgg0AADAGwQYAABiDYAMAAIxBsAEAAMYg2AAAAGMQbAAAgDEINgAAwBgEGwAAYAyCDQAAMAbBBgAAGINgAwAAjEGwAQAAxiDYIO44HI5YlwAAuEoRbBBXXC6X3ImJsS4DUTYzOVFDQ9YVx41lDIDJbUqsCwAuNmXKFCU4ndpTc0xn2rtHHJOzaLaKV3ujXBkm0nSPS06nY9R/93lzkrV9/c1RrgzA1YZgg7h0pr1bvrc7R1w3b/b0KFeDaBnt3x0AxoJDUQAAwBi2gs1rr72mDRs2KDc3V0uXLtVdd92lJ554Qt3doVPHDQ0NWrt2rbKzs7Vq1SrV1NSM+HrV1dVauXKlsrOzVVRUpKNHj4a/JwAAYNKzFWw6Ozu1bNkyPfbYY6qurtaDDz6ouro6PfLII8Exzc3N2rJli7xer55++mnde++92rVrl2pra0Neq7q6WhUVFVq/fr2qqqp0ww03aNOmTfrDH/4QmT0DAACTjq1zbAoKClRQUBD8fvny5XK73frWt76l9vZ2zZkzR5WVlfJ6vdq9e7ckKTc3V2fPntXevXtVVFQkp9Op/v5+HThwQMXFxSopKZEk3XLLLSosLNTBgwdVUVERwV0EAACTxbjPsZk5c6Yk6f3331d/f78aGxuVn58fMqawsFAdHR06fvy4JKmpqUnd3d0hISkhIUGrV69WQ0ODLItLOgEAgH1hBZvBwUH19fXpzTffVGVlpe68807NnTtXp06d0sDAgNLT00PGL1y4UJLk8/lCvl46LiMjQz09PWpvbw+nLAAAMMmFdbn3nXfeGQwft99+u8rLyyVdOAdHklJSUkLGB74PrO/q6pLb7VZSUlLIuBkzZkiSzp8/r7S0tHBKk2VZ6u3tDWvbSPP7/SFfMTr6hLHw+/1hzery+2gP/bKHftlzab8sy4rYXefDCjZVVVXq7e3ViRMntH//fj300EP60Y9+FFx/ueIuXj7SmMCb1Xh2bmBgQC0tLWFvPxHa2tpiXcJVw+PxxLoExLmTJ0+O648Hv4/20C976Jc9F/fL7XZH5DXDCjaLFi2SJOXk5Mjr9aqoqEgvv/xy8JBTYGYmoKurS9IHMzcpKSnq6+tTX1+fEi+6fX5gXGDmJhwulytYR6z5/X61tbVp/vz5/MEeA7/fz2FIXNGCBQvCnrHh93Hs6Jc99MueS/t14sSJiL32uO88vHjxYiUkJOjUqVNauXKlXC6XWltbdccddwTHBArOyMgI+erz+eT1fnBrfJ/Pp2nTpmnOnDlh1+NwODR16tSwt58IHo8n7moCrlbj/aPB76M99Mse+mVPoF+RfPjxuK+Kam5u1uDgoObNmye3263c3FzV19eHjDl8+LBmzZoVDDE5OTlKTk7WkSNHgmMGBwdVX1+vvLw8nu4MAADCYmvG5gtf+IKWLl2qrKwsJSUl6fe//72eeeYZZWVl6e6775Ykbd26VRs2bFBZWZkKCwvV1NSk2tpa7dy5U07nhRzldru1efNmVVRUKDU1VV6vV7W1tTp9+nTwRGQAAAC7bAWbj3zkIzpy5IiqqqpkWZbmzp2r+++/XyUlJcGTfpYtW6b9+/ervLxcdXV1SktLU1lZmdatWxfyWhs3bpRlWTp06JDOnTunzMxMVVVVKSsrK3J7BwAAJhVbwebzn/+8Pv/5z19xXF5envLy8kYd43A4VFpaqtLSUjslAAAAXBZP9wYAAMYg2AAAAGMQbAAAgDEINgAAwBgEGwAAYAyCDQAAMAbBBgAAGINgAwAAjEGwAQAAxiDYAAAAYxBsAACAMQg2AK4KM5MTNTRkXXHcWMYAMJeth2ACQKxM97jkdDq0p+aYzrR3jzhm3pxkbV9/c5QrAxBPCDYAripn2rvle7sz1mUAiFMcigIAAMYg2AAAAGMQbAAAgDEINgAAwBgEGwAAYAyCDQAAMAbBBgAAGINgAwAAjEGwQdxwOBxKSkqKdRkAgKsYwQZRNdpzfDwej9LT06NYDQDANDxSAVF1pWf95CyareLV3ihXBQAwBcEGUTfas37mzZ4e5WoAACbhUBQAADAGwQYAABiDYAMAAIxBsAEAAMYg2AAAAGMQbAAAgDEINgAAwBgEGwAAYAyCDYBJweFwyOPxyOFwxLoUABOIOw8DMMbM5EQNDVlyOoeHF4/HI6/3wuM6LjcGwNXPVrCpr6/Xiy++qDfffFOdnZ267rrr9OlPf1oPPPCAnM4PJn8aGhpUUVEhn8+ntLQ0fe5zn9P69euHvV51dbVqamrU0dGhzMxMffWrX9Xy5cvHv1cAJqXpHtcVn0c2b06ytq+/OcqVAYgWW8HmRz/6ka699lp99atf1TXXXKOjR4/q8ccf1+nTp/W1r31NktTc3KwtW7ZozZo12rFjh5qamrRr1y653W6tW7cu+FrV1dWqqKjQtm3b5PV6VVtbq02bNqm2tlZZWVmR3UsAk8pozyMDYDZbwebgwYNKTU0Nfp+bm6ve3l7V1NRo27ZtcrvdqqyslNfr1e7du4Njzp49q71796qoqEhOp1P9/f06cOCAiouLVVJSIkm65ZZbVFhYqIMHD6qioiKCuwgAACYLWycPXxxqAhYvXqy+vj6dP39e/f39amxsVH5+fsiYwsJCdXR06Pjx45KkpqYmdXd3q6CgIDgmISFBq1evVkNDgyzLCmdfAADAJDfuk4ePHTummTNn6pprrtHJkyc1MDCg9PT0kDELFy6UJPl8Pi1dulQ+n0+Sho3LyMhQT0+P2tvblZaWFlY9lmWpt7c3rG0jze/3h3yd7AJXpQDxwO/38yFqFLx/2UO/7Lm0X5ZlReyKxXEFmzfeeEMvvPCCtm7dqoSEBHV2XjimnZKSEjIu8H1gfVdXl9xut5KSkkLGzZgxQ5J0/vz5sIPNwMCAWlpawtp2orS1tcW6hLhw8VUpQKydPHmSP0JjwPuXPfTLnov75Xa7I/KaYQebjo4OffGLX1R2drY2bdoUsu5yqevi5SONCXx6Gk9qc7lcwRmiWPP7/Wpra9P8+fOZqdD4/l2BSFuwYAEzNqPg/cse+mXPpf06ceJExF47rGDT3d2tTZs2KSkpSQcOHJDL5ZL0wYxLYGYmoKurS9IHMzcpKSnq6+tTX1+fEhMTh40LvE44HA6Hpk6dGvb2E8Hj8cRdTcBkxx+fseH9yx76ZU+gX5H84Gv7zsN9fX3avHmzzp07p2eeeUYf+tCHguuuv/56uVwutba2hmwTSGIZGRkhXwPn2gT4fD5NmzZNc+bMsVsWAACAvWDz/vvv65FHHtHvf/97PfPMM5o7d27IerfbrdzcXNXX14csP3z4sGbNmhU8vyInJ0fJyck6cuRIcMzg4KDq6+uVl5fHIQsAABAWW4eidu7cqX/5l3/RV77yFb333nv6j//4j+C6hQsXavr06dq6das2bNigsrIyFRYWqqmpSbW1tdq5c2fw7sRut1ubN29WRUWFUlNTgzfoO336tMrLyyO6gwAAYPKwFWx+9atfSZKeeuqpYeuee+45LV++XMuWLdP+/ftVXl6uuro6paWlqaysLOSuw5K0ceNGWZalQ4cO6dy5c8rMzFRVVRV3HQYAAGGzFWxeeeWVMY3Ly8tTXl7eqGMcDodKS0tVWlpqpwQAAIDLsn3yMAAAQLwi2AAAAGMQbAAAgDEINgAAwBgEGwAAYAyCDQAAMAbBBhEzNMQDBQEAsRX2072BSzmdDu2pOaYz7d0jrs9ZNFvFq71RrgoAMJkQbBBRZ9q75Xu7c8R182ZPj3I1AIDJhkNRACaVmcmJYzpsyqFV4OrEjA2ASWW6x3XFw6bz5iRr+/qbo1wZgEgg2ACYlEY7bArg6sWhKAAAYAyCDQAAMAbBBgAAGINgAwAAjEGwAQAAxiDYAAAAYxBsAACAMQg2AADAGAQbAABgDIINAAAwBsEGAAAYg2ADAACMQbABAADGINgAwCVmJidqaMga09ixjgMQHVNiXQAAxJvpHpecTof21BzTmfbuy46bNydZ29ffHMXKAFwJwQYALuNMe7d8b3fGugwANnAoCgAAGINgAwAAjEGwAQAAxiDYAAAAYxBsAACAMQg2AADAGAQbAABgDNvB5o9//KMeffRRrVmzRl6vVwUFBSOOa2ho0Nq1a5Wdna1Vq1appqZmxHHV1dVauXKlsrOzVVRUpKNHj9otCVHA3VUBAFcD2zfoe+utt9TQ0KAbb7xRQ0NDsqzhf/Cam5u1ZcsWrVmzRjt27FBTU5N27dolt9utdevWBcdVV1eroqJC27Ztk9frVW1trTZt2qTa2lplZWWNb88QUVe6C2vOotkqXu2NclUAAISyHWxWrlypu+++W5K0Y8cO/e53vxs2prKyUl6vV7t375Yk5ebm6uzZs9q7d6+KiorkdDrV39+vAwcOqLi4WCUlJZKkW265RYWFhTp48KAqKirGs1+YAKPdhXXe7OlRrgYAgOFsH4pyOkffpL+/X42NjcrPzw9ZXlhYqI6ODh0/flyS1NTUpO7u7pBDWQkJCVq9erUaGhpGnAkCAAAYTcSfFXXq1CkNDAwoPT09ZPnChQslST6fT0uXLpXP55OkYeMyMjLU09Oj9vZ2paWl2f7/tyxLvb29YVYfWX6/P+Tr1crhcMjj8cS6DCBu+f1+4z6MmfL+FS30y55L+2VZlhwOR0ReO+LBprPzwqGKlJSUkOWB7wPru7q65Ha7lZSUFDJuxowZkqTz58+HFWwGBgbU0tJie7uJ1NbWFusSxsXj8cjr5fwZ4HJOnjxp7B+0q/39K9rolz0X98vtdkfkNSfs6d6XS14XLx9pTOBTT7jJzeVyBWeHYs3v96utrU3z58+/qmc8IpWiAVMtWLDAyBkbE96/ooV+2XNpv06cOBGx1454sAnMuARmZgK6urokfTBzk5KSor6+PvX19SkxMXHYuMDr2OVwODR16tSwtp0oHo8n7moCEDkm/yHj/cse+mVPoF+R/AAd8Rv0XX/99XK5XGptbQ1ZHkhjGRkZIV8D59oE+Hw+TZs2TXPmzIl0aQAAwHARDzZut1u5ubmqr68PWX748GHNmjUreK5GTk6OkpOTdeTIkeCYwcFB1dfXKy8vj8MfAOLezOTEMd28khtcAtFj+1CU3+9XQ0ODJOntt9/Wu+++q5deeknShfvQpKamauvWrdqwYYPKyspUWFiopqYm1dbWaufOncHLxd1utzZv3qyKigqlpqYGb9B3+vRplZeXR3AXAWBiTPe4rnjzynlzkrV9/c1RrgyYvGwHm3feeUePPPJIyLLA988995yWL1+uZcuWaf/+/SovL1ddXZ3S0tJUVlYWctdhSdq4caMsy9KhQ4d07tw5ZWZmqqqqirsOA7iqjHbzSgDRZTvYzJs3T3/4wx+uOC4vL095eXmjjnE4HCotLVVpaandMgAAAIbh6d4AAMAYBBsAAGAMgg0AADAGwQYAABiDYAMAAIxBsAEAAMYg2AAAAGMQbAAAgDEINuA5NgAAY9i+8zDMc6Vn3eQsmq3i1d4oVwUAgH0EG0ga/Vk382ZPj3I1AACEh0NRADCBZiYnjulwL4eEgchgxgYAJtB0j+uKh3vnzUnW9vU3R7kywEwEGwCIgtEO9wKIHA5FAQAAYxBsAACAMQg2AADAGAQbAABgDIINAAAwBsEGAAAYg2BjOG76BZiDG/0BV8Z9bAzHc6CA+Be4O7HT6Rh1HDf6A66MYDMJ8BwoIL6N5e7EgQ8h3OgPGB3BBgDiBB9CgPHjHBsAAGAMgg0AADAGwQYADBE4CflKuHIKJuMcm6vYWK6iADB5jOUkZK6cgukINlcxLuUGMBKunMJkRrC5ynEVBYCJ4HAwG4yrE+fYAMAkMtbzcNyJiXK5XFGoCIgsZmwAYBKxcx7OlCn8icDVh59aAJiEOA8HpuJQFAAAMAbBJk5xnwkAsZaUlMRJxLjqcCgqTnEpN4BYCZxgnJ6ePuo47qWFeBTTYHPy5Ent2rVLx44dk8fjUX5+vrZv366kpKRYlhU3uJQbQCxE8kZ/Ywk/BCREUsyCTVdXlz772c/q2muv1fe+9z396U9/0hNPPKHz589rz549sSoLAPD/RvtwFZjVuVIg4U7IiLaYBZt//Md/VFdXl+rq6pSamipJSkhI0Pbt27V582ZlZGTEqrSIcTgc8ng8HKMGYJyxzOoEDplHIiAxq4OxilmwefXVV7VixYpgqJGkj3/84/rGN76hhoaGuAw2Y/nFGhyylPD/Yzwej7ze4efBXDwGAK5m4z1kbsrzrfggGz8clmXF5PKbFStWqKioSNu3bw9Znp+fr5tuukmPP/647ddsamqSZVkTdrdMh8OhHv+ABi9zxZJrilOexCkRG9P5br/eHxwacUyiK0HTp7oYwxjGxGhMPNZk6hi3y6nkqe4R14XjSn/2IhlOovUnNtqBarz7ZVmW3n//fU2ZMkUOh0MDAwNyOBzKyckZd20xPccmJSVl2PKUlBR1doZ306jAP+xE/gNP81w5NEVqzIzpV/5FZgxjGBPbMfFYk6ljIiWaIcDUGZzx7pfD4ZDb7Q75PlK9irvLvS3LCnvnli1bFuFqAADA1SRmN+hLSUlRV1fXsOXd3d0jzuQAAABcScyCTUZGhnw+X8iy/v5+nTp1Ki5PHAYAAPEvZsHmjjvuUGNjo/785z8Hl7388svq7+9XXl5erMoCAABXsZhdFdXV1aWCggLNnTtXW7Zs0TvvvKMnn3xSt912GzfoAwAAYYlZsJFCH6mQlJSkgoICHqkAAADCFtNgAwAAEEkxO8cGAAAg0gg2AADAGAQbAABgDIINAAAwBsEGAAAYg2ADAACMQbCJsIaGBq1du1bZ2dlatWqVampqrrjNyZMn9dhjj2n16tW66aabdOedd+ob3/iGOjo6olBxbIXTL0mqrKzUgw8+qJtvvllZWVl64403JrjS6Dp58qRKSkp00003acWKFdq1a5fee++9MW37s5/9TPfcc4+ys7NVUFCg+vr6Ca429sLt15EjR/Twww/r9ttvV1ZWlqqrq6NQbeyF0693331X+/bt07p16/SXf/mXys3NVUlJid58880oVR074f58PfXUU8rPz9eyZcuUk5OjoqIi/fznP49CxbE1nvevgJdffllZWVkqKCiw/f9PsImg5uZmbdmyRV6vV08//bTuvfde7dq1S7W1taNu9/rrr+s3v/mN7r//fv3gBz/Q3/zN3+jf//3f9alPfUo9PT1Rqj76wu2XJP34xz/WwMCAbr311ihUGl1dXV367Gc/q56eHn3ve9/T1772Nb344osqKyu74rYvvfSSduzYoVWrVunpp59Wbm6utm3bpl/96ldRqDw2xtuv06dP684774xCpfEh3H7993//t3784x/rr/7qr1RRUaEnnnhCQ0NDeuCBB4wON+P5+fL7/XrggQe0b98+7d27V4sXL9aXvvQlvfjii1GoPDbG06+A9957T0888YQ+/OEPh1eEhYgpKSmxPvnJT4YsKysrs2699VZrcHDwstu988471tDQUMiylpYWKzMz03rhhRcmpNZ4EG6/LMsKrm9sbLQyMzOt3/72txNWZ7T94Ac/sG688UbrnXfeCS77p3/6JyszM9M6ceLEqNvec8891he/+MWQZRs3brTWrVs3IbXGg/H06+Kfs8zMTOuZZ56ZsDrjRbj96unpsXp7e0OWvffee9att95q7dixY8LqjbXx/HyN5FOf+pT14IMPRrLEuBKJfn33u9+11q9fb33ta1+z8vPzbdfAjE2E9Pf3q7GxUfn5+SHLCwsL1dHRoePHj19229TUVDkcjpBlWVlZSkhI0P/+7/9OSL2xNp5+SZLTae6P7quvvqoVK1YoNTU1uOzjH/+43G63GhoaLrvd6dOn1draOmzqtqCgQL/97W/1pz/9acJqjqVw+yWZ/XN0OeH2a+rUqfJ4PCHLEhMTlZGRYez7lDS+n6+RzJw5UwMDA5EsMa6Mt1+nTp3Sj370I1szPJeafL/VE+TUqVMaGBhQenp6yPKFCxdKknw+n63Xa25u1uDgoDIyMiJWYzyJdL9M4vP5hv27u91uXX/99aP2pbW1VZKG9TQjI0OWZQXXmybcfk1WkexXb2+vWlpahv3MmWS8/bIsS++//766urpUV1en119/XevXr5+ocmNuvP16/PHHtWbNGi1atCjsGqaEvSVCdHZ2SpJSUlJClge+D6wfi4GBAe3evVsLFizQxz72sYjVGE8i2S/TdHV1DeuLdKE3o/Xlcj2dMWNGyHrThNuvySqS/frud78rv9+vDRs2RKq8uDPefv3bv/2bHnzwQUnSlClT9K1vfUv33HNPxOuMF+Pp1yuvvKLm5ma99NJL46qBYDOK7u7uMU2xXnfddcH/fekhpSstH8ljjz2mt956S88//7ymTLl6/oli1a/JwrKsMfXl0jHW/z/ndrL1dKz9wgV2+/Xiiy/q2Wef1aOPPqobbrhhAiuLT2Pt10c+8hH99Kc/1bvvvqtXX31Vjz32mBISErRu3booVBk/rtSvvr4+7d69Ww8//HDIYaxwXD1/NWPg5Zdf1te//vUrjqurq7vsp+Kuri5Jwz9FX873v/99/fSnP9W+ffuUnZ1ts+LYikW/TJSSkhLsw8W6u7tHPTR5cU8vvprA9J6G26/JKhL9ev311/X1r39dJSUlRh9Wkcbfr+nTpwffy1esWKH+/n49+eSTuu+++5SQkBDxemMt3H49++yzcjqdys/PD24/MDCgoaEhdXV1KSkpSW63e0w1EGxGcd999+m+++4b09j+/n65XC61trbqjjvuCC4/ceKEJI3pF6Cmpkb79u3Tzp07ddddd4VXdAxFu1+mysjIGHYsur+/X6dOnVJRUdFltwuc59Da2hrSP5/PJ4fDYex5EOH2a7Iab79++9vf6gtf+ILuuecefeUrX5moMuNGpH++lixZoueff15/+tOfNGvWrEiVGTfC7Vdra6v++Mc/asWKFcPWffSjH9Xf/d3f6dOf/vSYauDk4Qhxu93Kzc0ddjO0w4cPa9asWfJ6vaNu//Of/1y7du3SF7/4RX3qU5+ayFLjwnj7ZbI77rhDjY2N+vOf/xxc9vLLL6u/v195eXmX3e66665Tenq6jhw5ErL88OHD+shHPjLu6d14FW6/Jqvx9Mvn82nTpk3KycnRE088MSkO9UX65+vYsWOaPn26PvShD0WyzLgRbr82bdqk5557LuS/2267TXPnztVzzz2nlStXjr0I2xeI47Kamposr9drffOb37QaGxut/fv3W4sWLbJ+8pOfhIy7++67reLi4uD3R48etZYsWWJt2LDBam5uDvnvj3/8Y7R3I2rC7ZdlXehZfX29tW/fPiszM9Oqqqqy6uvrjbifTWdnp3X77bdbDzzwgPXqq69aP/vZz6zly5dbX/7yl0PGff3rX7cWL14csuzIkSNWVlaWVV5ebjU2NlqPP/64lZWVZb322mvR3IWoGk+/3nrrLau+vt6qr6+3MjMzra9+9atWfX299a//+q/R3IWoCrdf586ds/Ly8qxbb73V+vWvfx3yPvXmm29GezeiJtx+tbS0WCUlJVZtba3161//2vrnf/5n65vf/KaVmZlp/eAHP4j2bkTNeH4fLxXufWw4FBVBy5Yt0/79+1VeXq66ujqlpaWprKxs2Elig4ODGhoaCn5/9OhRDQwM6De/+c2w2Zp7771XTz75ZFTqj7Zw+yVJ+/bt029+85vg93v27JFkRr9SUlL07LPPateuXXr44YeVlJSkgoICbd++PWTc0NCQBgcHQ5Z94hOf0HvvvaeDBw+qurpaN9xwgyoqKnTbbbdFcxeiajz9qq+v1/e///3g93V1daqrq9PcuXP1yiuvRKX+aAu3XydOnNDZs2clSZ/73OdCxtKv4f368Ic/rJSUFO3fv18dHR1KTk5Wenq6Kisrdffdd0d7N6JmPL+PkeKwrP+/ZAIAAOAqxzk2AADAGAQbAABgDIINAAAwBsEGAAAYg2ADAACMQbABAADGINgAAABjEGwAAIAxCDYAAMAYBBsAAGAMgg0AADDG/wGAc9NabxgNHQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query_sims_pca = cosine_similarity(emb_pcs, pca.transform(np.atleast_2d(query_emb)))\n",
    "plt.hist(query_sims_pca, bins=50);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.397\n",
      "In high-throughput studies, hundreds to millions of hypotheses are typically tested. Statistical methods that control the false discovery rate (FDR) have emerged as popular and powerful tools for error rate control. While classic FDR methods use only p values as input, more modern FDR methods have been shown to increase power by incorporating complementary information as informative covariates to prioritize, weight, and group hypotheses. However, there is currently no consensus on how the modern methods compare to one another. We investigate the accuracy, applicability, and ease of use of two classic and six modern FDR-controlling methods by performing a systematic benchmark comparison using simulation studies as well as six case studies in computational biology.\n",
      "Methods that incorporate informative covariates are modestly more powerful than classic approaches, and do not underperform classic approaches, even when the covariate is completely uninformative. The majority of methods are successful at controlling the FDR, with the exception of two modern methods under certain settings. Furthermore, we find that the improvement of the modern FDR methods over the classic methods increases with the informativeness of the covariate, total number of hypothesis tests, and proportion of truly non-null hypotheses.\n",
      "Modern FDR methods that use an informative covariate provide advantages over classic FDR-controlling procedures, with the relative gain dependent on the application and informativeness of available covariates. We present our findings as a practical guide and provide recommendations to aid researchers in their choice of methods to correct for false discoveries.\n",
      "\n",
      "0.383\n",
      "Using the model host/pathogen pair \n",
      "\n",
      "0.359\n",
      "Comparing multiple single-cell expression datasets such as cytometry and scRNA-seq data between case and control donors provides information to elucidate the mechanisms of disease. We propose a completely data-driven computational biological method for this task. This overcomes the challenges of conventional cellular subset-based comparisons and facilitates further analyses such as machine learning and gene set analysis of single-cell expression datasets.\n",
      "\n",
      "0.323\n",
      "A time course experiment is a widely used design in the study of cellular processes such as differentiation or response to stimuli. In this paper, we propose \n",
      "\n",
      "0.322\n",
      "Large-scale multiparameter screening has become increasingly feasible and straightforward to perform thanks to developments in technologies such as high-content microscopy and high-throughput flow cytometry. The automated toolkits for analyzing similarities and differences between large numbers of tested conditions have not kept pace with these technological developments. Thus, effective analysis of multiparameter screening datasets becomes a bottleneck and a limiting factor in unbiased interpretation of results. Here we introduce compaRe, a toolkit for large-scale multiparameter data analysis, which integrates quality control, data bias correction, and data visualization methods with a mass-aware gridding algorithm-based similarity analysis providing a much faster and more robust analyses than existing methods. Using mass and flow cytometry data from acute myeloid leukemia and myelodysplastic syndrome patients, we show that compaRe can reveal interpatient heterogeneity and recognizable phenotypic profiles. By applying compaRe to high-throughput flow cytometry drug response data in AML models, we robustly identified multiple types of both deep and subtle phenotypic response patterns, highlighting how this analysis could be used for therapeutic discoveries. In conclusion, compaRe is a toolkit that uniquely allows for automated, rapid, and precise comparisons of large-scale multiparameter datasets, including high-throughput screens.\n",
      "Biology has seen huge advances in technology in recent years. This has led to state-of-the-art techniques which can test hundreds of conditions simultaneously, such as how cancer cells respond to different drugs. In addition to this, each of the tens of thousands of cells studied can be screened for multiple variables, such as certain proteins or genes. This generates massive datasets with large numbers of parameters, which researchers can use to find similarities and differences between the tested conditions. Analyzing these ‘high-throughput’ experiments, however, is no easy task, as the data is often contaminated with meaningless information, or ‘background noise’, as well as sources of bias, such as non-biological variations between experiments. As a result, most analysis methods can only probe one parameter at a time, or are unautomated and require manual interpretation of the data. Here, Chalabi Hajkarim et al. have developed a new toolkit that can analyze multiparameter datasets faster and more robustly than current methods. The kit, which was named ‘compaRe’, combines a range of computational tools that automatically ‘clean’ the data of background noise or bias: the different conditions are then compared and any similarities are visually displayed using a graphical interface that is easy to explore. Chalabi Hajkarim et al. used their new method to study data from patients with acute myeloid leukemia (AML) and myelodysplastic syndrome, two forms of cancer that disrupt the production of functional immune cells. The toolkit was able to identify subtle differences between the patients and categorize them into groups based on the proteins present on immune cells. Chalabi Hajkarim et al. also applied compaRe to high-throughput data on cells from patients and mouse models with AML that had been treated with large numbers of specific drugs. This revealed that different cell types in the samples responded to the treatments in distinct ways. These findings suggest that the toolkit created by Chalabi Hajkarim et al. can automatically, rapidly and precisely compare large multiparameter datasets collected using high-throughput screens. In the future, compaRe could be used to identify drugs that illicit a specific response, or to predict how newly developed treatments impact different cell types in the body.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sdf = df[[\"abstract\"]].reset_index(drop=True).join(DataFrame(dict(sim=query_sims_pca[:,0]))).sort_values('sim', ascending=False)\n",
    "for i in range(5):\n",
    "    print(round(sdf.sim.values[i], 3))\n",
    "    print(sdf.abstract.values[i])\n",
    "    print()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "GPTAreaOverview",
   "language": "python",
   "name": "gptareaoverview"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "vscode": {
   "interpreter": {
    "hash": "e7370f93d1d0cde622a1f8e1c04877d8463912d04d973331ad4851f04de6915a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
